{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 1 - _Foundations of Information Retrieval 2023_\n",
    "\n",
    "This assignment is divided in 4 parts, which have to be delivered all together no later than 01/10/2023 at 23:59 (strict - no extensions will be granted!) via Canvas. Delivery of the assignment solutions is mandatory (_see grading conditions on Canvas and in slides of Lecture01_).\n",
    "\n",
    "We will use [ElasticSearch](https://www.elastic.co/) as search engine. It provides state-of-the-art tools to implement your own engine, index your documents, and let you focus on methodological aspects of search models and optimization. \n",
    "\n",
    "The assignment is about text-based Information Retrieval and it is structured in three parts:\n",
    "1. IR performance evaluation (implementation of performance metrics)\n",
    "2. Setting up a search engine, pre-processing and indexing using ElasticSearch (Indexing, Analyzers)\n",
    "3. Implementation and optimization of models of search (Similarity)\n",
    "\n",
    "\n",
    "This assignment contains exercises, marked with the section title __Exercise 01.(x)__, which are evaluated, and other sections that contain support code which you should study and use as it is. Write your answers between the comments `BEGIN ANSWER` and `END ANSWER`. \n",
    "\n",
    "_Note:_ the comment `#THIS IS GRADED!` in a section indicates that it will be graded.\n",
    "\n",
    "\n",
    "### Initial preparation (self-study)\n",
    "For the first part, it is good to acquire (or refresh) basic knowledge of Python. Please use the [Python tutorials](https://docs.python.org/3/tutorial/) if needed.\n",
    "\n",
    "For the second and third part of the assignment, please study yourself the [Getting Started guide](https://www.elastic.co/guide/en/elasticsearch/reference/current/getting-started.html) of ElasticSearch and get acquainted with the framework.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "***\n",
    "\n",
    "# PART 01 - Performance evaluation\n",
    "\n",
    "\n",
    "### Background information and reading\n",
    "Study the slides of Lecture 01 (available on Canvas) and the reference book chapter (Christopher D. Manning, Prabhakar Raghavan and Hinrich Schütze, Introduction to Information Retrieval, [Chapter 8, Evaluation in information retrieval](http://nlp.stanford.edu/IR-book/pdf/08eval.pdf), Cambridge University Press. 2008)\n",
    "\n",
    "### Basic concepts\n",
    "Suppose the set of relevant documents (the document identifiers - _doc-IDs_) is called `relevant`, then we  define it as follows (in Python):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "relevant = set([2, 3, 5, 8, 13, 17, 21, 34, 38])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A perfect run would retrieve exactly these 9 documents in any order. Now, suppose the list of retrieved documents (the document identifiers - _doc-IDs_) is called `retrieved`, and contains the following _doc-IDs_:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "retrieved = [14, 4, 2, 18, 16, 8, 46, 32, 17, 34, 33, 22, 47, 39, 11]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One of the simplest evaluation measures is the _Success at rank 1_, i.e. `Is the first document retrieved a relevant document?`\n",
    "\n",
    "_Success at rank 1_ returns 1 if the first document is relevant, and 0 otherwise. A possible implementation is: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def success_at_1 (relevant, retrieved):\n",
    "    if len(retrieved) > 0 and retrieved[0] in relevant:\n",
    "        return 1\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "success_at_1(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first retrieved documentid is 14 which is not in the set of relevant documents, so the `success_at_1` is 0.\n",
    "\n",
    "_________________\n",
    "\n",
    "> Note how easy it is to check if an item occurs in a Python set or list by using the keyword: `in`. Similarly, you can loop over all items in a set of list with: \n",
    "`for doc in retrieved:`, \n",
    "where doc will refer to each item in the set or list. \n",
    "\n",
    "Be sure to use the internet to sharpen your knowledge about Python constructs, for instance on [Python list slicing](https://duckduckgo.com/?q=python+list+slicing). Also note that the code above checks if at least one document is retrieved to avoid an index out of bounds exception (i.e. we avoid to access an empty vector).\n",
    "\n",
    "> ___Suggestion:___ _to be sure of the correctness of the implementation of the performance metrics, you can compute their values manually and compare them with those computed by your functions. This is important, as you will use these metrics for later exercises and to compare the results of differentmodels._"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation exercise: _Success at k_\n",
    "The measure _Success at k_ returns 1 if a relevant document is among the first _k_ documents retrieved and zero otherwise.\n",
    "\n",
    "> Success at _k_ measures are well-suited in case there is typically only one relevant document (or retrieving one relevant document is enough).\n",
    "\n",
    " __Implement _Success at 5_ below.__ \n",
    " > The correct result is 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def success_at_5(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    top_5_retrieved = retrieved[:5]\n",
    "    \n",
    "    for doc in top_5_retrieved:\n",
    "        if doc in relevant:\n",
    "            return 1\n",
    "    return 0\n",
    "    # END ANSWER\n",
    "    \n",
    "success_at_5(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly __implement success at rank 10__\n",
    "\n",
    "> The correct result is 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def success_at_10(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    top_10_retrieved = retrieved[:10]\n",
    "    \n",
    "    for doc in top_10_retrieved:\n",
    "        if doc in relevant:\n",
    "            return 1\n",
    "    return 0\n",
    "    # END ANSWER\n",
    "    \n",
    "success_at_10(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 01.A: _Precision, Recall and F-measure_\n",
    "__1. Implement _Precision_ using Formula 8.1 from [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book).__\n",
    "\n",
    ">_Hint:_ one can count the number of documents in a list using the built-in Python function [len()](https://docs.python.org/3/library/functions.html#len) \\\n",
    "> _example:_ `len(retrieved)` for the number of retrieved documents. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.26666666666666666"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def precision(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "\n",
    "    if len(retrieved )== 0:\n",
    "        return 0.0  # To handle the case where no documents are retrieved, thus avoiding division by zero\n",
    "\n",
    "    tp = len(set(retrieved).intersection(relevant))\n",
    "    return tp / len(retrieved)\n",
    "\n",
    "    # END ANSWER\n",
    "\n",
    "precision(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__2. Implement _Recall_ using Formula 8.2 from [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book).__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.4444444444444444"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def recall(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    if len(relevant) == 0:\n",
    "        return 0  # Return 0 if there are no relevant documents\n",
    "    \n",
    "    relevant_and_retrieved = set(relevant).intersection(set(retrieved))\n",
    "    return len(relevant_and_retrieved) / len(relevant)\n",
    "    # END ANSWER\n",
    "    \n",
    "recall(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__3. Implement the balanced F measure (_F_ with β=1) using Formula 8.6 from [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book).__\n",
    "\n",
    "> Tip: you may reuse your implementations of precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.33333333333333337"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def f_measure(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    p = precision(relevant, retrieved)\n",
    "    r = recall(relevant, retrieved)\n",
    "    \n",
    "    if p + r == 0:\n",
    "        return 0\n",
    "    \n",
    "    f1_score = (2 * p * r) / (p + r)\n",
    "    return f1_score\n",
    "    # END ANSWER\n",
    "    \n",
    "f_measure(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 01.B: _Precision at rank k_ and  _R-Precision_\n",
    "\n",
    "Precision, Recall and F are _set_-based measures and suited for unranked lists of documents. If our search system returns a ranked _list_ of results, we can measure precision for several cut-off levels _k_ in the ranked list, i.e. we evaluate the relevance of the TOP-_k_ retrieved documents _(see lecture 01 slides and the book chapter)_. \n",
    "\n",
    "\n",
    "**1. Implement the function `precision_at_k()` that measures the precision at rank _k_**\n",
    "\n",
    "> Interesting fact: For _k_=1, the _Precision at rank 1_ would be the samen as _Success at rank 1_ (why?) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pr@1: 0.00\n",
      "Pr@5: 0.20\n",
      "Pr@10: 0.40\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def precision_at_k(relevant, retrieved, k):\n",
    "    # BEGIN ANSWER\n",
    "    top_k_retrieved = retrieved[:k]\n",
    "    \n",
    "    if len(top_k_retrieved) == 0:\n",
    "        return 0\n",
    "\n",
    "    relevant_and_top_k = set(relevant).intersection(set(top_k_retrieved))\n",
    "    return len(relevant_and_top_k) / k\n",
    "    # END ANSWER\n",
    "\n",
    "# Example usage\n",
    "# relevant = ['doc1', 'doc2', 'doc3', 'doc4']\n",
    "# retrieved = ['doc1', 'doc5', 'doc2', 'doc6', 'doc7', 'doc3', 'doc8', 'doc9', 'doc10']\n",
    "\n",
    "print('Pr@1: %1.2f' % precision_at_k(relevant, retrieved, k=1))\n",
    "print('Pr@5: %1.2f' % precision_at_k(relevant, retrieved, k=5))\n",
    "print('Pr@10: %1.2f' % precision_at_k(relevant, retrieved, k=10))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__2. Implement R-Precision as defined in Chapter 8 of [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book)__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.3333333333333333"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def r_precision(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    R = len(relevant)\n",
    "    \n",
    "    if R == 0:\n",
    "        return 0\n",
    "    \n",
    "    top_R_retrieved = retrieved[:R]\n",
    "    relevant_and_top_R = set(relevant).intersection(set(top_R_retrieved))\n",
    "    \n",
    "    return len(relevant_and_top_R) / R\n",
    "    # END ANSWER\n",
    "    \n",
    "r_precision(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 01.D:  Interpolated precision at _recall_ X\n",
    "\n",
    "Another way to address ranked retrieval is to measure precision for several _recall_ levels _X_.\n",
    "\n",
    "__Implement the function `interpolated_precision_at_recall_X()` that measures the interpolated precision at recall level _X_ as defined by formula 8.7 of [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book).__\n",
    "\n",
    "> Tip: calculate for each rank the recall. If the recall is greater than or equal to X, \n",
    "> calculate the precision. Keep the highest (maximum) precision of those to be returned at the end."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pr_i@Re01: 0.40\n",
      "Pr_i@Re02: 0.40\n",
      "Pr_i@Re03: 0.40\n",
      "Pr_i@Re04: 0.40\n",
      "Pr_i@Re05: 0.00\n",
      "Pr_i@Re06: 0.00\n",
      "Pr_i@Re07: 0.00\n",
      "Pr_i@Re08: 0.00\n",
      "Pr_i@Re09: 0.00\n",
      "Pr_i@Re10: 0.00\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def interpolated_precision_at_recall_X (relevant, retrieved, X):\n",
    "    # BEGIN ANSWER\n",
    "    max_precision = 0\n",
    "    relevant_count = 0\n",
    "    total_relevant = len(relevant)\n",
    "    \n",
    "    for i, doc in enumerate(retrieved):\n",
    "        if doc in relevant:\n",
    "            current_recall = recall(relevant, retrieved[:i+1])\n",
    "            if current_recall >= X:\n",
    "                current_precision = precision(relevant, retrieved[:i+1])\n",
    "                max_precision = max(max_precision, current_precision)\n",
    "    \n",
    "    return max_precision\n",
    "    # END ANSWER\n",
    "    \n",
    " \n",
    "\n",
    "print('Pr_i@Re01: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.1))\n",
    "print('Pr_i@Re02: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.2))\n",
    "print('Pr_i@Re03: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.3))\n",
    "print('Pr_i@Re04: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.4))\n",
    "print('Pr_i@Re05: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.5))\n",
    "print('Pr_i@Re06: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.6))\n",
    "print('Pr_i@Re07: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.7))\n",
    "print('Pr_i@Re08: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.8))\n",
    "print('Pr_i@Re09: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.9))\n",
    "print('Pr_i@Re10: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 01.E:  _Average Precision_\n",
    "\n",
    "For a single information need, _Average Precision_ is the average of the precision value obtained for the set of top k documents existing after each relevant document is retrieved (see [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book), Pages 159 and 160). \n",
    "\n",
    "__Implement _Average Precision_ for a single information need.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.1835978835978836"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def average_precision(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    sum_precision = 0\n",
    "    R = len(relevant)\n",
    "    \n",
    "    if R == 0:\n",
    "        return 0\n",
    "    \n",
    "    for i in range(len(relevant)):\n",
    "#         if doc in relevant:\n",
    "        sum_precision += precision(relevant, retrieved[:i])\n",
    "    \n",
    "    return sum_precision / R\n",
    "    # END ANSWER\n",
    "\n",
    "average_precision(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Performance measures in TREC benchmarks\n",
    "\n",
    "The relevance judgments are provided by TREC in so-called _\"qrels\"_ files that look as follows:\n",
    "\n",
    "    1000 Q0 1341 1\n",
    "    1000 Q0 1231 0\n",
    "    1001 Q0 12332 1\n",
    "     ...\n",
    "\n",
    "The columns of the _qrels_ file contain:\n",
    "1. the query identifier\n",
    "2. the query number within that topic (currently unused and should always be Q0)\n",
    "3. the document identifier that was examined by the judges\n",
    "4. the relevance of the document (_1_:relevant; _0_: not relevant).\n",
    "\n",
    "Below we provide some Python code that reads the _qrels_ and the _run_. The qrels will be put in the Python dictionary `all_relevant`. A [Python dictionary](https://docs.python.org/3/tutorial/datastructures.html#dictionaries) provides quick lookup of a set of values given a key. We will use the `query_id` as a key, and a [Python set](https://docs.python.org/3/tutorial/datastructures.html#sets) of relevant document identifiers. For the partial qrels file above, `all_relevant` would look as follows:\n",
    "\n",
    "    {\n",
    "        \"1000\": set([\"1341\", \"1231\"]),\n",
    "        \"1001\": set([\"12332\"])\n",
    "    }\n",
    "    \n",
    "We will use a dictionary called `all_retrieved` with `query_id` as key, and as value a [Python list](https://docs.python.org/3/tutorial/introduction.html#lists) of document identifiers retrieved by the IR system:\n",
    "\n",
    "    {\n",
    "        \"1000\": [\"1341\", \"12346, \"2345\"],\n",
    "        \"1001\": [..., ..., ...],\n",
    "        ...\n",
    "    }\n",
    "\n",
    "Note that, with this data structure, for each `query_id` we can easily access the list of retrieved and relevant documents, and compute the performance metrics. We can then average these measures over all the queries to compute the mean performance of the IR system on the given retrieval task.\n",
    "\n",
    "Please examine the code below, and make sure you understand every line. Use the Python documentation where needed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DATA: the TREC genomics benchmark\n",
    "\n",
    "For the following exercises, we will use a subset of the TREC genomics document collection and queries. \n",
    "It is stored in the folder `data01/` in the directory where you have been instructed to place the assignment notebooks (`/`).\n",
    "\n",
    "The collections contains:\n",
    "\n",
    "* `FIR-s05-medline.json` (the collection in Elasticsearch batch format - because of its size it cannot be indexed with a single curl command!)\n",
    "* `FIR-s05-training-queries-simple.txt` (test queries)\n",
    "* `FIR-s05-training-qrels.txt` (the \"relevance judgements\" for the test queries, i.e. the correct answers)\n",
    "\n",
    "> ___Note___ that these files contain a subset of the documents and queries of the TREC genomics track benchmark, to facilitate experimentations with less computation time needed.\n",
    "> The original files are also included in the `data01/` directory, withouth the `FIR-s05-` prefix (you may use them for the final project).\n",
    "\n",
    "To make things easy, the data is already provided in Elasticsearch' batch processing format. \n",
    "Inspect the collection file in the terminal:\n",
    "\n",
    "`head FIR-s05-medline.json`\n",
    "\n",
    "This shows the first 5 documents in the collection (in JSON format prepared for ElasticSearch, as you have seen in the tutorial)\n",
    "\n",
    "#### Baseline model and results\n",
    "We also provide the list of retrieved documents by a _baseline_ model, in the file `data01/baseline.run`. For each query, it contains the list of document IDs of the retrieved documents (to be compared with those in the qrels file). We use this file in the examples and evaluation exercises below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_qrels_file(qrels_file):  # reads the content of he qrels file\n",
    "    trec_relevant = dict()  # query_id -> set([docid1, docid2, ...])\n",
    "    with open(qrels_file, 'r') as qrels:\n",
    "        for line in qrels:\n",
    "            (qid, q0, doc_id, rel) = line.strip().split()\n",
    "            if qid not in trec_relevant:\n",
    "                trec_relevant[qid] = set()\n",
    "            if (rel == \"1\"):\n",
    "                trec_relevant[qid].add(doc_id)\n",
    "    return trec_relevant\n",
    "\n",
    "def read_run_file(run_file):  \n",
    "    # read the content of the run file produced by our IR system \n",
    "    # (in the following exercises you will create your own run_files)\n",
    "    trec_retrieved = dict()  # query_id -> [docid1, docid2, ...]\n",
    "    with open(run_file, 'r') as run:\n",
    "        for line in run:\n",
    "            (qid, q0, doc_id, rank, score, tag) = line.strip().split()\n",
    "            if qid not in trec_retrieved:\n",
    "                trec_retrieved[qid] = []\n",
    "            trec_retrieved[qid].append(doc_id) \n",
    "    return trec_retrieved\n",
    "    \n",
    "\n",
    "def read_eval_files(qrels_file, run_file):\n",
    "    return read_qrels_file(qrels_file), read_run_file(run_file)\n",
    "\n",
    "(all_relevant, all_retrieved) = read_eval_files('data01/FIR-s05-training-qrels.txt', 'data01/baseline.run')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### _Number of queries_ and _number of retrieved documents per query_\n",
    " \n",
    "The following code counts the number of queries evaluated in the file `baseline.run` (provided in the `data01/` folder, containing the list of doc-ids retrieved using a baseline model) and prints it (use the result from the cell above). For each query, it also prints the number of documents that were retrieved for that query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of retrieved documents: 38\n",
      "Docs retrieved for query #1: 1000\n",
      "Docs retrieved for query #3: 1000\n",
      "Docs retrieved for query #4: 1000\n",
      "Docs retrieved for query #5: 1000\n",
      "Docs retrieved for query #6: 1000\n",
      "Docs retrieved for query #7: 1000\n",
      "Docs retrieved for query #8: 1000\n",
      "Docs retrieved for query #9: 1000\n",
      "Docs retrieved for query #10: 1000\n",
      "Docs retrieved for query #11: 1000\n",
      "Docs retrieved for query #12: 1000\n",
      "Docs retrieved for query #13: 1000\n",
      "Docs retrieved for query #14: 1000\n",
      "Docs retrieved for query #15: 1000\n",
      "Docs retrieved for query #16: 1000\n",
      "Docs retrieved for query #18: 1000\n",
      "Docs retrieved for query #20: 1000\n",
      "Docs retrieved for query #22: 1000\n",
      "Docs retrieved for query #23: 1000\n",
      "Docs retrieved for query #24: 1000\n",
      "Docs retrieved for query #25: 1000\n",
      "Docs retrieved for query #27: 1000\n",
      "Docs retrieved for query #28: 1000\n",
      "Docs retrieved for query #29: 1000\n",
      "Docs retrieved for query #31: 1000\n",
      "Docs retrieved for query #32: 1000\n",
      "Docs retrieved for query #34: 1000\n",
      "Docs retrieved for query #36: 1000\n",
      "Docs retrieved for query #37: 1000\n",
      "Docs retrieved for query #38: 1000\n",
      "Docs retrieved for query #39: 1000\n",
      "Docs retrieved for query #40: 1000\n",
      "Docs retrieved for query #42: 1000\n",
      "Docs retrieved for query #44: 1000\n",
      "Docs retrieved for query #45: 1000\n",
      "Docs retrieved for query #46: 1000\n",
      "Docs retrieved for query #48: 1000\n",
      "Docs retrieved for query #50: 1000\n"
     ]
    }
   ],
   "source": [
    "print('Number of retrieved documents: %d' % len(all_retrieved))\n",
    "\n",
    "for qid in all_retrieved:\n",
    "    print ('Docs retrieved for query #{}: {}'.format(qid, str(len(all_retrieved[qid]))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For your own understanding, __inspect the structure and content of the `all_retrieved` and `all_relevant` data structures__ to understand them better. Use the `print()` function to see the content of the data structures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write here the code to inspect the data structures\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 01.F: _mean average precision_\n",
    "__Using the `average_precision()` function you implemented above, write the code to compute the _Mean Average Precision_ for the `baseline.run` results.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean Average Precision (MAP): 0.763\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def mean_average_precision(all_relevant, all_retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    count = 0\n",
    "    total = 0\n",
    "    for retrieved in all_retrieved:\n",
    "        total += precision(all_relevant, retrieved)\n",
    "        count += 1\n",
    "    # END ANSWER\n",
    "    return total / count\n",
    "\n",
    "mapr = mean_average_precision(all_relevant, all_retrieved)\n",
    "print('Mean Average Precision (MAP): %1.3f\\n' % mapr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## TREC benchmark evaluation\n",
    "\n",
    "Below you find a function that take `all_relevant` and `all_retrieved` to compute the mean value of the `measure` over all queries. \n",
    "\n",
    "The function `mean_metric()`'s first function argument, `measure`, is a special argument: it is a function too! The `mean_metric` function sums the total score for the particular measure and divides it by the number of queries. It computes the average measures over all the query results.\n",
    "\n",
    "_This part will be reused later to compare the results of different models._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('mean r_precision', 0.09155954402134368)"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def mean_metric(measure, all_relevant, all_retrieved):\n",
    "    total = 0\n",
    "    count = 0\n",
    "    for qid in all_relevant:\n",
    "        relevant  = all_relevant[qid]\n",
    "        retrieved = all_retrieved.get(qid, [])\n",
    "        value = measure(relevant, retrieved)\n",
    "        total += value\n",
    "        count += 1\n",
    "    return \"mean \" + measure.__name__, total / count\n",
    "\n",
    "# Example of use of the mean_metric function: computing the average r_precision\n",
    "mean_metric(r_precision, all_relevant, all_retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TREC overview of the results of a run\n",
    "The following two functions use your implementation of the metrics to create an overview of the performance metrics on the TREC benchmark data. Give a look at the numbers and make your own interpretations of the results. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for data01/baseline.run\n",
      "mean success_at_1              0.1053\n",
      "mean success_at_5              0.2632\n",
      "mean success_at_10             0.3158\n",
      "mean r_precision               0.09156\n",
      "mean precision_at_1            0.1053\n",
      "mean precision_at_5            0.07895\n",
      "mean precision_at_10           0.04737\n",
      "mean precision_at_50           0.01947\n",
      "mean precision_at_100          0.01395\n",
      "mean precision_at_recall_00    0.2015\n",
      "mean precision_at_recall_01    0.1898\n",
      "mean precision_at_recall_02    0.1683\n",
      "mean precision_at_recall_03    0.1333\n",
      "mean precision_at_recall_04    0.1236\n",
      "mean precision_at_recall_05    0.1227\n",
      "mean precision_at_recall_06    0.08744\n",
      "mean precision_at_recall_07    0.08435\n",
      "mean precision_at_recall_08    0.05999\n",
      "mean precision_at_recall_09    0.05803\n",
      "mean precision_at_recall_10    0.05803\n",
      "mean average_precision         0.05687\n"
     ]
    }
   ],
   "source": [
    "def trec_eval(qrels_file, run_file):\n",
    "\n",
    "    def precision_at_1(rel, ret): return precision_at_k(rel, ret, k=1)\n",
    "    def precision_at_5(rel, ret): return precision_at_k(rel, ret, k=5)\n",
    "    def precision_at_10(rel, ret): return precision_at_k(rel, ret, k=10)\n",
    "    def precision_at_50(rel, ret): return precision_at_k(rel, ret, k=50)\n",
    "    def precision_at_100(rel, ret): return precision_at_k(rel, ret, k=100)\n",
    "    def precision_at_recall_00(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.0)\n",
    "    def precision_at_recall_01(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.1)\n",
    "    def precision_at_recall_02(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.2)\n",
    "    def precision_at_recall_03(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.3)\n",
    "    def precision_at_recall_04(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.4)\n",
    "    def precision_at_recall_05(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.5)\n",
    "    def precision_at_recall_06(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.6)\n",
    "    def precision_at_recall_07(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.7)\n",
    "    def precision_at_recall_08(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.8)\n",
    "    def precision_at_recall_09(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.9)\n",
    "    def precision_at_recall_10(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=1.0)\n",
    "\n",
    "    (all_relevant, all_retrieved) = read_eval_files(qrels_file, run_file)\n",
    "    \n",
    "    unknown_qids = set(all_retrieved.keys()).difference(all_relevant.keys())\n",
    "    if len(unknown_qids) > 0:\n",
    "        raise ValueError(\"Unknown qids in run: {}\".format(sorted(list(unknown_qids))))\n",
    "\n",
    "    metrics = [success_at_1,\n",
    "               success_at_5,\n",
    "               success_at_10,\n",
    "               r_precision,\n",
    "               precision_at_1,\n",
    "               precision_at_5,\n",
    "               precision_at_10,\n",
    "               precision_at_50,\n",
    "               precision_at_100,\n",
    "               precision_at_recall_00,\n",
    "               precision_at_recall_01,\n",
    "               precision_at_recall_02,\n",
    "               precision_at_recall_03,\n",
    "               precision_at_recall_04,\n",
    "               precision_at_recall_05,\n",
    "               precision_at_recall_06,\n",
    "               precision_at_recall_07,\n",
    "               precision_at_recall_08,\n",
    "               precision_at_recall_09,\n",
    "               precision_at_recall_10,\n",
    "               average_precision]\n",
    "\n",
    "    return [mean_metric(metric, all_relevant, all_retrieved) for metric in metrics]\n",
    "\n",
    "\n",
    "def print_trec_eval(qrels_file, run_file):\n",
    "    results = trec_eval(qrels_file, run_file)\n",
    "    print(\"Results for {}\".format(run_file))\n",
    "    for (metric, score) in results:\n",
    "        print(\"{:<30} {:.4}\".format(metric, score))\n",
    "\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'data01/baseline.run')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Exercise 01.G: _Significance testing_\n",
    "\n",
    "Testing the statistical significance of differences of the results of different IR systems is important (see slides of lecture 01 and course book, Section 8.8). One of the basic tests one can perform is the two-tailed [sign test](https://en.wikipedia.org/wiki/Sign_test).\n",
    "\n",
    "Only for this exercise, we use the run files obtained by  [Hiemstra and Aly](https://djoerdhiemstra.com/wp-content/uploads/trec2014mirex-draft.pdf) for the TREC Web track 2014 benchmark (note these files are from a different benchmark from what we have been working with so far). The `utbase.run` file was generated using Language Modeling, while `utexact.run` was generated using an IR system based on mathing the exact query string, and ranking the documents by  the number of exact matches found. The exact run improves the _Precision at 5_ to 0.456 (compared to 0.440 for the baseline run).  \n",
    "\n",
    "__Implement the code to perform the _sign test_ of statistical significance.__\n",
    "> _Hint:_ for each sign, compute the number of queries that increase/descrease performance (called `better, worse` in the code below). How would you use these values to compute the _p_ value of the two-tailed sign test? Is the difference between _utbase_ and _utexact_ significant?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9, 9)"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def sign_test_values(measure, qrels_file, run_file_1, run_file_2):\n",
    "    all_relevant = read_qrels_file(qrels_file)\n",
    "    all_retrieved_1 = read_run_file(run_file_1)\n",
    "    all_retrieved_2 = read_run_file(run_file_2)\n",
    "    better = 0\n",
    "    worse  = 0\n",
    "    # BEGIN ANSWER\n",
    "    for i in all_relevant:\n",
    "        m1 = measure(all_relevant[i], all_retrieved_1[i])\n",
    "        m2 = measure(all_relevant[i], all_retrieved_2[i])\n",
    "        if m1 > m2:\n",
    "            worse += 1\n",
    "        if m1 < m2:\n",
    "            better += 1\n",
    "    # END ANSWER\n",
    "    return(better, worse)\n",
    "    \n",
    "def precision_at_rank_5(rel, ret):\n",
    "    return precision_at_k(rel, ret, k=5)\n",
    "\n",
    "sign_test_values(precision_at_rank_5, 'data01/trec.qrels', 'data01/utbase.run', 'data01/utexact.run')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "***\n",
    "***\n",
    "***\n",
    "\n",
    "# Part 02 - Indexing and querying with ElasticSearch"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation: Getting started with Elasticsearch\n",
    "\n",
    "The following parts of the assignment will be based on ElasticSearch. you are adviced to go through the \"Elasticsearch, [reference guide](https://www.elastic.co/guide/en/elasticsearch/reference/current/getting-started.html)\", and work on the tutorials. You can skip the section on [Installation](https://www.elastic.co/guide/en/elasticsearch/reference/current/_installation.html), as we provide it already installed in the Virtual Machine.\n",
    "\n",
    "> If you want (disclaimer: we do __not__ give help with this!), you can \n",
    "> follow the [Installation](https://www.elastic.co/guide/en/elasticsearch/reference/current/_installation.html) to run Elasticsearch on your laptop without VM. Beware your system will likely be different from the \n",
    "> one of your colleagues and they might not be able to help you if \n",
    "> you have problems that are specific to your system, your operating\n",
    "> system, or your Elasticsearch version.\n",
    "\n",
    "### Starting/Stopping ElasticSearch\n",
    "To start ElasticSearch on the virtual machine, you can type `sudo service elasticsearch start` in a Terminal.\n",
    "To stop the ElasticSearch server, instead, you can type `sudo service elasticsearch stop`. Refer at the [the official guide](https://www.elastic.co/guide/en/elasticsearch/reference/current/deb.html#deb-running-init), for more information.\n",
    "\n",
    "### The REST API\n",
    "\n",
    "Elasticsearch runs its own server that can be accessed by a regular web browser by opening this link: http://localhost:9200. \n",
    "\n",
    "Elasticsearch will respond with something like:\n",
    "\n",
    "    {\n",
    "        \"name\" : \"fir-machine\",\n",
    "        \"cluster_name\" : \"elasticsearch\",\n",
    "        \"cluster_uuid\" : \"w7SBVo1ESVivMApbLIqRvA\",\n",
    "        \"version\" : {\n",
    "            \"number\" : \"7.9.0\",\n",
    "            \"build_flavor\" : \"default\",\n",
    "            \"build_type\" : \"deb\",\n",
    "            \"build_hash\" : \"a479a2a7fce0389512d6a9361301708b92dff667\",\n",
    "            \"build_date\" : \"2020-08-11T21:36:48.204330Z\",\n",
    "            \"build_snapshot\" : false,\n",
    "            \"lucene_version\" : \"8.6.0\",\n",
    "            \"minimum_wire_compatibility_version\" : \"6.8.0\",\n",
    "            \"minimum_index_compatibility_version\" : \"6.0.0-beta1\"\n",
    "        },\n",
    "        \"tagline\" : \"You Know, for Search\"\n",
    "    }\n",
    "\n",
    "\n",
    "If you see this, then your Elasticsearch node is up and running. The RESTful API uses simple text or JSON over HTTP. \n",
    "\n",
    "> REST, API, JSON, HTTP, that's a lot of abbreviations! It is good to\n",
    "> be familiar with the terminology. Let us explain: The Elasticsearch\n",
    "> response is not (only) intended for humans. It is supposed to be used \n",
    "> by applications that run on the client machines, and therefore the\n",
    "> interface is called an Application Programming Interface (API). The \n",
    "> API uses a format called JSON (JavaScript Object Notation), which \n",
    "> can be easily read by machines (and humans). The API sends its JSON\n",
    "> response using the same method as your web browser displays web\n",
    "> pages. This method is called HTTP (Hyper Text Transfer Protocol), \n",
    "> and it is the reason you can inspect the response in a normal web\n",
    "> browser. APIs that use HTTP are called RESTful interfaces. REST \n",
    "> stands for REpresentational State Transfer, arguably one of the\n",
    "> simplest ways to define an API.\n",
    "\n",
    "\n",
    "### Interacting with the ElasticSearch server\n",
    "\n",
    "You can interact with your Elasticsearch service in different ways. In this first part we explore Kibana, a dashboard for inspection of your indices. Later during the practical work we will use the Python Elasticsearch client or the DSL library. You can also start yourself with Python.\n",
    "\n",
    "#### Kibana\n",
    "Kibana provides a web interface to interact with your Elasticsearch service. It's available from http://localhost:5601. You can use Kibana to create interactive dashboards visualizing data in your Elasticsearch indices. It also provides a console to execute Elasticsearch commands. It's available from http://localhost:5601/app/kibana#/dev_tools\n",
    "\n",
    "To start Kibana on the virtual machine, you can type `sudo service kibana start` in a Terminal. \\\n",
    "To stop the Kibana server, instead, you can type `sudo service kibana stop`.\n",
    "\n",
    "Many examples from the Elasticsearch user guide can be directly executed in Kibana by clicking on the `CONSOLE` button.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Indexing and queries (Exercises - Part 02)\n",
    "\n",
    "_You can work on this part after Lecture 01 already_\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collection indexing: useful code\n",
    "\n",
    "We provide some code to read the TREC collection documents and index them into the ElasticSearch engine.\n",
    "As we need to re-index the document collection when we use a different indexing configurations (called Mappings in ElasticSearch), we developed some functions to support a quick re-indexing in the following exercises.\n",
    "\n",
    "Below you find the Python code for bulk-indexing our (FIR)Medline collection. Execute the following cells to index the collection in an Elasticsearch index called `genomics'. Study the code carefully, as you will use the indexing functions later for the completion of the assignment.\n",
    "\n",
    "> The code uses additional helper functions \n",
    "> (`elasticsearch.helpers`) and a library for processing JSON.\n",
    "> The function `read_documents()` reads the bulk collection file: The \n",
    "> function is a [Python generator](https://wiki.python.org/moin/Generators) function. It generates an 'on-demand' list\n",
    "> by using the statement `yield` for every item of the list. It\n",
    "> is used in the helper function `elasticsearch.helpers.bulk()`.\n",
    "> The statement `raise` is Python's approach to throw exceptions: it exits the program with an error.\n",
    "> Note the (keyword) arguments to bulk:\n",
    "> `chunk_size` indicates the number of documents to be processed by\n",
    "> elasticsearch in one batch. \n",
    "> The request_timeout is set to 30 seconds because processing a single batch\n",
    "> of documents can take some time.\n",
    "\n",
    "> __Note:__ _when processing a bulk index, be sure to have few GigaBytes free on the hard drive of the VM. If you get a BulkIndexError with read-only/FORBIDDEN errors, you probably have too little hard drive space available for ElasticSearch to work properly._\n",
    "\n",
    "\n",
    "**_Note:_ indexing the (FIR)TREC genomics collection can take some time, be patient.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "import elasticsearch\n",
    "import elasticsearch.helpers\n",
    "import json\n",
    "\n",
    "def read_documents(file_name):\n",
    "    \"\"\"\n",
    "    Returns a generator of documents to be indexed by elastic, read from file_name\n",
    "    \"\"\"\n",
    "    with open(file_name, 'r') as documents:\n",
    "        for line in documents:\n",
    "            doc_line = json.loads(line)\n",
    "            if ('index' in doc_line):\n",
    "                id = doc_line['index']['_id']\n",
    "            elif ('PMID' in doc_line):\n",
    "                doc_line['_id'] = id\n",
    "                yield doc_line\n",
    "            else:\n",
    "                raise ValueError('Woops, error in index file')\n",
    "\n",
    "def create_index(es, index_name, body={}):\n",
    "    # delete index when it already exists\n",
    "    es.indices.delete(index=index_name, ignore=[400, 404])\n",
    "    # create the index \n",
    "    es.indices.create(index=index_name, body=body)\n",
    "                \n",
    "def index_documents(es, collection_file_name, index_name, body={}):\n",
    "    create_index(es, index_name, body)\n",
    "    # bulk index the documents from file_name\n",
    "    return elasticsearch.helpers.bulk(\n",
    "        es, \n",
    "        read_documents(collection_file_name),\n",
    "        index=index_name,\n",
    "        chunk_size=2000,\n",
    "        request_timeout=30\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ut-student/.local/lib/python3.8/site-packages/elasticsearch/connection/base.py:208: ElasticsearchWarning: Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.17/security-minimal-setup.html to enable security.\n",
      "  warnings.warn(message, category=ElasticsearchWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(263080, [])"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Connect to the ElasticSearch server\n",
    "es = elasticsearch.Elasticsearch(host='localhost')  # in case you use Docker, the host is 'elasticsearch'\n",
    "\n",
    "# Index the collection into the index called 'genomics'\n",
    "body = {} # no indexing options (leave default)\n",
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-base', body)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> You can change the name of the index, in case you want to have different indices of the same collection created with different indexing settings, and compare the performance on the test queries. \n",
    "\n",
    "> E.g. you create two indices 'genomics01' and 'genomics02': genomics01 uses the default options, while genomics02 uses custom tokenizers. You will then have two indices with different characteristics (and probably different performance). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 02.A: index properties and querying"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__1. Query the index called 'genomics-base' and determine how many documents are indexed.__\n",
    "\n",
    "You can use Kibana (suggested for the time being - you can use the command line in Kibana), the Python ElasticSearch library or DSL. Report the code you implemented and the resulting number of documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Documents in 'genomics-base' is 263080\n",
      "{'AB': 'The global fold of maltose binding protein in complex with beta-cyclodextrin has been determined using a CNS-based torsion angle molecular dynamics protocol involving direct refinement against dipolar couplings and carbonyl chemical shift changes that occur upon alignment. The shift changes have been included as structural restraints using a new module, CANI, that has been incorporated into CNS. Force constants and timesteps have been determined that are particularly effective in structure refinement applications involving high molecular weight proteins with small to moderate numbers of NOE restraints. Solution structures of the N', 'AD': 'Protein Engineering Network Center of Excellence and Department of Medical Genetics and Microbiology, University of Toronto, Ontario, Canada.', 'CY': 'Netherlands', 'DA': '20011105', 'DCOM': '20020401', 'DP': '2001 Sep', 'EDAT': '2001/11/06 10:00', 'IP': '1', 'IS': '0925-2738', 'JID': '9110829', 'LA': 'eng', 'LR': '20021101', 'MHDA': '2002/04/02 10:01', 'PG': '31-40', 'PMID': '11693566', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'J Biomol NMR 2001 Sep;21(1):31-40.', 'TA': 'J Biomol NMR', 'TI': 'Direct structure refinement of high molecular weight proteins against residual dipolar couplings and carbonyl chemical shift changes upon alignment: an application to maltose binding protein.', 'UI': '21544926', 'VI': '21', 'MH': ['Carrier Proteins/*chemistry/metabolism', 'Crystallography, X-Ray', 'Cyclodextrins/chemistry/metabolism', 'Hydrogen Bonding', 'Models, Molecular', 'Molecular Weight', 'Nuclear Magnetic Resonance, Biomolecular/*methods', 'Protein Binding', 'Protein Structure, Tertiary', \"Support, Non-U.S. Gov't\", 'Temperature', 'Thermodynamics'], 'RN': ['0 (Carrier Proteins)', '0 (Cyclodextrins)', '0 (maltose-binding protein)', '7585-39-9 (betadex)'], 'FAU': ['Choy, W Y', 'Tollinger, M', 'Mueller, G A', 'Kay, L E']}\n",
      "{'AB': 'The dependence between the anomeric carbon chemical shift and the glycosidic bond (phi, psi) dihedral angles in oligosaccharide and glycopeptide model compounds was studied by Gauge-Including Atomic Orbital (GIAO) ab initio calculations. Complete chemical shift surfaces versus phi and psi for D-Glcp-D-Glcp disaccharides with (1-->1), (1-->2), (1-->3), and (1-->4) linkages in both alpha', 'AD': 'Department of Chemistry & Biochemistry, University of the Sciences in Philadelphia, PA 19104-4495, USA.', 'CY': 'Netherlands', 'DA': '20011105', 'DCOM': '20020401', 'DP': '2001 Sep', 'EDAT': '2001/11/06 10:00', 'IP': '1', 'IS': '0925-2738', 'JID': '9110829', 'LA': 'eng', 'MHDA': '2002/04/02 10:01', 'PG': '49-61', 'PMID': '11693568', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'J Biomol NMR 2001 Sep;21(1):49-61.', 'TA': 'J Biomol NMR', 'TI': 'Derivation of 13C chemical shift surfaces for the anomeric carbons of oligosaccharides and glycopeptides using ab initio methodology.', 'UI': '21544928', 'VI': '21', 'MH': ['Carbohydrate Conformation', 'Carbon/*chemistry', 'Carbon Isotopes/chemistry', 'Computer Simulation', 'Glycopeptides/*chemistry', 'Magnetic Resonance Spectroscopy/*methods', 'Models, Molecular', 'Oligosaccharides/*chemistry', 'Protein Conformation', \"Support, Non-U.S. Gov't\"], 'RN': ['0 (Carbon Isotopes)', '0 (Glycopeptides)', '0 (Oligosaccharides)', '7440-44-0 (Carbon)'], 'FAU': ['Swalina, C W', 'Zauhar, R J', 'DeGrazia, M J', 'Moyna, G']}\n",
      "{'CY': 'Netherlands', 'DA': '20011105', 'DCOM': '20020401', 'DP': '2001 Sep', 'EDAT': '2001/11/06 10:00', 'IP': '1', 'IS': '0925-2738', 'JID': '9110829', 'LA': 'eng', 'MHDA': '2002/04/02 10:01', 'PG': '63-6', 'PMID': '11693569', 'PST': 'ppublish', 'PT': 'Letter', 'SB': 'IM', 'SI': 'PDB/1IIO', 'SO': 'J Biomol NMR 2001 Sep;21(1):63-6.', 'TA': 'J Biomol NMR', 'TI': 'NMR-based structure of the conserved protein MTH865 from the archaeon Methanobacterium thermoautotrophicum.', 'UI': '21544929', 'VI': '21', 'MH': ['Amino Acid Sequence', 'Archaeal Proteins/*chemistry', 'Conserved Sequence', 'Methanobacterium/*chemistry', 'Models, Molecular', 'Molecular Sequence Data', 'Nuclear Magnetic Resonance, Biomolecular', 'Protein Conformation', \"Support, Non-U.S. Gov't\"], 'RN': ['0 (Archaeal Proteins)'], 'FAU': ['Lee, G M', 'Edwards, A M', 'Arrowsmith, C H', 'McIntosh, L P']}\n",
      "{'CY': 'Netherlands', 'DA': '20011105', 'DCOM': '20020401', 'DP': '2001 Sep', 'EDAT': '2001/11/06 10:00', 'IP': '1', 'IS': '0925-2738', 'JID': '9110829', 'LA': 'eng', 'MHDA': '2002/04/02 10:01', 'PG': '67-8', 'PMID': '11693570', 'PST': 'ppublish', 'PT': 'Letter', 'SB': 'IM', 'SO': 'J Biomol NMR 2001 Sep;21(1):67-8.', 'TA': 'J Biomol NMR', 'TI': 'Assignment of selectively 13C-labeled cellopentaose synthesized using an engineered glycosynthase.', 'UI': '21544930', 'VI': '21', 'MH': ['Agrobacterium/enzymology', 'Carbon Isotopes/chemistry', 'Chromatography, High Pressure Liquid', 'Magnetic Resonance Spectroscopy/methods', 'Oligosaccharides/*biosynthesis/*chemistry', '*Protein Engineering', \"Support, Non-U.S. Gov't\", 'beta-Glucosidase/*metabolism'], 'RN': ['0 (Carbon Isotopes)', '0 (Oligosaccharides)', '34620-76-3 (maltopentaose)', 'EC 3.2.1.21 (beta-Glucosidase)'], 'FAU': ['Brun, E', 'Brumer, H 3rd', 'MacKenzie, L F', 'Withers, S G', 'McIntosh, L P']}\n",
      "{'CY': 'Netherlands', 'DA': '20011105', 'DCOM': '20020401', 'DP': '2001 Sep', 'EDAT': '2001/11/06 10:00', 'ID': 'GM-52421/GM/NIGMS', 'IP': '1', 'IS': '0925-2738', 'JID': '9110829', 'LA': 'eng', 'MHDA': '2002/04/02 10:01', 'PG': '69-70', 'PMID': '11693571', 'PST': 'ppublish', 'PT': 'Letter', 'SB': 'IM', 'SO': 'J Biomol NMR 2001 Sep;21(1):69-70.', 'TA': 'J Biomol NMR', 'TI': '1H, 15N and 13C assignments of the N-terminal domain of Yersinia outer protein H in its apo form and in complex with a phosphotyrosine peptide.', 'UI': '21544931', 'VI': '21', 'MH': ['Apoproteins/*chemistry', 'Bacterial Outer Membrane Proteins/*chemistry/*metabolism', 'Macromolecular Systems', 'Nuclear Magnetic Resonance, Biomolecular/methods', 'Phosphopeptides/chemistry/*metabolism', 'Phosphotyrosine/chemistry/*metabolism', 'Protein Binding', 'Protein Structure, Tertiary', 'Protein-Tyrosine-Phosphatase/*chemistry/*metabolism', \"Support, Non-U.S. Gov't\", \"Support, U.S. Gov't, Non-P.H.S.\", \"Support, U.S. Gov't, P.H.S.\", 'Titrimetry', 'Yersinia/*chemistry'], 'RN': ['0 (Apoproteins)', '0 (Bacterial Outer Membrane Proteins)', '0 (Macromolecular Systems)', '0 (Phosphopeptides)', '123781-29-3 (yopH protein, Yersinia)', '21820-51-9 (Phosphotyrosine)', 'EC 3.1.3.48 (Protein-Tyrosine-Phosphatase)'], 'FAU': ['Khandelwal, P', 'Keliikuli, K', 'Smit, C L', 'Saper, M A', 'Zuiderweg, E R']}\n",
      "{'CY': 'Netherlands', 'DA': '20011105', 'DCOM': '20020401', 'DP': '2001 Sep', 'EDAT': '2001/11/06 10:00', 'IP': '1', 'IS': '0925-2738', 'JID': '9110829', 'LA': 'eng', 'MHDA': '2002/04/02 10:01', 'PG': '77-8', 'PMID': '11693575', 'PST': 'ppublish', 'PT': 'Letter', 'SB': 'IM', 'SO': 'J Biomol NMR 2001 Sep;21(1):77-8.', 'TA': 'J Biomol NMR', 'TI': 'Sequential assignment and secondary structure of saratin, an inhibitor of von Willebrand factor-dependent platelet adhesion to collagen.', 'UI': '21544935', 'VI': '21', 'MH': ['Collagen/*metabolism', 'Nuclear Magnetic Resonance, Biomolecular/methods', 'Pichia/*chemistry', '*Platelet Adhesiveness', 'Platelet Aggregation Inhibitors/*chemistry', 'Protein Structure, Secondary', 'Salivary Proteins/*chemistry', \"Support, Non-U.S. Gov't\", 'von Willebrand Factor/*antagonists & inhibitors'], 'RN': ['0 (Platelet Aggregation Inhibitors)', '0 (Salivary Proteins)', '0 (saratin)', '0 (von Willebrand Factor)', '9007-34-5 (Collagen)'], 'FAU': ['Maurer, T', 'Bomke, J', 'Frech, M', 'Rysiok, T', 'Kalbitzer, H B']}\n",
      "{'AB': 'There is a great need for an integrated international effort in research and training using rapid, easy to use, biomarker and microscale ecotoxicity techniques. These techniques must be directed, coordinated and formulated into protocols that contribute to the prevention and reduction of marine pollution world-wide and the improvement of ocean and human health. This need should be considered as urgent by marine environmental scientists, managers and policy makers throughout the world. Our paper discusses such techniques and suggests a four-point framework for advancing work towards their wider use, particularly in developing coastal nations.', 'AD': 'Environmental Conservation Branch, Environment Canada, Dartmouth, NS. peter.wells@ec.gc.ca', 'CY': 'England', 'DA': '20011105', 'DCOM': '20020702', 'DP': '2001 Oct', 'EDAT': '2001/11/06 10:00', 'IP': '10', 'IS': '0025-326X', 'JID': '0260231', 'LA': 'eng', 'MHDA': '2002/07/03 10:01', 'PG': '799-804', 'PMID': '11693633', 'PST': 'ppublish', 'PT': 'Review, Tutorial', 'RF': '44', 'SB': 'IM', 'SO': 'Mar Pollut Bull 2001 Oct;42(10):799-804.', 'TA': 'Mar Pollut Bull', 'TI': 'Rapid toxicity assessment and biomonitoring of marine contaminants--exploiting the potential of rapid biomarker assays and microscale toxicity tests.', 'UI': '21544993', 'VI': '42', 'MH': ['Animal', 'Biological Markers/*analysis', 'Developing Countries', 'Environmental Monitoring/*methods/standards', 'Human', 'International Cooperation', 'Risk Assessment', \"Support, Non-U.S. Gov't\", 'Toxicity Tests/*methods', 'Water Pollutants/analysis/*toxicity', 'Water Pollution/*prevention & control'], 'RN': ['0 (Biological Markers)', '0 (Water Pollutants)'], 'FAU': ['Wells, P G', 'Depledge, M H', 'Butler, J N', 'Manock, J J', 'Knap, A H']}\n",
      "{'AB': 'Discharges of tritium (3H) into the Severn estuary/Bristol Channel (UK) arise from the authorized release of wastes from nuclear power plants at Hinkley Point and Berkley/Oldbury and from the Nycomed-Amersham radiochemical plant, via the sewer system, at Cardiff. The wastes from the nuclear power plants probably consist almost entirely of 3H2O, whereas those from the radiochemical plant also include uncharacterized 3H labelled organic compounds. The total 3H concentrations in demersal fish and other benthic organisms in the vicinity of the Cardiff Eastern sewer outfall are significantly elevated compared to those observed around other UK nuclear establishments. Concentrations in filtered seawater were approximately 10 Bq kg(-1) whilst levels in surface sediment, seaweed (Fucus vesiculosis) and mussels (Mytilus edulis)/flounder (Platichthys flesus) were in the order of 6 x 10(2), 2 x 10(3), and 10(5) Bq kg(-1) (dry weight), respectively. Almost all the 3H found in sediment and biota were organically bound tritium (OBT). The high concentration in these materials, relative to that in seawater, is due to the presence of bioavailable organic 3H labelled compounds in the radiochemical waste. It is suggested that bioaccumulation of 3H by benthic organisms and demersal fish occurs primarily via a pathway of physico-chemical sorption/bacterial transformation of dissolved 3H labelled organic compounds into particulate organic matter, and subsequent transfer up a web of sediment dwelling microbes and meiofauna. Variations in 3H accumulation between individual organisms have been interpreted in terms of their different feeding behaviour. Relatively low concentrations were observed in the herbivorous winkle (Littorina littorea) and the pelagic Sprat (Spratus spratus) compared with other benthic organisms and demersal fish. The elevated 3H concentrations in seafood, due to bioaccumulation of OBT, have low radiological significance even for the local critical group of seafood consumers.', 'AD': 'The Centre for Environment, Fisheries & Aquaculture Science, CEFAS Lowestoft Laboratory, Suffolk, UK. d.mccubbin@cefas.co.uk', 'CY': 'England', 'DA': '20011105', 'DCOM': '20020702', 'DP': '2001 Oct', 'EDAT': '2001/11/06 10:00', 'IP': '10', 'IS': '0025-326X', 'JID': '0260231', 'LA': 'eng', 'MHDA': '2002/07/03 10:01', 'PG': '852-63', 'PMID': '11693639', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Mar Pollut Bull 2001 Oct;42(10):852-63.', 'TA': 'Mar Pollut Bull', 'TI': 'Incorporation of organic tritium (3H) by marine organisms and sediment in the severn estuary/Bristol channel (UK).', 'UI': '21544999', 'VI': '42', 'MH': ['Animal', 'Bacteria/metabolism', 'Biological Availability', 'Flounder/*metabolism', 'Food Chain', 'Geologic Sediments/*chemistry', 'Great Britain', 'Mussels/*metabolism', 'Polychaeta/metabolism', 'Power Plants', 'Radioactive Waste', 'Seawater/chemistry', 'Seaweed/*metabolism', \"Support, Non-U.S. Gov't\", 'Tissue Distribution', 'Tritium/metabolism/*pharmacokinetics', 'Water Pollutants, Radioactive/metabolism/*pharmacokinetics'], 'RN': ['0 (Radioactive Waste)', '0 (Water Pollutants, Radioactive)', '10028-17-8 (Tritium)'], 'FAU': ['McCubbin, D', 'Leonard, K S', 'Bailey, T A', 'Williams, J', 'Tossell, P']}\n",
      "{'AB': \"Land-use changes and associated river discharges in coastal tropical regions present a global threat to coral reef environments. This study investigated the temporal variation in biological oxygen demand (BOD5) and suspended particulate matter (SPM) at the mouths of seven rivers on Okinawa Island (Japan) over 20 years. We report strong positive relationships between human population densities within river catchment areas and both average BOD5 concentration (r2 = 0.968; p < 0.001) and SPM (r2 = 0.659; p < 0.003) at the mouths of the rivers. At the reef adjacent to one river (Hija River, 50.2 km2 catchment area) we applied moving window analysis to assess an optimal sampling strategy for elucidating transitional boundaries in coral composition from the river mouth to a point where the effect of river discharge was minimal. The optimal window width for Okinawan rivers was five 1 m2 quadrats spaced over 5 m intervals. This sampling strategy clearly showed dissimilarity spikes in coral community composition up to 400 m from the Hija River mouth, beyond which no significant differences in coral composition were detected using analysis of similarities (ANOSIM). We developed a simple diffusion model linking the rivers' maximum discharge rate, and the average concentration of BOD5 and SPM with the spatial impact on the coral communities. The diffusion model can aid in predicting negative shifts in coral communities expected to result from detrimental land-use changes and is an important tool for monitoring coral reefs.\", 'AD': 'Department of Marine Sciences, University of the Ryukyus, Nishihara, Okinawa, Japan. kristiw@hawaii.edu', 'CY': 'England', 'DA': '20011105', 'DCOM': '20020702', 'DP': '2001 Oct', 'EDAT': '2001/11/06 10:00', 'IP': '10', 'IS': '0025-326X', 'JID': '0260231', 'LA': 'eng', 'MHDA': '2002/07/03 10:01', 'PG': '864-72', 'PMID': '11693640', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Mar Pollut Bull 2001 Oct;42(10):864-72.', 'TA': 'Mar Pollut Bull', 'TI': 'Spatial and temporal variance of river discharge on Okinawa (Japan): inferring the temporal impact on adjacent coral reefs.', 'UI': '21545000', 'VI': '42', 'MH': ['Animal', 'Cnidaria/*growth & development/physiology', 'Conservation of Natural Resources', 'Diffusion', 'Ecosystem', 'Environmental Monitoring', 'Fresh Water/*chemistry', 'Human', 'Japan', 'Oxygen/metabolism', 'Population Density', \"Support, Non-U.S. Gov't\", 'Time Factors', 'Water Pollutants/analysis/metabolism'], 'RN': ['0 (Water Pollutants)', '7782-44-7 (Oxygen)'], 'FAU': ['West, K', 'van Woesik, R']}\n",
      "{'AB': 'The spatial distribution and fate of petrogenic and pyrogenic hydrocarbons in coastal sediments following the Aegean Sea oil spill (Galicia, NW Spain) was investigated through a detailed study of chemical markers. Alkanes and acyclic isoprenoid hydrocarbons were degraded within six months of the accident, but triterpane and sterane distributions were still detectable and were useful in monitoring, respectively, the oil source and weathering. Aromatic steranes were also useful source indicators, and oil degradation was clear from the decrease of certain alkyl dibenzothiophene and phenanthrene isomers. The pyrogenic polycyclic aromatic hydrocarbons produced in the tanker wreck were found more persistent than the petrogenic ones. However, the presence of the oil in the sediments was recognized even one year after the accident and was found particularly preserved in the subsurface layers.', 'AD': 'Department of Environmental Chemistry, CID-CSIC, Barcelona, Spain.', 'CY': 'England', 'DA': '20011105', 'DCOM': '20020702', 'DP': '2001 Oct', 'EDAT': '2001/11/06 10:00', 'IP': '10', 'IS': '0025-326X', 'JID': '0260231', 'LA': 'eng', 'MHDA': '2002/07/03 10:01', 'PG': '895-904', 'PMID': '11693644', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Mar Pollut Bull 2001 Oct;42(10):895-904.', 'TA': 'Mar Pollut Bull', 'TI': 'The aegean sea oil spill in the Galicia Coast (NW Spain). I. Distribution and fate of the crude oil and combustion products in subtidal sediments.', 'UI': '21545004', 'VI': '42', 'MH': ['Environmental Monitoring', 'Geologic Sediments/*chemistry', 'Oceans and Seas', 'Petroleum/*analysis', 'Polycyclic Hydrocarbons, Aromatic/*analysis', 'Spain', \"Support, Non-U.S. Gov't\", 'Triterpenes/analysis', 'Water Pollutants, Chemical/*analysis'], 'RN': ['0 (Polycyclic Hydrocarbons, Aromatic)', '0 (Triterpenes)', '0 (Water Pollutants, Chemical)'], 'FAU': ['Pastor, D', 'Sanchez, J', 'Porte, C', 'Albaiges, J']}\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# write the code here\n",
    "# BEGIN ANSWER\n",
    "index = \"genomics-base\"\n",
    "document_count = es.count(index=index)['count']\n",
    "print(f\"Number of Documents in '{index}' is {document_count}\")\n",
    "\n",
    "\n",
    "query = {\n",
    "    \"query\": {\n",
    "        \"match_all\": {}\n",
    "    }\n",
    "}\n",
    "\n",
    "result = es.search(index=\"genomics-base\", body=query, track_total_hits=True)\n",
    "\n",
    "for hit in result['hits']['hits']:\n",
    "    print(hit['_source'])\n",
    "\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__2. How many documents containing the term `molecule` are there in your index? (searching all fields of the documents).__\n",
    "\n",
    "You can use Kibana (suggested for the time being - you can use the command line in Kibana), the Python ElasticSearch library or DSL. Report the code you implemented and the resulting number of documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of documents is 3404\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# write the code that generates the answer here (you may also use Kibana)\n",
    "# BEGIN ANSWER\n",
    "body = {\n",
    "    \"query\": {\n",
    "        \"query_string\": {\n",
    "            \"query\": \"molecule\"\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "result = es.search(index=index, body=body, size=0, track_total_hits=True)  # size=0 returns no hits, just the count\n",
    "hits_total_value = result['hits']['total']['value']\n",
    "\n",
    "print(f\"Number of documents is {hits_total_value}\")\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__3. How many documents containing the term `molecular` are there in your index? (searching all fields of the documents).__\n",
    "\n",
    "You can use Kibana (suggested for the time being - you can use the command line in Kibana), the Python ElasticSearch library or DSL. Report the code you implemented and the resulting number of documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of documents is 31556\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# write the code that generates the answer here (you may also use Kibana)\n",
    "# BEGIN ANSWER\n",
    "body = {\n",
    "    \"query\": {\n",
    "        \"query_string\": {\n",
    "            \"query\": \"molecular\"\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "result = es.search(index=index, body=body, size=0, track_total_hits=True)  # size=0 returns no hits, just the count\n",
    "hits_total_value = result['hits']['total']['value']\n",
    "\n",
    "print(f\"Number of documents is {hits_total_value}\")\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__4. How many documents containing the terms `cell` AND `blood` are there in your index? (searching all fields of the documents).__\n",
    "\n",
    "You can use Kibana (suggested for the time being - you can use the command line in Kibana), the Python ElasticSearch library or DSL. Report the code you implemented and the resulting number of documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of documents is 6865\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# write the code that generates the answer here (you may also use Kibana)\n",
    "# BEGIN ANSWER\n",
    "\n",
    "body = {\n",
    "    \"query\": {\n",
    "        \"query_string\": {\n",
    "            \"query\": \"cell AND blood\"\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "result = es.search(index=index, body=body, size=0, track_total_hits=True)  # size=0 returns no hits, just the count\n",
    "hits_total_value = result['hits']['total']['value']\n",
    "\n",
    "print(f\"Number of documents is {hits_total_value}\")\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of results: 6865\n"
     ]
    }
   ],
   "source": [
    "import elasticsearch\n",
    "es = elasticsearch.Elasticsearch(host='localhost')  # in case you use Docker, the host is 'elasticsearch'\n",
    "\n",
    "# this is another solution (if query_string is used, be sure that AND is in the query, otherwise it will not search properly)\n",
    "term = 'blood AND cell'\n",
    "body = {\"track_total_hits\": True, \"query\": {\"query_string\": \n",
    "                                            {\"query\": term, \n",
    "                                             \"default_operator\":\"AND\", \n",
    "                                             \"auto_generate_synonyms_phrase_query\": True }}}\n",
    "result = es.search(index='genomics-base', body=body, track_total_hits=True)\n",
    "print(\"Number of results: {}\".format(result['hits']['total']['value']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 02.B: the Python ElasticSearch library\n",
    "\n",
    "#### Preparation\n",
    "The command line is fine for doing basic operations on your Elasticsearch indices, but as soon as things get more complex, you better use custom client programs.\n",
    "We will use the [Elasticsearch client library for Python](https://elasticsearch-py.readthedocs.io). This library will execute the HTTP requests that you have used before (with CURL or Kibana). The library is pre-installed on the VM.\n",
    "\n",
    "#### Exercise\n",
    "\n",
    "__Write the code that searches the index for _\"molecule\"_ using the [search()](https://elasticsearch-py.readthedocs.io/en/master/api.html#elasticsearch.Elasticsearch.search) function.__ Your code will take at minimum the following steps:\n",
    "\n",
    "1. import the python library `elasticsearch`.\n",
    "2. open a connection with the Elasticsearch host `'elasticsearch'` with `Elasticsearch()`.\n",
    "3. execute a search with `search()` using the index `genomics-base`, and a correct query body.\n",
    "4. print the JSON output of Elasticsearch \n",
    "\n",
    "How many hits are there in your index? Is the result the same as in Exercise 02.A?\n",
    "\n",
    "> Elasticsearch runs on localhost on your laptop, at port 9200 (so as http://localhost:9200)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of documents (with term molecule) is 3404\n",
      "Json output: {'took': 19, 'timed_out': False, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}, 'hits': {'total': {'value': 3404, 'relation': 'eq'}, 'max_score': 10.767412, 'hits': [{'_index': 'genomics-base', '_type': '_doc', '_id': '21807', '_score': 10.767412, '_ignored': ['AB.keyword'], '_source': {'AB': \"OBJECTIVE: With Helios gene gun, the report genes EGFP and Lac-Z were transfected to several cultured mammalian tumor cell lines in vitro and mice skin in vivo, respectively. A stable long time exogene's expression were got. Then, the pWRG3142, which carried mGM-CSF gene was delivered to the abdominal skin of C57BL/6 mice using gene gun. Pathological sections showed the local transproteins' expression companying with a profound inflammation reaction characterized by neutrophilic infiltration. ELISA assay of transfected mouse's serum sample indicated a high improvement of transgenic proteins level, which demonstrated that transgenic GM-CSF secreted from treated skin into the bloodstream effectively. In B16 melanoma tumor model, mice immunized with mGM-CSF expression plasmids could be partially protected from 1 x 10(5) B16 cells challenge and exhibited a drastically reduced tumor growth rate. Finally, we conclucled that exogenes could be transfected into cultured cell lines and mice skin effectively by gene gun technology, and gene gun mediated in vivo delivery of GM-CSF cDNA should be further developed for potential clinical testing as an approach for human cancer gene therapy.\", 'AD': 'National Lab of Molecule Oncology, Cancer Institute, PUMC & CAMS, Beijing, 100021 China.', 'CY': 'China', 'DA': '20020118', 'DCOM': '20020711', 'DP': '2000 Jul', 'EDAT': '2002/01/19 10:00', 'IP': '7', 'IS': '0376-2491', 'JID': '7511141', 'LA': 'chi', 'MHDA': '2002/07/12 10:01', 'PG': '522-5', 'PMID': '11798811', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Zhonghua Yi Xue Za Zhi 2000 Jul;80(7):522-5.', 'TA': 'Zhonghua Yi Xue Za Zhi', 'TI': '[The use of gene gun in cancer gene therapy]', 'UI': '21658031', 'VI': '80', 'MH': ['Animal', 'Biolistics/*methods', 'Cell Division/genetics', 'English Abstract', 'Gene Therapy/*methods', 'Granulocyte-Macrophage Colony-Stimulating Factor/genetics', 'Human', 'Melanoma, Experimental/genetics/pathology/*therapy', 'Mice', 'Mice, Inbred C57BL', 'Plasmids/genetics', 'Tumor Cells, Cultured'], 'RN': ['0 (Plasmids)', '83869-56-1 (Granulocyte-Macrophage Colony-Stimulating Factor)'], 'FAU': ['Zha, Y', 'Lin, C', 'Liang, X']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '132671', '_score': 9.435863, '_ignored': ['AB.keyword'], '_source': {'AB': 'OBJECTIVE: To determine the effect of stage III and IV endometriosis on serum concentration of soluble intracellular adhesion molecule-1 (ICAM-1), vascular cellular adhesion molecule-1 (VCAM-1), and E-selectin. METHODS: Sera from women with laparoscopically proven stage III and IV endometriosis were analyzed for the presence of soluble ICAM-1, VCAM-1, and E-selectin using enzyme-linked immunosorbent assays. Concentrations of these molecules were compared with those of laparoscopically proven disease-free controls. RESULTS: When compared with controls, women with stage III and IV endometriosis had higher serum concentrations of soluble VCAM-1, lower serum concentration of soluble ICAM-1, and no difference in serum concentration of soluble E-selectin. CONCLUSION: Significant aberrations in levels of circulating adhesion molecules were found in women with stage III and IV endometriosis. These findings might shed light on the pathogenesis of endometriosis and be useful in the development of biochemical markers for disease stage.', 'AD': 'Department of Obstetrics and Gynecology, University of Missouri, Columbia, Missouri, USA. breton.barrier@lackland.af.mil', 'CY': 'United States', 'DA': '20020419', 'DCOM': '20021025', 'DP': '2002 Mar-Apr', 'EDAT': '2002/04/20 10:00', 'IP': '2', 'IS': '1071-5576', 'JID': '9433806', 'LA': 'eng', 'MHDA': '2002/10/31 04:00', 'PG': '98-101', 'PMID': '11963839', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'J Soc Gynecol Investig 2002 Mar-Apr;9(2):98-101.', 'TA': 'J Soc Gynecol Investig', 'TI': 'Expression of soluble adhesion molecules in sera of women with stage III and IV endometriosis.', 'UI': '21961150', 'VI': '9', 'MH': ['Biological Markers/blood', 'E-Selectin/*blood', 'Endometriosis/*blood/pathology', 'Female', 'Human', 'Intercellular Adhesion Molecule-1/*blood', 'Vascular Cell Adhesion Molecule-1/*blood'], 'RN': ['0 (Biological Markers)', '0 (E-Selectin)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Barrier, Breton F', 'Sharpe-Timms, Kathy L']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '148107', '_score': 9.326713, '_source': {'AID': '106195 [pii]', 'CY': 'France', 'DA': '20020430', 'DCOM': '20020627', 'DP': '2002 Apr', 'EDAT': '2002/05/01 10:00', 'IP': '2 Pt 2', 'IS': '0368-2315', 'JID': '0322206', 'LA': 'fre', 'LR': '20021101', 'MHDA': '2002/06/28 10:01', 'PG': '1S5-1S6', 'PMID': '11981482', 'PST': 'ppublish', 'PT': 'Editorial', 'SB': 'IM', 'SO': 'J Gynecol Obstet Biol Reprod (Paris) 2002 Apr;31(2 Pt 2):1S5-1S6.', 'TA': 'J Gynecol Obstet Biol Reprod (Paris)', 'TI': '[LH Day: floodlights on a forgotten compound]', 'TT': 'LH Day: coup de projecteur sur une molecule oubliee.', 'UI': '21977764', 'VI': '31', 'MH': ['Female', 'Human', 'Luteinizing Hormone/*physiology', 'Menstrual Cycle', 'Oocytes/physiology', 'Ovarian Follicle/physiology', 'Ovulation Induction', 'Polycystic Ovary Syndrome', 'Receptors, LH/physiology'], 'RN': ['0 (Receptors, LH)', '9002-67-9 (Luteinizing Hormone)'], 'FAU': ['Fanchin, R']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '249043', '_score': 9.312457, '_ignored': ['AB.keyword'], '_source': {'AB': 'The structures of two potential anti-human immunodeficiency virus type 1 (HIV-1) non-nucleoside reverse transcriptase inhibitors (NNRTI), namely 1-benzyl-5-oxo-2-phenylpyrrolidine-2-carboxamide, C(18)H(18)N(2)O(2), (III), and 2-(4-isopropoxyphenyl)-5-oxo-1-(4-tolyl)pyrrolidine-2-carbonitrile, C(21)H(22)N(2)O(2), (IV), have been investigated by X-ray diffraction, confirming the butterfly-like conformation of both compounds. The pyrrolidine ring is in an envelope conformation in (III) and a half-chair conformation in (IV). Two intermolecular N[bond]H...O hydrogen bonds are present in the crystal structure of (III), with N...O distances of 2.995 (2) and 2.927 (2) A.', 'AD': 'Molecule Structure Research Centre, National Academy of Sciences RA, Azatutyan Ave. 26, 375014 Yerevan, Republic of Armenia.', 'AID': 'S0108270102009241 [pii]', 'CY': 'Denmark', 'DA': '20020702', 'DCOM': '20020917', 'DP': '2002 Jul', 'EDAT': '2002/07/03 10:00', 'IP': 'Pt 7', 'IS': '0108-2701', 'JID': '8305826', 'LA': 'eng', 'MHDA': '2002/07/03 10:00', 'PG': 'o399-401', 'PHST': '2002/Jun/20 [epublish]', 'PMID': '12094058', 'PST': 'ppublish', 'PT': 'Journal Article', 'SO': 'Acta Crystallogr C 2002 Jul;58(Pt 7):o399-401.', 'TA': 'Acta Crystallogr C', 'TI': '1-Substituted derivatives of 2-aryl-5-oxopyrrolidine-2-carboxylic acid.', 'UI': '22088002', 'VI': '58', 'FAU': ['Karapetyan, Harutyun', 'Tamazyan, Rafael', 'Martirosyan, Ashot', 'Hovhannesyan, Vahagn', 'Gasparyan, Sahak']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '212437', '_score': 9.013664, '_source': {'AD': 'Department of Physiology and Biophysics, University of Miami School of Medicine, Miami, Florida 33136, USA.', 'CY': 'United States', 'DA': '20020610', 'DCOM': '20030131', 'DP': '2002', 'EDAT': '2002/06/11 10:00', 'ID': '1 R29 GM55611-01/GM/NIGMS', 'IS': '0091-679X', 'JID': '0373334', 'LA': 'eng', 'MHDA': '2003/02/01 04:00', 'PG': '301-9', 'PMID': '12053735', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Methods Cell Biol 2002;68:301-9.', 'TA': 'Methods Cell Biol', 'TI': 'Single-molecule force measurements.', 'UI': '22049610', 'VI': '68', 'MH': ['3T3 Cells', 'Animal', 'Avidin/chemistry/metabolism', 'Biotin/chemistry/metabolism', 'Ligands', 'Mice', 'Microscopy, Atomic Force/instrumentation/*methods', 'Protein Binding', 'Receptors, Cell Surface/metabolism', \"Support, Non-U.S. Gov't\", \"Support, U.S. Gov't, P.H.S.\"], 'RN': ['0 (Ligands)', '0 (Receptors, Cell Surface)', '1405-69-2 (Avidin)', '58-85-5 (Biotin)'], 'FAU': ['Chen, Aileen', 'Moy, Vincent T']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '467640', '_score': 9.013664, '_ignored': ['AB.keyword'], '_source': {'AB': 'The Alliance for Cellular Signaling (AfCS)-Nature Molecule Pages will be a comprehensive database of key facts about more than 3,000 proteins involved in cell signalling. Each entry will be created by invited experts and be peer-reviewed. Alongside the large-scale experiments being conducted by the AfCS scientists, the wealth of information contained in this database offers the potential of accelerating the pace of discovery in signal transduction research.', 'AD': 'San Diego Supercomputer Center, University of California at San Diego, 9500 Gilman Drive, La Jolla, California 92093, USA.', 'AID': 'nature01307 [pii]', 'CY': 'England', 'DA': '20021212', 'DCOM': '20030113', 'DP': '2002 Dec 12', 'EDAT': '2002/12/13 04:00', 'IP': '6916', 'IS': '0028-0836', 'JID': '0410462', 'LA': 'eng', 'MHDA': '2003/01/14 04:00', 'PG': '716-7', 'PMID': '12478304', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Nature 2002 Dec 12;420(6916):716-7.', 'TA': 'Nature', 'TI': 'The Molecule Pages database.', 'UI': '22366428', 'VI': '420', 'MH': ['Automation', '*Databases, Factual', 'Heterotrimeric GTP-Binding Proteins/metabolism', 'Information Management', 'Peer Review, Research', 'Protein Binding', '*Research', '*Signal Transduction'], 'RN': ['EC 3.6.1.'], 'FAU': ['Li, Joshua', 'Ning, Yuhong', 'Hedley, Warren', 'Saunders, Brian', 'Chen, Yongsheng', 'Tindill, Nicole', 'Hannay, Timo', 'Subramaniam, Shankar']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '280977', '_score': 8.912884, '_ignored': ['AB.keyword'], '_source': {'AB': 'To determine whether soluble adhesion molecules are affected in muscular dystrophy, we measured serum levels of creatine kinase (CK), soluble vascular cell adhesion molecule-1 (sVCAM-1), soluble intercellular adhesion molecule-1 (sICAM-1), soluble (s) E-selectin, and fibrin and fibrinogen degradation products (FDP) in 25 patients with Duchenne muscular dystrophy (DMD), 7 with Becker muscular dystrophy, 7 with Fukuyama type congenital muscular dystrophy, 6 with myotonic dystrophy (MyD), and 5 with spinal muscular atrophy (SMA) type 2, and also serum sVCAM-1, sICAM-1, and sE-selectin in 9 healthy controls. The levels of sVCAM-1 in the patients with DMD were 367.0-852.0 ng/ml (552.8 +/', 'AD': 'Department of Neurology, Toneyama National Hospital, Toyonaka, Osaka.', 'CY': 'Japan', 'DA': '20020723', 'DCOM': '20020730', 'DP': '2002 Jul', 'EDAT': '2002/07/24 10:00', 'IP': '4', 'IS': '0029-0831', 'JID': '0215224', 'LA': 'jpn', 'LR': '20021101', 'MHDA': '2002/07/31 10:01', 'PG': '326-9', 'PMID': '12134684', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'No To Hattatsu 2002 Jul;34(4):326-9.', 'TA': 'No To Hattatsu', 'TI': '[Soluble adhesion molecules in muscular dystrophy]', 'UI': '22130743', 'VI': '34', 'MH': ['Adolescent', 'Adult', 'Child', 'Child, Preschool', 'E-Selectin/blood', 'English Abstract', 'Female', 'Human', 'Intercellular Adhesion Molecule-1/*blood', 'Male', 'Muscular Dystrophies/*blood', 'Solubility', 'Vascular Cell Adhesion Molecule-1/*blood'], 'RN': ['0 (E-Selectin)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Saito, Toshio', 'Yamamoto, Yuko', 'Shinno, Susumu']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '502100', '_score': 8.912884, '_source': {'CY': 'China', 'DA': '20030127', 'DCOM': '20030212', 'DP': '2000 Jul', 'EDAT': '2003/01/28 04:00', 'IP': '3', 'IS': '0559-7765', 'JID': '20730140R', 'LA': 'chi', 'MHDA': '2003/02/14 04:00', 'PG': '257-60', 'PMID': '12545718', 'PST': 'ppublish', 'PT': 'Review, Tutorial', 'RF': '8', 'SB': 'IM', 'SO': 'Sheng Li Ke Xue Jin Zhan 2000 Jul;31(3):257-60.', 'TA': 'Sheng Li Ke Xue Jin Zhan', 'TI': '[Adhesion molecules and inflammation following cerebral ischemia]', 'UI': '22434060', 'VI': '31', 'MH': ['Animal', 'Brain Ischemia/complications/*metabolism', 'Cell Adhesion Molecules/biosynthesis', 'Human', 'Inflammation/etiology/*metabolism', 'Integrins/biosynthesis', 'Intercellular Adhesion Molecule-1/*biosynthesis', 'Selectins/*biosynthesis', 'Vascular Cell Adhesion Molecule-1/biosynthesis'], 'RN': ['0 (Cell Adhesion Molecules)', '0 (Integrins)', '0 (Selectins)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Xu, H L', 'Fen, Y P']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '249039', '_score': 8.834926, '_ignored': ['AB.keyword'], '_source': {'AB': 'The structures of two potential anti-human immunodeficiency virus type 1 (HIV-1) non-nucleoside reverse transcriptase inhibitors (NNRTI), namely 1-(2-chlorobenzoyl)-2-phenylpyrrolidine-2-carboxamide, C(18)H(17)ClN(2)O(2), and 1-(2-furoyl)-2-phenylpyrrolidine-2-carboxamide, C(16)H(16)N(2)O(3), have been investigated by X-ray diffraction and the butterfly-like conformation established in each case. The pyrrolidine ring has the same half-chair conformation in both structures.', 'AD': 'Molecule Structure Research Centre, National Academy of Sciences RA, Azatutyan Ave. 26, 375014 Yerevan, Republic of Armenia. rafael@msrc.am', 'AID': 'S0108270102008636 [pii]', 'CY': 'Denmark', 'DA': '20020702', 'DCOM': '20020917', 'DP': '2002 Jul', 'EDAT': '2002/07/03 10:00', 'IP': 'Pt 7', 'IS': '0108-2701', 'JID': '8305826', 'LA': 'eng', 'MHDA': '2002/07/03 10:00', 'PG': 'o386-8', 'PHST': '2002/Jun/12 [epublish]', 'PMID': '12094054', 'PST': 'ppublish', 'PT': 'Journal Article', 'SO': 'Acta Crystallogr C 2002 Jul;58(Pt 7):o386-8.', 'TA': 'Acta Crystallogr C', 'TI': '1-Substituted derivatives of 2-phenylpyrrolidine-2-carboxamide.', 'UI': '22087998', 'VI': '58', 'FAU': ['Tamazyan, Rafael', 'Karapetyan, Harutyun', 'Martirosyan, Ashot', 'Hovhannesyan, Vahagn', 'Gasparyan, Sahak']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '63151', '_score': 8.751207, '_source': {'AID': 'S0021915001007572 [pii]', 'CY': 'Ireland', 'DA': '20020307', 'DCOM': '20020716', 'DP': '2002 Mar', 'EDAT': '2002/03/08 10:00', 'IP': '1', 'IS': '0021-9150', 'JID': '0242543', 'LA': 'eng', 'LR': '20020808', 'MHDA': '2002/07/18 10:01', 'PG': '243-4', 'PMID': '11882338', 'PST': 'ppublish', 'PT': 'Letter', 'SB': 'IM', 'SO': 'Atherosclerosis 2002 Mar;161(1):243-4.', 'TA': 'Atherosclerosis', 'TI': 'Serum levels of adhesion molecules correlate with insulin resistance.', 'UI': '21876766', 'VI': '161', 'MH': ['Cell Adhesion Molecules/*blood', 'Diabetes Mellitus, Non-Insulin-Dependent/blood', 'E-Selectin/blood', 'Human', '*Insulin Resistance', 'Intercellular Adhesion Molecule-1/blood', 'Male', 'Middle Age', 'Vascular Cell Adhesion Molecule-1/blood'], 'RN': ['0 (Cell Adhesion Molecules)', '0 (E-Selectin)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Matsumoto, Kazunari', 'Sera, Yasunori', 'Miyake, Seibei', 'Ueki, Yukitaka']}}]}}\n",
      "Number of documents (with term molecular) is 31556\n",
      "Json output: {'took': 19, 'timed_out': False, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}, 'hits': {'total': {'value': 3404, 'relation': 'eq'}, 'max_score': 10.767412, 'hits': [{'_index': 'genomics-base', '_type': '_doc', '_id': '21807', '_score': 10.767412, '_ignored': ['AB.keyword'], '_source': {'AB': \"OBJECTIVE: With Helios gene gun, the report genes EGFP and Lac-Z were transfected to several cultured mammalian tumor cell lines in vitro and mice skin in vivo, respectively. A stable long time exogene's expression were got. Then, the pWRG3142, which carried mGM-CSF gene was delivered to the abdominal skin of C57BL/6 mice using gene gun. Pathological sections showed the local transproteins' expression companying with a profound inflammation reaction characterized by neutrophilic infiltration. ELISA assay of transfected mouse's serum sample indicated a high improvement of transgenic proteins level, which demonstrated that transgenic GM-CSF secreted from treated skin into the bloodstream effectively. In B16 melanoma tumor model, mice immunized with mGM-CSF expression plasmids could be partially protected from 1 x 10(5) B16 cells challenge and exhibited a drastically reduced tumor growth rate. Finally, we conclucled that exogenes could be transfected into cultured cell lines and mice skin effectively by gene gun technology, and gene gun mediated in vivo delivery of GM-CSF cDNA should be further developed for potential clinical testing as an approach for human cancer gene therapy.\", 'AD': 'National Lab of Molecule Oncology, Cancer Institute, PUMC & CAMS, Beijing, 100021 China.', 'CY': 'China', 'DA': '20020118', 'DCOM': '20020711', 'DP': '2000 Jul', 'EDAT': '2002/01/19 10:00', 'IP': '7', 'IS': '0376-2491', 'JID': '7511141', 'LA': 'chi', 'MHDA': '2002/07/12 10:01', 'PG': '522-5', 'PMID': '11798811', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Zhonghua Yi Xue Za Zhi 2000 Jul;80(7):522-5.', 'TA': 'Zhonghua Yi Xue Za Zhi', 'TI': '[The use of gene gun in cancer gene therapy]', 'UI': '21658031', 'VI': '80', 'MH': ['Animal', 'Biolistics/*methods', 'Cell Division/genetics', 'English Abstract', 'Gene Therapy/*methods', 'Granulocyte-Macrophage Colony-Stimulating Factor/genetics', 'Human', 'Melanoma, Experimental/genetics/pathology/*therapy', 'Mice', 'Mice, Inbred C57BL', 'Plasmids/genetics', 'Tumor Cells, Cultured'], 'RN': ['0 (Plasmids)', '83869-56-1 (Granulocyte-Macrophage Colony-Stimulating Factor)'], 'FAU': ['Zha, Y', 'Lin, C', 'Liang, X']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '132671', '_score': 9.435863, '_ignored': ['AB.keyword'], '_source': {'AB': 'OBJECTIVE: To determine the effect of stage III and IV endometriosis on serum concentration of soluble intracellular adhesion molecule-1 (ICAM-1), vascular cellular adhesion molecule-1 (VCAM-1), and E-selectin. METHODS: Sera from women with laparoscopically proven stage III and IV endometriosis were analyzed for the presence of soluble ICAM-1, VCAM-1, and E-selectin using enzyme-linked immunosorbent assays. Concentrations of these molecules were compared with those of laparoscopically proven disease-free controls. RESULTS: When compared with controls, women with stage III and IV endometriosis had higher serum concentrations of soluble VCAM-1, lower serum concentration of soluble ICAM-1, and no difference in serum concentration of soluble E-selectin. CONCLUSION: Significant aberrations in levels of circulating adhesion molecules were found in women with stage III and IV endometriosis. These findings might shed light on the pathogenesis of endometriosis and be useful in the development of biochemical markers for disease stage.', 'AD': 'Department of Obstetrics and Gynecology, University of Missouri, Columbia, Missouri, USA. breton.barrier@lackland.af.mil', 'CY': 'United States', 'DA': '20020419', 'DCOM': '20021025', 'DP': '2002 Mar-Apr', 'EDAT': '2002/04/20 10:00', 'IP': '2', 'IS': '1071-5576', 'JID': '9433806', 'LA': 'eng', 'MHDA': '2002/10/31 04:00', 'PG': '98-101', 'PMID': '11963839', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'J Soc Gynecol Investig 2002 Mar-Apr;9(2):98-101.', 'TA': 'J Soc Gynecol Investig', 'TI': 'Expression of soluble adhesion molecules in sera of women with stage III and IV endometriosis.', 'UI': '21961150', 'VI': '9', 'MH': ['Biological Markers/blood', 'E-Selectin/*blood', 'Endometriosis/*blood/pathology', 'Female', 'Human', 'Intercellular Adhesion Molecule-1/*blood', 'Vascular Cell Adhesion Molecule-1/*blood'], 'RN': ['0 (Biological Markers)', '0 (E-Selectin)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Barrier, Breton F', 'Sharpe-Timms, Kathy L']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '148107', '_score': 9.326713, '_source': {'AID': '106195 [pii]', 'CY': 'France', 'DA': '20020430', 'DCOM': '20020627', 'DP': '2002 Apr', 'EDAT': '2002/05/01 10:00', 'IP': '2 Pt 2', 'IS': '0368-2315', 'JID': '0322206', 'LA': 'fre', 'LR': '20021101', 'MHDA': '2002/06/28 10:01', 'PG': '1S5-1S6', 'PMID': '11981482', 'PST': 'ppublish', 'PT': 'Editorial', 'SB': 'IM', 'SO': 'J Gynecol Obstet Biol Reprod (Paris) 2002 Apr;31(2 Pt 2):1S5-1S6.', 'TA': 'J Gynecol Obstet Biol Reprod (Paris)', 'TI': '[LH Day: floodlights on a forgotten compound]', 'TT': 'LH Day: coup de projecteur sur une molecule oubliee.', 'UI': '21977764', 'VI': '31', 'MH': ['Female', 'Human', 'Luteinizing Hormone/*physiology', 'Menstrual Cycle', 'Oocytes/physiology', 'Ovarian Follicle/physiology', 'Ovulation Induction', 'Polycystic Ovary Syndrome', 'Receptors, LH/physiology'], 'RN': ['0 (Receptors, LH)', '9002-67-9 (Luteinizing Hormone)'], 'FAU': ['Fanchin, R']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '249043', '_score': 9.312457, '_ignored': ['AB.keyword'], '_source': {'AB': 'The structures of two potential anti-human immunodeficiency virus type 1 (HIV-1) non-nucleoside reverse transcriptase inhibitors (NNRTI), namely 1-benzyl-5-oxo-2-phenylpyrrolidine-2-carboxamide, C(18)H(18)N(2)O(2), (III), and 2-(4-isopropoxyphenyl)-5-oxo-1-(4-tolyl)pyrrolidine-2-carbonitrile, C(21)H(22)N(2)O(2), (IV), have been investigated by X-ray diffraction, confirming the butterfly-like conformation of both compounds. The pyrrolidine ring is in an envelope conformation in (III) and a half-chair conformation in (IV). Two intermolecular N[bond]H...O hydrogen bonds are present in the crystal structure of (III), with N...O distances of 2.995 (2) and 2.927 (2) A.', 'AD': 'Molecule Structure Research Centre, National Academy of Sciences RA, Azatutyan Ave. 26, 375014 Yerevan, Republic of Armenia.', 'AID': 'S0108270102009241 [pii]', 'CY': 'Denmark', 'DA': '20020702', 'DCOM': '20020917', 'DP': '2002 Jul', 'EDAT': '2002/07/03 10:00', 'IP': 'Pt 7', 'IS': '0108-2701', 'JID': '8305826', 'LA': 'eng', 'MHDA': '2002/07/03 10:00', 'PG': 'o399-401', 'PHST': '2002/Jun/20 [epublish]', 'PMID': '12094058', 'PST': 'ppublish', 'PT': 'Journal Article', 'SO': 'Acta Crystallogr C 2002 Jul;58(Pt 7):o399-401.', 'TA': 'Acta Crystallogr C', 'TI': '1-Substituted derivatives of 2-aryl-5-oxopyrrolidine-2-carboxylic acid.', 'UI': '22088002', 'VI': '58', 'FAU': ['Karapetyan, Harutyun', 'Tamazyan, Rafael', 'Martirosyan, Ashot', 'Hovhannesyan, Vahagn', 'Gasparyan, Sahak']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '212437', '_score': 9.013664, '_source': {'AD': 'Department of Physiology and Biophysics, University of Miami School of Medicine, Miami, Florida 33136, USA.', 'CY': 'United States', 'DA': '20020610', 'DCOM': '20030131', 'DP': '2002', 'EDAT': '2002/06/11 10:00', 'ID': '1 R29 GM55611-01/GM/NIGMS', 'IS': '0091-679X', 'JID': '0373334', 'LA': 'eng', 'MHDA': '2003/02/01 04:00', 'PG': '301-9', 'PMID': '12053735', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Methods Cell Biol 2002;68:301-9.', 'TA': 'Methods Cell Biol', 'TI': 'Single-molecule force measurements.', 'UI': '22049610', 'VI': '68', 'MH': ['3T3 Cells', 'Animal', 'Avidin/chemistry/metabolism', 'Biotin/chemistry/metabolism', 'Ligands', 'Mice', 'Microscopy, Atomic Force/instrumentation/*methods', 'Protein Binding', 'Receptors, Cell Surface/metabolism', \"Support, Non-U.S. Gov't\", \"Support, U.S. Gov't, P.H.S.\"], 'RN': ['0 (Ligands)', '0 (Receptors, Cell Surface)', '1405-69-2 (Avidin)', '58-85-5 (Biotin)'], 'FAU': ['Chen, Aileen', 'Moy, Vincent T']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '467640', '_score': 9.013664, '_ignored': ['AB.keyword'], '_source': {'AB': 'The Alliance for Cellular Signaling (AfCS)-Nature Molecule Pages will be a comprehensive database of key facts about more than 3,000 proteins involved in cell signalling. Each entry will be created by invited experts and be peer-reviewed. Alongside the large-scale experiments being conducted by the AfCS scientists, the wealth of information contained in this database offers the potential of accelerating the pace of discovery in signal transduction research.', 'AD': 'San Diego Supercomputer Center, University of California at San Diego, 9500 Gilman Drive, La Jolla, California 92093, USA.', 'AID': 'nature01307 [pii]', 'CY': 'England', 'DA': '20021212', 'DCOM': '20030113', 'DP': '2002 Dec 12', 'EDAT': '2002/12/13 04:00', 'IP': '6916', 'IS': '0028-0836', 'JID': '0410462', 'LA': 'eng', 'MHDA': '2003/01/14 04:00', 'PG': '716-7', 'PMID': '12478304', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Nature 2002 Dec 12;420(6916):716-7.', 'TA': 'Nature', 'TI': 'The Molecule Pages database.', 'UI': '22366428', 'VI': '420', 'MH': ['Automation', '*Databases, Factual', 'Heterotrimeric GTP-Binding Proteins/metabolism', 'Information Management', 'Peer Review, Research', 'Protein Binding', '*Research', '*Signal Transduction'], 'RN': ['EC 3.6.1.'], 'FAU': ['Li, Joshua', 'Ning, Yuhong', 'Hedley, Warren', 'Saunders, Brian', 'Chen, Yongsheng', 'Tindill, Nicole', 'Hannay, Timo', 'Subramaniam, Shankar']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '280977', '_score': 8.912884, '_ignored': ['AB.keyword'], '_source': {'AB': 'To determine whether soluble adhesion molecules are affected in muscular dystrophy, we measured serum levels of creatine kinase (CK), soluble vascular cell adhesion molecule-1 (sVCAM-1), soluble intercellular adhesion molecule-1 (sICAM-1), soluble (s) E-selectin, and fibrin and fibrinogen degradation products (FDP) in 25 patients with Duchenne muscular dystrophy (DMD), 7 with Becker muscular dystrophy, 7 with Fukuyama type congenital muscular dystrophy, 6 with myotonic dystrophy (MyD), and 5 with spinal muscular atrophy (SMA) type 2, and also serum sVCAM-1, sICAM-1, and sE-selectin in 9 healthy controls. The levels of sVCAM-1 in the patients with DMD were 367.0-852.0 ng/ml (552.8 +/', 'AD': 'Department of Neurology, Toneyama National Hospital, Toyonaka, Osaka.', 'CY': 'Japan', 'DA': '20020723', 'DCOM': '20020730', 'DP': '2002 Jul', 'EDAT': '2002/07/24 10:00', 'IP': '4', 'IS': '0029-0831', 'JID': '0215224', 'LA': 'jpn', 'LR': '20021101', 'MHDA': '2002/07/31 10:01', 'PG': '326-9', 'PMID': '12134684', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'No To Hattatsu 2002 Jul;34(4):326-9.', 'TA': 'No To Hattatsu', 'TI': '[Soluble adhesion molecules in muscular dystrophy]', 'UI': '22130743', 'VI': '34', 'MH': ['Adolescent', 'Adult', 'Child', 'Child, Preschool', 'E-Selectin/blood', 'English Abstract', 'Female', 'Human', 'Intercellular Adhesion Molecule-1/*blood', 'Male', 'Muscular Dystrophies/*blood', 'Solubility', 'Vascular Cell Adhesion Molecule-1/*blood'], 'RN': ['0 (E-Selectin)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Saito, Toshio', 'Yamamoto, Yuko', 'Shinno, Susumu']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '502100', '_score': 8.912884, '_source': {'CY': 'China', 'DA': '20030127', 'DCOM': '20030212', 'DP': '2000 Jul', 'EDAT': '2003/01/28 04:00', 'IP': '3', 'IS': '0559-7765', 'JID': '20730140R', 'LA': 'chi', 'MHDA': '2003/02/14 04:00', 'PG': '257-60', 'PMID': '12545718', 'PST': 'ppublish', 'PT': 'Review, Tutorial', 'RF': '8', 'SB': 'IM', 'SO': 'Sheng Li Ke Xue Jin Zhan 2000 Jul;31(3):257-60.', 'TA': 'Sheng Li Ke Xue Jin Zhan', 'TI': '[Adhesion molecules and inflammation following cerebral ischemia]', 'UI': '22434060', 'VI': '31', 'MH': ['Animal', 'Brain Ischemia/complications/*metabolism', 'Cell Adhesion Molecules/biosynthesis', 'Human', 'Inflammation/etiology/*metabolism', 'Integrins/biosynthesis', 'Intercellular Adhesion Molecule-1/*biosynthesis', 'Selectins/*biosynthesis', 'Vascular Cell Adhesion Molecule-1/biosynthesis'], 'RN': ['0 (Cell Adhesion Molecules)', '0 (Integrins)', '0 (Selectins)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Xu, H L', 'Fen, Y P']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '249039', '_score': 8.834926, '_ignored': ['AB.keyword'], '_source': {'AB': 'The structures of two potential anti-human immunodeficiency virus type 1 (HIV-1) non-nucleoside reverse transcriptase inhibitors (NNRTI), namely 1-(2-chlorobenzoyl)-2-phenylpyrrolidine-2-carboxamide, C(18)H(17)ClN(2)O(2), and 1-(2-furoyl)-2-phenylpyrrolidine-2-carboxamide, C(16)H(16)N(2)O(3), have been investigated by X-ray diffraction and the butterfly-like conformation established in each case. The pyrrolidine ring has the same half-chair conformation in both structures.', 'AD': 'Molecule Structure Research Centre, National Academy of Sciences RA, Azatutyan Ave. 26, 375014 Yerevan, Republic of Armenia. rafael@msrc.am', 'AID': 'S0108270102008636 [pii]', 'CY': 'Denmark', 'DA': '20020702', 'DCOM': '20020917', 'DP': '2002 Jul', 'EDAT': '2002/07/03 10:00', 'IP': 'Pt 7', 'IS': '0108-2701', 'JID': '8305826', 'LA': 'eng', 'MHDA': '2002/07/03 10:00', 'PG': 'o386-8', 'PHST': '2002/Jun/12 [epublish]', 'PMID': '12094054', 'PST': 'ppublish', 'PT': 'Journal Article', 'SO': 'Acta Crystallogr C 2002 Jul;58(Pt 7):o386-8.', 'TA': 'Acta Crystallogr C', 'TI': '1-Substituted derivatives of 2-phenylpyrrolidine-2-carboxamide.', 'UI': '22087998', 'VI': '58', 'FAU': ['Tamazyan, Rafael', 'Karapetyan, Harutyun', 'Martirosyan, Ashot', 'Hovhannesyan, Vahagn', 'Gasparyan, Sahak']}}, {'_index': 'genomics-base', '_type': '_doc', '_id': '63151', '_score': 8.751207, '_source': {'AID': 'S0021915001007572 [pii]', 'CY': 'Ireland', 'DA': '20020307', 'DCOM': '20020716', 'DP': '2002 Mar', 'EDAT': '2002/03/08 10:00', 'IP': '1', 'IS': '0021-9150', 'JID': '0242543', 'LA': 'eng', 'LR': '20020808', 'MHDA': '2002/07/18 10:01', 'PG': '243-4', 'PMID': '11882338', 'PST': 'ppublish', 'PT': 'Letter', 'SB': 'IM', 'SO': 'Atherosclerosis 2002 Mar;161(1):243-4.', 'TA': 'Atherosclerosis', 'TI': 'Serum levels of adhesion molecules correlate with insulin resistance.', 'UI': '21876766', 'VI': '161', 'MH': ['Cell Adhesion Molecules/*blood', 'Diabetes Mellitus, Non-Insulin-Dependent/blood', 'E-Selectin/blood', 'Human', '*Insulin Resistance', 'Intercellular Adhesion Molecule-1/blood', 'Male', 'Middle Age', 'Vascular Cell Adhesion Molecule-1/blood'], 'RN': ['0 (Cell Adhesion Molecules)', '0 (E-Selectin)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Matsumoto, Kazunari', 'Sera, Yasunori', 'Miyake, Seibei', 'Ueki, Yukitaka']}}]}}\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "import elasticsearch\n",
    "\n",
    "# your code below\n",
    "# BEGIN ANSWER\n",
    "\n",
    "es = elasticsearch.Elasticsearch(host='localhost')\n",
    "\n",
    "body = {\n",
    "    \"query\": {\n",
    "        \"query_string\": {\n",
    "            \"query\": \"molecule\"\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "molecular_body = {\n",
    "    \"query\": {\n",
    "        \"query_string\": {\n",
    "            \"query\": \"molecular\"\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "result = es.search(index=index, body=body)\n",
    "hits_total_value = result['hits']['total']['value']\n",
    "\n",
    "print(f\"Number of documents (with term molecule) is {hits_total_value}\")\n",
    "print(f\"Json output: {result}\")\n",
    "\n",
    "molecular_result = es.search(index=index, body=molecular_body, track_total_hits=True)\n",
    "hits_total_value = molecular_result['hits']['total']['value']\n",
    "\n",
    "print(f\"Number of documents (with term molecular) is {hits_total_value}\")\n",
    "print(f\"Json output: {result}\")\n",
    "# END ANSWER\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Python client library returns Python objects, that use [dictionaries](https://docs.python.org/3.6/tutorial/datastructures.html#dictionaries) and [lists](https://docs.python.org/3.6/tutorial/introduction.html#lists).\n",
    "Use a [for loop](https://docs.python.org/3.6/tutorial/controlflow.html#for-statements) to inspect each hit, and print the retrieved document's titles one by one. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of results: 3404\n",
      "[The use of gene gun in cancer gene therapy]\n",
      "Expression of soluble adhesion molecules in sera of women with stage III and IV endometriosis.\n",
      "[LH Day: floodlights on a forgotten compound]\n",
      "1-Substituted derivatives of 2-aryl-5-oxopyrrolidine-2-carboxylic acid.\n",
      "Single-molecule force measurements.\n",
      "The Molecule Pages database.\n",
      "[Soluble adhesion molecules in muscular dystrophy]\n",
      "[Adhesion molecules and inflammation following cerebral ischemia]\n",
      "1-Substituted derivatives of 2-phenylpyrrolidine-2-carboxamide.\n",
      "Serum levels of adhesion molecules correlate with insulin resistance.\n"
     ]
    }
   ],
   "source": [
    "#example\n",
    "print(\"Number of results: {}\".format(result['hits']['total']['value']))\n",
    "# your code below\n",
    "\n",
    "for hit in result['hits']['hits']:\n",
    "    print(hit['_source'][\"TI\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 02.C: _Search using the Elasticsearch DSL_\n",
    "\n",
    "You will notice that the native query format of Elasticsearch can be quite verbose.\n",
    "Elasticsearh provides the Python library `elasticsearch_dsl` to write more concise Elasticsearch queries. \n",
    "This is only to simplify the syntax: the library still issues Elasticsearch queries.\n",
    "\n",
    "For example, a simple `multi_match` query looks as follows:\n",
    "```python\n",
    "query = {\n",
    "   \"query\": {\n",
    "       \"multi_match\": {}\n",
    "   }\n",
    "}\n",
    "```\n",
    "\n",
    "The same query can be created with the DSL as follows:\n",
    "```python\n",
    "query = Q(\"multi_match\")\n",
    "```\n",
    "\n",
    "Especially for more complicated boolean queries, to use the native query format can become complicated.\n",
    "Read more about the DSL [here](https://elasticsearch-dsl.readthedocs.io/en/latest/search_dsl.html)\n",
    "\n",
    "__1. Search for the query `molecule` and check whether you get the same number of results as for exercise 02.A(2).__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of documents is 3404\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# your code here\n",
    "# BEGIN ANSWER\n",
    "from elasticsearch_dsl import Search, Q\n",
    "\n",
    "search = Search(using=es, index=index)\n",
    "query = Q(\"query_string\", query=\"molecule\")\n",
    "search = search.query(query)\n",
    "result = search.execute()\n",
    "\n",
    "print(\"Number of documents is\", result.hits.total.value)\n",
    "\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__2. Search for the documents that contain the words `cell` AND `blood`, using the DSL library. Check whether you get the same number of results as for exercise 02.A(4).__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of documents is 6865\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# your code here\n",
    "# BEGIN ANSWER\n",
    "search = Search(using=es, index=index)\n",
    "query = Q(\"query_string\", query=\"cell AND blood\")\n",
    "search = search.query(query)\n",
    "result = search.execute()\n",
    "print(\"Number of documents is\", result.hits.total.value)\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "##  Exercise 02.D: Making your own TREC run\n",
    "\n",
    "We will adopt a scientific approach to building search engines. That is, we are not only going to build a search engine and see that it works, but we are also going to _measure_ how well it works, by measuring the search engine's quality. We will adopt the method from the [Text Retrieval Conference](http://trec.nist.gov) (TREC). TREC provides researchers with test collections, that consists of 3 parts:\n",
    "\n",
    "1. the document collection (in our case a part of the MEDLINE database)\n",
    "2. the topics (which are natural language descriptions of what the user is searching for: you can think of the as the _queries_)\n",
    "3. the relevance judgments (for each topic, what documents are relevant)\n",
    "\n",
    "\n",
    "\n",
    "__Exercise: Complete the code of the Python function `make_trec_run()` that reads the topics [FIR-s05-training-queries-simple.txt](data01/FIR-s05-training-queries-simple.txt), and for each topic does a search using Elasticsearch.__ The program should output a file in the [TREC submission format](https://trec-core.github.io/2017/#submission-guidelines). We already provided the first  lines for this exercise, which include:\n",
    "\n",
    "1. Open the file `'run_file_name'`' for writing and call it `run_file`.\n",
    "2. Open the file `'topics_file_name'` for reading, call it `test_queries`.\n",
    "3. For each line in `test_queries`:\n",
    "4. Remove the newline using `strip()`, then split the string on the tab character (`'\\t'`). The first part of the line is now `qid` (the query identifier) and the last part is `query` (a textual description of the query).\n",
    "5. complete the Python program such that the correct TREC run file is written to `'run_file_name'`.\n",
    "\n",
    "> **Note**: Make sure you output the `PMID` (pubmed identifier) of the document `hit['_source']['PMID']`. Do **not** use the elasticsearch identifier `_id` because they do not match the document identifiers in the relevance judgements. They were randomly generated by Elasticsearch during indexing.\n",
    "\n",
    "\n",
    "__Make sure to search in the fiels `TI` and `AB`, which correspond to the title and abstract, respectivelt, of the scientific papers of the MEDLINE collection.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def make_trec_run(es, topics_file_name, run_file_name, index_name=\"genomics\", run_name=\"test\"):\n",
    "    with open(run_file_name, 'w') as run_file:\n",
    "        with open(topics_file_name, 'r') as test_queries:\n",
    "            for line in test_queries:\n",
    "                (qid, query) = line.strip().split('\\t')\n",
    "                # BEGIN ANSWER\n",
    "                # search_query = {\n",
    "                #     \"query\": {\n",
    "                #         \"bool\": {\n",
    "                #             \"should\": [\n",
    "                #                 {\"match\": {\"TI\": query}},\n",
    "                #                 {\"match\": {\"AB\": query}}\n",
    "                #             ]\n",
    "                #         }\n",
    "                #     }\n",
    "                # }\n",
    "                # response = es.search(index=index_name, body=search_query)\n",
    "\n",
    "                # It seems that the result can be different when using DSL vs elastic API.\n",
    "\n",
    "                s = Search(using=es, index = index_name)\n",
    "                s = s.query(Q(\"simple_query_string\", query = query, fields=['TI', 'AB']))\n",
    "                response = s.execute()\n",
    "\n",
    "                # Write results to run file\n",
    "                rank = 1\n",
    "                for hit in response['hits']['hits']:\n",
    "                    pmid = hit['_source']['PMID']  # Use 'UNKNOWN' if PMID is not present\n",
    "                    score = hit['_score']\n",
    "                    line = f\"{qid}\\tQ0\\t{pmid}\\t{str(rank)}\\t{str(score)}\\t{run_name}\\n\"\n",
    "                    run_file.write(line)\n",
    "                    rank += 1\n",
    "                # END ANSWER\n",
    "                \n",
    "# connect to ES server             \n",
    "es = elasticsearch.Elasticsearch('localhost')\n",
    "# Write the results of the queries contained in the topic file `'data/training-queries-simple.txt'` \n",
    "# to the run file `'baseline.run'`, and name this test as `test01`\n",
    "make_trec_run(es, 'data01/FIR-s05-training-queries-simple.txt', 'baseline.run', \"genomics-base\", run_name='test01')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "1\tQ0\t11907576\t2\t28.477657\ttest01\n",
      "1\tQ0\t12167152\t3\t27.794035\ttest01\n",
      "1\tQ0\t11971762\t4\t26.188068\ttest01\n",
      "1\tQ0\t12645611\t5\t24.40221\ttest01\n",
      "1\tQ0\t12393857\t6\t24.205702\ttest01\n",
      "1\tQ0\t12018119\t7\t23.876041\ttest01\n",
      "1\tQ0\t12148460\t8\t23.304388\ttest01\n",
      "1\tQ0\t12399461\t9\t23.044468\ttest01\n",
      "1\tQ0\t12203811\t10\t22.933348\ttest01\n",
      "3\tQ0\t12435397\t1\t39.12735\ttest01\n",
      "3\tQ0\t11761710\t2\t38.898323\ttest01\n",
      "3\tQ0\t11824903\t3\t36.454147\ttest01\n",
      "3\tQ0\t12036890\t4\t35.52736\ttest01\n",
      "3\tQ0\t12370291\t5\t34.95769\ttest01\n",
      "3\tQ0\t11796228\t6\t33.702114\ttest01\n",
      "3\tQ0\t12058964\t7\t31.147223\ttest01\n",
      "3\tQ0\t11991951\t8\t30.934141\ttest01\n",
      "3\tQ0\t12039986\t9\t30.622875\ttest01\n",
      "3\tQ0\t12063983\t10\t30.516993\ttest01\n",
      "4\tQ0\t12455885\t1\t57.972248\ttest01\n",
      "4\tQ0\t12565793\t2\t55.840343\ttest01\n",
      "4\tQ0\t12438323\t3\t54.975372\ttest01\n",
      "4\tQ0\t12384427\t4\t54.577736\ttest01\n",
      "4\tQ0\t12190869\t5\t54.34951\ttest01\n",
      "4\tQ0\t12072406\t6\t53.700294\ttest01\n",
      "4\tQ0\t12475802\t7\t53.19451\ttest01\n",
      "4\tQ0\t12225820\t8\t53.149975\ttest01\n",
      "4\tQ0\t11904459\t9\t53.016304\ttest01\n",
      "4\tQ0\t12061391\t10\t52.522118\ttest01\n",
      "5\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "5\tQ0\t11907576\t2\t28.477657\ttest01\n",
      "5\tQ0\t12167152\t3\t27.794035\ttest01\n",
      "5\tQ0\t11971762\t4\t26.188068\ttest01\n",
      "5\tQ0\t12645611\t5\t24.40221\ttest01\n",
      "5\tQ0\t12393857\t6\t24.205702\ttest01\n",
      "5\tQ0\t12018119\t7\t23.876041\ttest01\n",
      "5\tQ0\t12148460\t8\t23.304388\ttest01\n",
      "5\tQ0\t12399461\t9\t23.044468\ttest01\n",
      "5\tQ0\t12203811\t10\t22.933348\ttest01\n",
      "6\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "6\tQ0\t11907576\t2\t28.477657\ttest01\n",
      "6\tQ0\t12167152\t3\t27.794035\ttest01\n",
      "6\tQ0\t11971762\t4\t26.188068\ttest01\n",
      "6\tQ0\t12645611\t5\t24.40221\ttest01\n",
      "6\tQ0\t12393857\t6\t24.205702\ttest01\n",
      "6\tQ0\t12018119\t7\t23.876041\ttest01\n",
      "6\tQ0\t12148460\t8\t23.304388\ttest01\n",
      "6\tQ0\t12399461\t9\t23.044468\ttest01\n",
      "6\tQ0\t12203811\t10\t22.933348\ttest01\n",
      "7\tQ0\t12459483\t1\t49.43423\ttest01\n",
      "7\tQ0\t12147240\t2\t49.26383\ttest01\n",
      "7\tQ0\t12559972\t3\t48.35612\ttest01\n",
      "7\tQ0\t12538627\t4\t47.924484\ttest01\n",
      "7\tQ0\t11926591\t5\t47.0282\ttest01\n",
      "7\tQ0\t12388236\t6\t46.888424\ttest01\n",
      "7\tQ0\t11812652\t7\t45.7469\ttest01\n",
      "7\tQ0\t12055272\t8\t44.81004\ttest01\n",
      "7\tQ0\t12445196\t9\t44.80458\ttest01\n",
      "7\tQ0\t12009519\t10\t44.559628\ttest01\n",
      "8\tQ0\t12356766\t1\t39.800034\ttest01\n",
      "8\tQ0\t12067715\t2\t37.250828\ttest01\n",
      "8\tQ0\t11716781\t3\t37.16111\ttest01\n",
      "8\tQ0\t11700864\t4\t36.664738\ttest01\n",
      "8\tQ0\t11824903\t5\t36.454147\ttest01\n",
      "8\tQ0\t11859079\t6\t36.22301\ttest01\n",
      "8\tQ0\t12032381\t7\t36.18384\ttest01\n",
      "8\tQ0\t12011080\t8\t31.440596\ttest01\n",
      "8\tQ0\t11882240\t9\t30.865894\ttest01\n",
      "8\tQ0\t12038609\t10\t30.218628\ttest01\n",
      "9\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "9\tQ0\t12009766\t2\t35.69858\ttest01\n",
      "9\tQ0\t12072966\t3\t34.629223\ttest01\n",
      "9\tQ0\t12071967\t4\t33.174267\ttest01\n",
      "9\tQ0\t12218530\t5\t32.040787\ttest01\n",
      "9\tQ0\t11904459\t6\t31.2541\ttest01\n",
      "9\tQ0\t11932979\t7\t31.218292\ttest01\n",
      "9\tQ0\t12417648\t8\t31.156477\ttest01\n",
      "9\tQ0\t12460607\t9\t31.085634\ttest01\n",
      "9\tQ0\t11940769\t10\t30.858912\ttest01\n",
      "10\tQ0\t12063259\t1\t49.770027\ttest01\n",
      "10\tQ0\t12091346\t2\t46.776196\ttest01\n",
      "10\tQ0\t12004254\t3\t41.50456\ttest01\n",
      "10\tQ0\t11824903\t4\t40.070126\ttest01\n",
      "10\tQ0\t12221080\t5\t35.705334\ttest01\n",
      "10\tQ0\t11799130\t6\t34.24811\ttest01\n",
      "10\tQ0\t11776319\t7\t31.517563\ttest01\n",
      "10\tQ0\t11951068\t8\t30.718279\ttest01\n",
      "10\tQ0\t12192065\t9\t30.716959\ttest01\n",
      "10\tQ0\t11856727\t10\t29.872877\ttest01\n",
      "11\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "11\tQ0\t11907576\t2\t28.477657\ttest01\n",
      "11\tQ0\t12167152\t3\t27.794035\ttest01\n",
      "11\tQ0\t11971762\t4\t26.188068\ttest01\n",
      "11\tQ0\t12645611\t5\t24.40221\ttest01\n",
      "11\tQ0\t12393857\t6\t24.205702\ttest01\n",
      "11\tQ0\t12018119\t7\t23.876041\ttest01\n",
      "11\tQ0\t12148460\t8\t23.304388\ttest01\n",
      "11\tQ0\t12399461\t9\t23.044468\ttest01\n",
      "11\tQ0\t12203811\t10\t22.933348\ttest01\n",
      "12\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "12\tQ0\t12163017\t2\t30.334925\ttest01\n",
      "12\tQ0\t11907576\t3\t28.477657\ttest01\n",
      "12\tQ0\t12048198\t4\t28.080452\ttest01\n",
      "12\tQ0\t12167152\t5\t27.794035\ttest01\n",
      "12\tQ0\t12511681\t6\t27.481863\ttest01\n",
      "12\tQ0\t12399461\t7\t27.434668\ttest01\n",
      "12\tQ0\t12196661\t8\t27.133131\ttest01\n",
      "12\tQ0\t11971762\t9\t26.188068\ttest01\n",
      "12\tQ0\t12410593\t10\t25.09195\ttest01\n",
      "13\tQ0\t11937514\t1\t37.42442\ttest01\n",
      "13\tQ0\t11824903\t2\t36.454147\ttest01\n",
      "13\tQ0\t12015117\t3\t35.84235\ttest01\n",
      "13\tQ0\t11973308\t4\t35.60395\ttest01\n",
      "13\tQ0\t11893338\t5\t33.773247\ttest01\n",
      "13\tQ0\t12172548\t6\t33.682465\ttest01\n",
      "13\tQ0\t12205098\t7\t30.203524\ttest01\n",
      "13\tQ0\t11997106\t8\t29.507065\ttest01\n",
      "13\tQ0\t11907576\t9\t28.477657\ttest01\n",
      "13\tQ0\t12242240\t10\t28.430386\ttest01\n",
      "14\tQ0\t11927599\t1\t60.162376\ttest01\n",
      "14\tQ0\t11956181\t2\t60.04175\ttest01\n",
      "14\tQ0\t12061551\t3\t57.458027\ttest01\n",
      "14\tQ0\t12011422\t4\t50.312134\ttest01\n",
      "14\tQ0\t12244330\t5\t45.684574\ttest01\n",
      "14\tQ0\t12130658\t6\t43.090042\ttest01\n",
      "14\tQ0\t12016229\t7\t42.57526\ttest01\n",
      "14\tQ0\t12359246\t8\t39.912643\ttest01\n",
      "14\tQ0\t12097643\t9\t39.07212\ttest01\n",
      "14\tQ0\t12392560\t10\t38.401424\ttest01\n",
      "15\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "15\tQ0\t12610302\t2\t32.253513\ttest01\n",
      "15\tQ0\t11855821\t3\t32.01107\ttest01\n",
      "15\tQ0\t12355291\t4\t29.548512\ttest01\n",
      "15\tQ0\t11907576\t5\t28.477657\ttest01\n",
      "15\tQ0\t12167152\t6\t27.794035\ttest01\n",
      "15\tQ0\t12393857\t7\t27.100195\ttest01\n",
      "15\tQ0\t11971762\t8\t26.188068\ttest01\n",
      "15\tQ0\t11975970\t9\t25.896713\ttest01\n",
      "15\tQ0\t12218795\t10\t25.017607\ttest01\n",
      "16\tQ0\t11871146\t1\t47.699116\ttest01\n",
      "16\tQ0\t11872956\t2\t47.6396\ttest01\n",
      "16\tQ0\t12234081\t3\t36.72107\ttest01\n",
      "16\tQ0\t12130825\t4\t33.45511\ttest01\n",
      "16\tQ0\t12079642\t5\t33.164093\ttest01\n",
      "16\tQ0\t11901185\t6\t27.668995\ttest01\n",
      "16\tQ0\t12409760\t7\t24.682886\ttest01\n",
      "16\tQ0\t11950994\t8\t24.678429\ttest01\n",
      "16\tQ0\t12101252\t9\t23.815731\ttest01\n",
      "16\tQ0\t12453616\t10\t21.900188\ttest01\n",
      "18\tQ0\t11871146\t1\t47.699116\ttest01\n",
      "18\tQ0\t11872956\t2\t47.6396\ttest01\n",
      "18\tQ0\t12234081\t3\t36.72107\ttest01\n",
      "18\tQ0\t12130825\t4\t33.45511\ttest01\n",
      "18\tQ0\t12079642\t5\t33.164093\ttest01\n",
      "18\tQ0\t12409760\t6\t24.682886\ttest01\n",
      "18\tQ0\t11950994\t7\t24.678429\ttest01\n",
      "18\tQ0\t12101252\t8\t23.815731\ttest01\n",
      "18\tQ0\t12466852\t9\t21.632769\ttest01\n",
      "18\tQ0\t12437987\t10\t21.206905\ttest01\n",
      "20\tQ0\t11871146\t1\t47.699116\ttest01\n",
      "20\tQ0\t11872956\t2\t47.6396\ttest01\n",
      "20\tQ0\t12234081\t3\t36.72107\ttest01\n",
      "20\tQ0\t12130825\t4\t33.45511\ttest01\n",
      "20\tQ0\t12079642\t5\t33.164093\ttest01\n",
      "20\tQ0\t12409760\t6\t24.682886\ttest01\n",
      "20\tQ0\t11950994\t7\t24.678429\ttest01\n",
      "20\tQ0\t12101252\t8\t23.815731\ttest01\n",
      "20\tQ0\t12466852\t9\t21.632769\ttest01\n",
      "20\tQ0\t12437987\t10\t21.206905\ttest01\n",
      "22\tQ0\t11871146\t1\t58.254944\ttest01\n",
      "22\tQ0\t11872956\t2\t47.6396\ttest01\n",
      "22\tQ0\t12234081\t3\t36.72107\ttest01\n",
      "22\tQ0\t12130825\t4\t33.45511\ttest01\n",
      "22\tQ0\t12079642\t5\t33.164093\ttest01\n",
      "22\tQ0\t12142670\t6\t31.421957\ttest01\n",
      "22\tQ0\t12050144\t7\t28.295504\ttest01\n",
      "22\tQ0\t12569124\t8\t28.24082\ttest01\n",
      "22\tQ0\t11934857\t9\t25.33506\ttest01\n",
      "22\tQ0\t12586063\t10\t25.24517\ttest01\n",
      "23\tQ0\t11871146\t1\t47.699116\ttest01\n",
      "23\tQ0\t11872956\t2\t47.6396\ttest01\n",
      "23\tQ0\t12234081\t3\t36.72107\ttest01\n",
      "23\tQ0\t12130825\t4\t33.45511\ttest01\n",
      "23\tQ0\t12079642\t5\t33.164093\ttest01\n",
      "23\tQ0\t12409760\t6\t24.682886\ttest01\n",
      "23\tQ0\t11950994\t7\t24.678429\ttest01\n",
      "23\tQ0\t12101252\t8\t23.815731\ttest01\n",
      "23\tQ0\t12466852\t9\t21.632769\ttest01\n",
      "23\tQ0\t12437987\t10\t21.206905\ttest01\n",
      "24\tQ0\t11871146\t1\t47.79355\ttest01\n",
      "24\tQ0\t11872956\t2\t47.71553\ttest01\n",
      "24\tQ0\t11786535\t3\t39.035267\ttest01\n",
      "24\tQ0\t12234081\t4\t38.641384\ttest01\n",
      "24\tQ0\t12130825\t5\t38.190712\ttest01\n",
      "24\tQ0\t12079642\t6\t36.928703\ttest01\n",
      "24\tQ0\t12468090\t7\t26.8997\ttest01\n",
      "24\tQ0\t11950994\t8\t25.550508\ttest01\n",
      "24\tQ0\t12409760\t9\t25.544056\ttest01\n",
      "24\tQ0\t12101252\t10\t23.815731\ttest01\n",
      "25\tQ0\t11871146\t1\t47.699116\ttest01\n",
      "25\tQ0\t11872956\t2\t47.6396\ttest01\n",
      "25\tQ0\t12234081\t3\t36.72107\ttest01\n",
      "25\tQ0\t12130825\t4\t33.45511\ttest01\n",
      "25\tQ0\t12079642\t5\t33.164093\ttest01\n",
      "25\tQ0\t12409760\t6\t24.682886\ttest01\n",
      "25\tQ0\t11950994\t7\t24.678429\ttest01\n",
      "25\tQ0\t12101252\t8\t23.815731\ttest01\n",
      "25\tQ0\t12466852\t9\t21.632769\ttest01\n",
      "25\tQ0\t12437987\t10\t21.206905\ttest01\n",
      "27\tQ0\t11919687\t1\t42.909748\ttest01\n",
      "27\tQ0\t12234080\t2\t41.72815\ttest01\n",
      "27\tQ0\t12048574\t3\t39.798225\ttest01\n",
      "27\tQ0\t12234076\t4\t37.98059\ttest01\n",
      "27\tQ0\t12234074\t5\t37.35634\ttest01\n",
      "27\tQ0\t12018594\t6\t37.208927\ttest01\n",
      "27\tQ0\t12163239\t7\t36.241283\ttest01\n",
      "27\tQ0\t12234073\t8\t35.617508\ttest01\n",
      "27\tQ0\t12136153\t9\t35.03256\ttest01\n",
      "27\tQ0\t12234079\t10\t33.306328\ttest01\n",
      "28\tQ0\t11919687\t1\t44.656723\ttest01\n",
      "28\tQ0\t12234080\t2\t41.72815\ttest01\n",
      "28\tQ0\t12048574\t3\t39.798225\ttest01\n",
      "28\tQ0\t12234074\t4\t39.771023\ttest01\n",
      "28\tQ0\t12018594\t5\t38.87433\ttest01\n",
      "28\tQ0\t12234076\t6\t37.98059\ttest01\n",
      "28\tQ0\t12136153\t7\t37.539948\ttest01\n",
      "28\tQ0\t12163239\t8\t36.241283\ttest01\n",
      "28\tQ0\t12234079\t9\t35.998463\ttest01\n",
      "28\tQ0\t12234073\t10\t35.617508\ttest01\n",
      "29\tQ0\t11919687\t1\t42.909748\ttest01\n",
      "29\tQ0\t12234080\t2\t41.72815\ttest01\n",
      "29\tQ0\t12048574\t3\t39.798225\ttest01\n",
      "29\tQ0\t12234076\t4\t37.98059\ttest01\n",
      "29\tQ0\t12234074\t5\t37.35634\ttest01\n",
      "29\tQ0\t12018594\t6\t37.208927\ttest01\n",
      "29\tQ0\t12163239\t7\t36.241283\ttest01\n",
      "29\tQ0\t12234073\t8\t35.617508\ttest01\n",
      "29\tQ0\t12136153\t9\t35.03256\ttest01\n",
      "29\tQ0\t12234079\t10\t33.306328\ttest01\n",
      "31\tQ0\t11919687\t1\t42.909748\ttest01\n",
      "31\tQ0\t12234080\t2\t41.72815\ttest01\n",
      "31\tQ0\t12048574\t3\t39.798225\ttest01\n",
      "31\tQ0\t12234076\t4\t37.98059\ttest01\n",
      "31\tQ0\t12234074\t5\t37.35634\ttest01\n",
      "31\tQ0\t12018594\t6\t37.208927\ttest01\n",
      "31\tQ0\t12163239\t7\t36.241283\ttest01\n",
      "31\tQ0\t12234073\t8\t35.617508\ttest01\n",
      "31\tQ0\t12136153\t9\t35.03256\ttest01\n",
      "31\tQ0\t12234079\t10\t33.306328\ttest01\n",
      "32\tQ0\t11919687\t1\t44.656723\ttest01\n",
      "32\tQ0\t12234080\t2\t41.72815\ttest01\n",
      "32\tQ0\t12048574\t3\t39.798225\ttest01\n",
      "32\tQ0\t12234074\t4\t39.771023\ttest01\n",
      "32\tQ0\t12018594\t5\t38.87433\ttest01\n",
      "32\tQ0\t12234079\t6\t38.835094\ttest01\n",
      "32\tQ0\t12234076\t7\t37.98059\ttest01\n",
      "32\tQ0\t12136153\t8\t37.539948\ttest01\n",
      "32\tQ0\t12163239\t9\t36.241283\ttest01\n",
      "32\tQ0\t12234073\t10\t35.617508\ttest01\n",
      "34\tQ0\t12213131\t1\t47.860058\ttest01\n",
      "34\tQ0\t11919687\t2\t42.98556\ttest01\n",
      "34\tQ0\t12421341\t3\t42.754757\ttest01\n",
      "34\tQ0\t12234080\t4\t41.79381\ttest01\n",
      "34\tQ0\t12048574\t5\t39.870846\ttest01\n",
      "34\tQ0\t12234076\t6\t38.959396\ttest01\n",
      "34\tQ0\t12234079\t7\t38.688793\ttest01\n",
      "34\tQ0\t12018594\t8\t38.183117\ttest01\n",
      "34\tQ0\t12234074\t9\t37.414772\ttest01\n",
      "34\tQ0\t12163239\t10\t37.241684\ttest01\n",
      "36\tQ0\t11919715\t1\t41.292984\ttest01\n",
      "36\tQ0\t11710527\t2\t34.847103\ttest01\n",
      "36\tQ0\t12451171\t3\t33.318195\ttest01\n",
      "36\tQ0\t12018165\t4\t33.16485\ttest01\n",
      "36\tQ0\t12018164\t5\t33.16225\ttest01\n",
      "36\tQ0\t12386932\t6\t33.062145\ttest01\n",
      "36\tQ0\t12610309\t7\t32.714046\ttest01\n",
      "36\tQ0\t11972156\t8\t32.515545\ttest01\n",
      "36\tQ0\t11985896\t9\t32.481308\ttest01\n",
      "36\tQ0\t12384287\t10\t31.532597\ttest01\n",
      "37\tQ0\t11710527\t1\t37.873096\ttest01\n",
      "37\tQ0\t12438685\t2\t36.74741\ttest01\n",
      "37\tQ0\t11940677\t3\t35.551914\ttest01\n",
      "37\tQ0\t12384287\t4\t35.092297\ttest01\n",
      "37\tQ0\t11827180\t5\t34.522964\ttest01\n",
      "37\tQ0\t12119098\t6\t34.174465\ttest01\n",
      "37\tQ0\t11809773\t7\t33.42332\ttest01\n",
      "37\tQ0\t12471440\t8\t33.350483\ttest01\n",
      "37\tQ0\t12451171\t9\t33.318195\ttest01\n",
      "37\tQ0\t11783007\t10\t33.24755\ttest01\n",
      "38\tQ0\t12065414\t1\t48.827408\ttest01\n",
      "38\tQ0\t12084706\t2\t42.96778\ttest01\n",
      "38\tQ0\t11710527\t3\t34.847103\ttest01\n",
      "38\tQ0\t12451171\t4\t33.318195\ttest01\n",
      "38\tQ0\t12018165\t5\t33.16485\ttest01\n",
      "38\tQ0\t12018164\t6\t33.16225\ttest01\n",
      "38\tQ0\t12386932\t7\t33.062145\ttest01\n",
      "38\tQ0\t12610309\t8\t32.714046\ttest01\n",
      "38\tQ0\t11972156\t9\t32.515545\ttest01\n",
      "38\tQ0\t11985896\t10\t32.481308\ttest01\n",
      "39\tQ0\t11710527\t1\t34.847103\ttest01\n",
      "39\tQ0\t12451171\t2\t33.318195\ttest01\n",
      "39\tQ0\t12018165\t3\t33.16485\ttest01\n",
      "39\tQ0\t12018164\t4\t33.16225\ttest01\n",
      "39\tQ0\t12386932\t5\t33.062145\ttest01\n",
      "39\tQ0\t12610309\t6\t32.714046\ttest01\n",
      "39\tQ0\t11972156\t7\t32.515545\ttest01\n",
      "39\tQ0\t11985896\t8\t32.481308\ttest01\n",
      "39\tQ0\t12384287\t9\t31.532597\ttest01\n",
      "39\tQ0\t12419898\t10\t31.500788\ttest01\n",
      "40\tQ0\t12089351\t1\t50.6517\ttest01\n",
      "40\tQ0\t11861304\t2\t49.104404\ttest01\n",
      "40\tQ0\t12040024\t3\t48.432964\ttest01\n",
      "40\tQ0\t11875714\t4\t48.07624\ttest01\n",
      "40\tQ0\t12130529\t5\t46.389465\ttest01\n",
      "40\tQ0\t12538627\t6\t44.52735\ttest01\n",
      "40\tQ0\t11981045\t7\t44.452984\ttest01\n",
      "40\tQ0\t11827180\t8\t42.74169\ttest01\n",
      "40\tQ0\t11994705\t9\t42.56342\ttest01\n",
      "40\tQ0\t12171960\t10\t42.531982\ttest01\n",
      "42\tQ0\t12213838\t1\t40.432198\ttest01\n",
      "42\tQ0\t11744384\t2\t35.598637\ttest01\n",
      "42\tQ0\t11850624\t3\t35.117916\ttest01\n",
      "42\tQ0\t11710527\t4\t34.847103\ttest01\n",
      "42\tQ0\t12510193\t5\t33.673378\ttest01\n",
      "42\tQ0\t12451171\t6\t33.318195\ttest01\n",
      "42\tQ0\t12018165\t7\t33.16485\ttest01\n",
      "42\tQ0\t12018164\t8\t33.16225\ttest01\n",
      "42\tQ0\t12386932\t9\t33.062145\ttest01\n",
      "42\tQ0\t12610309\t10\t32.714046\ttest01\n",
      "44\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "44\tQ0\t11907576\t2\t28.477657\ttest01\n",
      "44\tQ0\t12167152\t3\t27.794035\ttest01\n",
      "44\tQ0\t11971762\t4\t26.188068\ttest01\n",
      "44\tQ0\t12645611\t5\t24.40221\ttest01\n",
      "44\tQ0\t12393857\t6\t24.205702\ttest01\n",
      "44\tQ0\t12018119\t7\t23.876041\ttest01\n",
      "44\tQ0\t12223891\t8\t23.63093\ttest01\n",
      "44\tQ0\t11935290\t9\t23.387932\ttest01\n",
      "44\tQ0\t12148460\t10\t23.304388\ttest01\n",
      "45\tQ0\t11799118\t1\t44.866287\ttest01\n",
      "45\tQ0\t12459172\t2\t41.760845\ttest01\n",
      "45\tQ0\t12058063\t3\t37.182938\ttest01\n",
      "45\tQ0\t11997453\t4\t36.65457\ttest01\n",
      "45\tQ0\t11824903\t5\t36.454147\ttest01\n",
      "45\tQ0\t11879186\t6\t34.5406\ttest01\n",
      "45\tQ0\t12554655\t7\t32.66574\ttest01\n",
      "45\tQ0\t11907576\t8\t28.477657\ttest01\n",
      "45\tQ0\t12167152\t9\t27.794035\ttest01\n",
      "45\tQ0\t11971762\t10\t26.188068\ttest01\n",
      "46\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "46\tQ0\t11907576\t2\t28.477657\ttest01\n",
      "46\tQ0\t12167152\t3\t27.794035\ttest01\n",
      "46\tQ0\t11971762\t4\t26.188068\ttest01\n",
      "46\tQ0\t12645611\t5\t24.40221\ttest01\n",
      "46\tQ0\t12393857\t6\t24.205702\ttest01\n",
      "46\tQ0\t12018119\t7\t23.876041\ttest01\n",
      "46\tQ0\t12148460\t8\t23.304388\ttest01\n",
      "46\tQ0\t12399461\t9\t23.044468\ttest01\n",
      "46\tQ0\t12203811\t10\t22.933348\ttest01\n",
      "48\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "48\tQ0\t11907576\t2\t28.477657\ttest01\n",
      "48\tQ0\t12167152\t3\t27.794035\ttest01\n",
      "48\tQ0\t11971762\t4\t26.188068\ttest01\n",
      "48\tQ0\t12645611\t5\t24.40221\ttest01\n",
      "48\tQ0\t12393857\t6\t24.205702\ttest01\n",
      "48\tQ0\t12018119\t7\t23.876041\ttest01\n",
      "48\tQ0\t12148460\t8\t23.304388\ttest01\n",
      "48\tQ0\t12399461\t9\t23.044468\ttest01\n",
      "48\tQ0\t12203811\t10\t22.933348\ttest01\n",
      "50\tQ0\t12234080\t1\t45.42011\ttest01\n",
      "50\tQ0\t11919687\t2\t45.18249\ttest01\n",
      "50\tQ0\t12048574\t3\t39.798225\ttest01\n",
      "50\tQ0\t12234073\t4\t38.63184\ttest01\n",
      "50\tQ0\t12136153\t5\t38.411358\ttest01\n",
      "50\tQ0\t12234076\t6\t37.98059\ttest01\n",
      "50\tQ0\t12234074\t7\t37.35634\ttest01\n",
      "50\tQ0\t12018594\t8\t37.208927\ttest01\n",
      "50\tQ0\t12163239\t9\t36.241283\ttest01\n",
      "50\tQ0\t12237116\t10\t34.60788\ttest01\n"
     ]
    }
   ],
   "source": [
    "# this prints out (it is a shell command) the content of the file baseline.run \n",
    "!cat baseline.run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Tip: Write a line to `run_file` using `run_file.write(line)`. \n",
    "> The newline character is: `'\\n'`. Before writing a number to\n",
    "> the file, cast it to a string using `str()`.\n",
    ">\n",
    "> The TREC Submission guidelines allow you to submit up to 1000\n",
    "> documents per topic. Keep this in mind!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------------------------------------------------------------------\n",
    "--------------------------------------------------------------------------\n",
    "--------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 03: Index improvements: Tokenization\n",
    "<span style=\"background:red; color: white;\">__You are advised to work on this part after Lecture 02 (Conceptual Indexing)__</span>\n",
    "\n",
    "\n",
    "\n",
    "## Background\n",
    "The following part of the assignment requires some self-study of the ElasticSearch tools to support the improvemnet of the indexing. Please read the:\n",
    "* [Index Settings and Mappings](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/indices-create-index.html).\n",
    "* Elasticsearch [Analyzers](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/analysis.html) contain many options for improving your search engine.\n",
    "\n",
    "> You are suggested to use the [Python Elasticsearch Client](https://elasticsearch-py.readthedocs.io) library documentation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Exercise 03.A: _chat language analyzer_\n",
    "\n",
    "Read the documentation for [Custom Analyzer](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/analysis-custom-analyzer.html). \n",
    "Make a custom analyzer for _English chat language_. The analyzer should do the following:\n",
    "* change common abbreviations to the full forms: \n",
    "  * _b4_ to _before_, \n",
    "  * _abt_ to _about_, \n",
    "  * _chk_ to _check_, \n",
    "  * _dm_ to _direct message_,\n",
    "  * _f2f_ to _face-to-face_\n",
    "* use the _standard_ tokenizer;\n",
    "* put everything to lower-case;\n",
    "* filter English stopwords."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'tokens': [{'end_offset': 4,\n",
      "             'position': 0,\n",
      "             'start_offset': 0,\n",
      "             'token': 'done',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 10,\n",
      "             'position': 2,\n",
      "             'start_offset': 8,\n",
      "             'token': 'before',\n",
      "             'type': 'SYNONYM'},\n",
      "            {'end_offset': 16,\n",
      "             'position': 3,\n",
      "             'start_offset': 12,\n",
      "             'token': 'what',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 20,\n",
      "             'position': 4,\n",
      "             'start_offset': 17,\n",
      "             'token': 'about',\n",
      "             'type': 'SYNONYM'},\n",
      "            {'end_offset': 23,\n",
      "             'position': 5,\n",
      "             'start_offset': 21,\n",
      "             'token': 'direct',\n",
      "             'type': 'SYNONYM'},\n",
      "            {'end_offset': 26,\n",
      "             'position': 6,\n",
      "             'start_offset': 24,\n",
      "             'token': 'me',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 26,\n",
      "             'position': 6,\n",
      "             'start_offset': 24,\n",
      "             'token': 'message',\n",
      "             'type': 'SYNONYM'}]}\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "import pprint\n",
    "tweet_analyzer = {\n",
    "    # BEGIN ANSWER\n",
    "    \"settings\": {\n",
    "        \"analysis\": {\n",
    "            \"filter\": {\n",
    "                \"english_stop\": {\n",
    "                    \"type\": \"stop\",\n",
    "                    \"stopwords\": \"_english_\"\n",
    "                },\n",
    "                \"chat_abbreviations\": {\n",
    "                    \"type\": \"synonym\",\n",
    "                    \"synonyms\": [\n",
    "                        \"b4 => before\",\n",
    "                        \"abt => about\",\n",
    "                        \"chk => check\",\n",
    "                        \"dm => direct message\",\n",
    "                        \"f2f => face to face\"\n",
    "                    ]\n",
    "                }\n",
    "            },\n",
    "            \"analyzer\": {\n",
    "                \"tweet_analyzer\": {\n",
    "                    \"type\": \"custom\",\n",
    "                    \"tokenizer\": \"standard\",\n",
    "                    \"filter\": [\n",
    "                        \"lowercase\",\n",
    "                        \"chat_abbreviations\",\n",
    "                        \"english_stop\"\n",
    "                    ]\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    },\n",
    "    \"mappings\": {\n",
    "    \"properties\": {\n",
    "      \"TI\": {\n",
    "          \"type\": \"text\",\n",
    "          \"analyzer\": \"tweet_analyzer\"\n",
    "      },\n",
    "      \"AB\": {\n",
    "          \"type\": \"text\",\n",
    "          \"analyzer\": \"tweet_analyzer\"\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "    # END ANSWER\n",
    "}\n",
    "\n",
    "# create the index, but don't index any documents:\n",
    "create_index(es, 'genomics', body=tweet_analyzer)\n",
    "body = { \"field\": \"all\", \"analyzer\": \"tweet_analyzer\", \"text\": \"done it b4! what abt dm me?\"}\n",
    "tokens = es.indices.analyze(index='genomics', body=body)\n",
    "pprint.pprint(tokens)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 03.B: Stemmers\n",
    "\n",
    "Referring at Exercise 02.A, we have seen that queries like `molecule` and `molecular` retrieve different sets of documents. Lemmatizer and stemmers can help the indexing and search of 'similar' terms, and retrieve more consistent sets of documents.\n",
    "\n",
    "__Use the ElasticSearch [Stemming](https://www.elastic.co/guide/en/elasticsearch/reference/current/stemming.html) to index the document collection. Then retrieve documents with the queries `molecule` and `molecular` and comment on the eventual differences with the previous query results.__\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# body = {\n",
    "#   # BEGIN ANSWER\n",
    "\n",
    "#   # END ANSWER\n",
    "# }\n",
    "\n",
    "# body = {\n",
    "#   # BEGIN ANSWER\n",
    "\n",
    "#   # END ANSWER\n",
    "# }\n",
    "\n",
    "body = {\n",
    "  \"settings\": {\n",
    "    \"analysis\": {\n",
    "      \"analyzer\": {\n",
    "        \"english_stemming_analyzer\": {\n",
    "          \"tokenizer\": \"standard\",\n",
    "          \"filter\": [\"lowercase\", \"english_stemmer\"]\n",
    "        }\n",
    "      },\n",
    "      \"filter\": {\n",
    "        \"english_stemmer\": {\n",
    "          \"type\": \"stemmer\",\n",
    "          \"name\": \"english\"\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "  },\n",
    "  \"mappings\": {\n",
    "    \"properties\": {\n",
    "      \"TI\": {\n",
    "          \"type\": \"text\",\n",
    "          \"analyzer\": \"english_stemming_analyzer\"\n",
    "      },\n",
    "      \"AB\": {\n",
    "          \"type\": \"text\",\n",
    "          \"analyzer\": \"english_stemming_analyzer\"\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(263080, [])"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Connect to the ElasticSearch server\n",
    "es = elasticsearch.Elasticsearch(host='localhost')  # in case you use Docker, the host is 'elasticsearch'\n",
    "\n",
    "# Index the collection into the index called 'genomics-stem'\n",
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-stem', body)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Retrieve documents with the queries `molecule` and `molecular` and comment on the eventual differences with the previous query results.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'took': 5910, 'timed_out': False, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}, 'hits': {'total': {'value': 7345, 'relation': 'eq'}, 'max_score': 10.767412, 'hits': [{'_index': 'genomics-stem', '_type': '_doc', '_id': '21807', '_score': 10.767412, '_source': {'AB': \"OBJECTIVE: With Helios gene gun, the report genes EGFP and Lac-Z were transfected to several cultured mammalian tumor cell lines in vitro and mice skin in vivo, respectively. A stable long time exogene's expression were got. Then, the pWRG3142, which carried mGM-CSF gene was delivered to the abdominal skin of C57BL/6 mice using gene gun. Pathological sections showed the local transproteins' expression companying with a profound inflammation reaction characterized by neutrophilic infiltration. ELISA assay of transfected mouse's serum sample indicated a high improvement of transgenic proteins level, which demonstrated that transgenic GM-CSF secreted from treated skin into the bloodstream effectively. In B16 melanoma tumor model, mice immunized with mGM-CSF expression plasmids could be partially protected from 1 x 10(5) B16 cells challenge and exhibited a drastically reduced tumor growth rate. Finally, we conclucled that exogenes could be transfected into cultured cell lines and mice skin effectively by gene gun technology, and gene gun mediated in vivo delivery of GM-CSF cDNA should be further developed for potential clinical testing as an approach for human cancer gene therapy.\", 'AD': 'National Lab of Molecule Oncology, Cancer Institute, PUMC & CAMS, Beijing, 100021 China.', 'CY': 'China', 'DA': '20020118', 'DCOM': '20020711', 'DP': '2000 Jul', 'EDAT': '2002/01/19 10:00', 'IP': '7', 'IS': '0376-2491', 'JID': '7511141', 'LA': 'chi', 'MHDA': '2002/07/12 10:01', 'PG': '522-5', 'PMID': '11798811', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Zhonghua Yi Xue Za Zhi 2000 Jul;80(7):522-5.', 'TA': 'Zhonghua Yi Xue Za Zhi', 'TI': '[The use of gene gun in cancer gene therapy]', 'UI': '21658031', 'VI': '80', 'MH': ['Animal', 'Biolistics/*methods', 'Cell Division/genetics', 'English Abstract', 'Gene Therapy/*methods', 'Granulocyte-Macrophage Colony-Stimulating Factor/genetics', 'Human', 'Melanoma, Experimental/genetics/pathology/*therapy', 'Mice', 'Mice, Inbred C57BL', 'Plasmids/genetics', 'Tumor Cells, Cultured'], 'RN': ['0 (Plasmids)', '83869-56-1 (Granulocyte-Macrophage Colony-Stimulating Factor)'], 'FAU': ['Zha, Y', 'Lin, C', 'Liang, X']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '132671', '_score': 9.435863, '_source': {'AB': 'OBJECTIVE: To determine the effect of stage III and IV endometriosis on serum concentration of soluble intracellular adhesion molecule-1 (ICAM-1), vascular cellular adhesion molecule-1 (VCAM-1), and E-selectin. METHODS: Sera from women with laparoscopically proven stage III and IV endometriosis were analyzed for the presence of soluble ICAM-1, VCAM-1, and E-selectin using enzyme-linked immunosorbent assays. Concentrations of these molecules were compared with those of laparoscopically proven disease-free controls. RESULTS: When compared with controls, women with stage III and IV endometriosis had higher serum concentrations of soluble VCAM-1, lower serum concentration of soluble ICAM-1, and no difference in serum concentration of soluble E-selectin. CONCLUSION: Significant aberrations in levels of circulating adhesion molecules were found in women with stage III and IV endometriosis. These findings might shed light on the pathogenesis of endometriosis and be useful in the development of biochemical markers for disease stage.', 'AD': 'Department of Obstetrics and Gynecology, University of Missouri, Columbia, Missouri, USA. breton.barrier@lackland.af.mil', 'CY': 'United States', 'DA': '20020419', 'DCOM': '20021025', 'DP': '2002 Mar-Apr', 'EDAT': '2002/04/20 10:00', 'IP': '2', 'IS': '1071-5576', 'JID': '9433806', 'LA': 'eng', 'MHDA': '2002/10/31 04:00', 'PG': '98-101', 'PMID': '11963839', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'J Soc Gynecol Investig 2002 Mar-Apr;9(2):98-101.', 'TA': 'J Soc Gynecol Investig', 'TI': 'Expression of soluble adhesion molecules in sera of women with stage III and IV endometriosis.', 'UI': '21961150', 'VI': '9', 'MH': ['Biological Markers/blood', 'E-Selectin/*blood', 'Endometriosis/*blood/pathology', 'Female', 'Human', 'Intercellular Adhesion Molecule-1/*blood', 'Vascular Cell Adhesion Molecule-1/*blood'], 'RN': ['0 (Biological Markers)', '0 (E-Selectin)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Barrier, Breton F', 'Sharpe-Timms, Kathy L']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '148107', '_score': 9.326713, '_source': {'AID': '106195 [pii]', 'CY': 'France', 'DA': '20020430', 'DCOM': '20020627', 'DP': '2002 Apr', 'EDAT': '2002/05/01 10:00', 'IP': '2 Pt 2', 'IS': '0368-2315', 'JID': '0322206', 'LA': 'fre', 'LR': '20021101', 'MHDA': '2002/06/28 10:01', 'PG': '1S5-1S6', 'PMID': '11981482', 'PST': 'ppublish', 'PT': 'Editorial', 'SB': 'IM', 'SO': 'J Gynecol Obstet Biol Reprod (Paris) 2002 Apr;31(2 Pt 2):1S5-1S6.', 'TA': 'J Gynecol Obstet Biol Reprod (Paris)', 'TI': '[LH Day: floodlights on a forgotten compound]', 'TT': 'LH Day: coup de projecteur sur une molecule oubliee.', 'UI': '21977764', 'VI': '31', 'MH': ['Female', 'Human', 'Luteinizing Hormone/*physiology', 'Menstrual Cycle', 'Oocytes/physiology', 'Ovarian Follicle/physiology', 'Ovulation Induction', 'Polycystic Ovary Syndrome', 'Receptors, LH/physiology'], 'RN': ['0 (Receptors, LH)', '9002-67-9 (Luteinizing Hormone)'], 'FAU': ['Fanchin, R']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '249043', '_score': 9.312457, '_source': {'AB': 'The structures of two potential anti-human immunodeficiency virus type 1 (HIV-1) non-nucleoside reverse transcriptase inhibitors (NNRTI), namely 1-benzyl-5-oxo-2-phenylpyrrolidine-2-carboxamide, C(18)H(18)N(2)O(2), (III), and 2-(4-isopropoxyphenyl)-5-oxo-1-(4-tolyl)pyrrolidine-2-carbonitrile, C(21)H(22)N(2)O(2), (IV), have been investigated by X-ray diffraction, confirming the butterfly-like conformation of both compounds. The pyrrolidine ring is in an envelope conformation in (III) and a half-chair conformation in (IV). Two intermolecular N[bond]H...O hydrogen bonds are present in the crystal structure of (III), with N...O distances of 2.995 (2) and 2.927 (2) A.', 'AD': 'Molecule Structure Research Centre, National Academy of Sciences RA, Azatutyan Ave. 26, 375014 Yerevan, Republic of Armenia.', 'AID': 'S0108270102009241 [pii]', 'CY': 'Denmark', 'DA': '20020702', 'DCOM': '20020917', 'DP': '2002 Jul', 'EDAT': '2002/07/03 10:00', 'IP': 'Pt 7', 'IS': '0108-2701', 'JID': '8305826', 'LA': 'eng', 'MHDA': '2002/07/03 10:00', 'PG': 'o399-401', 'PHST': '2002/Jun/20 [epublish]', 'PMID': '12094058', 'PST': 'ppublish', 'PT': 'Journal Article', 'SO': 'Acta Crystallogr C 2002 Jul;58(Pt 7):o399-401.', 'TA': 'Acta Crystallogr C', 'TI': '1-Substituted derivatives of 2-aryl-5-oxopyrrolidine-2-carboxylic acid.', 'UI': '22088002', 'VI': '58', 'FAU': ['Karapetyan, Harutyun', 'Tamazyan, Rafael', 'Martirosyan, Ashot', 'Hovhannesyan, Vahagn', 'Gasparyan, Sahak']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '61882', '_score': 8.993986, '_source': {'AB': 'Synthetic molecular receptors that completely surround their target molecules can be created through the use of noncovalent interactions. These molecular capsules selectively sequester guest molecules from the influence of bulk solvent and other molecules on the basis of size, shape, and chemical complementarity. This reversible isolation spawns unique behavior within the confines of the host; the catalysis of chemical reactions and the stabilization of reactive species are possible outcomes that have been recently demonstrated. Compartmentalization of reagents can also have a dramatic effect on reactions that take place outside of the capsule, producing nonlinear kinetics in relatively simple reaction systems.', 'AD': 'The Skaggs Institute for Chemical Biology and the Department of Chemistry, The Scripps Research Institute, 10550 North Torrey Pines Road, MB-26, La Jolla, CA 92037, USA.', 'AID': '042663699 [pii]', 'CY': 'United States', 'DA': '20020417', 'DCOM': '20020614', 'DP': '2002 Apr 16', 'EDAT': '2002/03/07 10:00', 'IP': '8', 'IS': '0027-8424', 'JID': '7505876', 'LA': 'eng', 'LR': '20030103', 'MHDA': '2002/06/18 10:01', 'PG': '4775-7', 'PHST': '2002/Mar/05 [aheadofprint]', 'PMID': '11880604', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Proc Natl Acad Sci U S A 2002 Apr 16;99(8):4775-7.', 'TA': 'Proc Natl Acad Sci U S A', 'TI': 'Molecules within molecules: recognition through self-assembly.', 'UI': '21957116', 'VI': '99', 'MH': ['Catalysis', 'Chemistry/*methods', 'Chemistry, Physical/methods', 'Kinetics', 'Macromolecular Systems', 'Models, Molecular', 'Time Factors'], 'RN': ['0 (Macromolecular Systems)'], 'FAU': ['Hof, Fraser', 'Rebek, Julius Jr']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '280977', '_score': 8.912884, '_source': {'AB': 'To determine whether soluble adhesion molecules are affected in muscular dystrophy, we measured serum levels of creatine kinase (CK), soluble vascular cell adhesion molecule-1 (sVCAM-1), soluble intercellular adhesion molecule-1 (sICAM-1), soluble (s) E-selectin, and fibrin and fibrinogen degradation products (FDP) in 25 patients with Duchenne muscular dystrophy (DMD), 7 with Becker muscular dystrophy, 7 with Fukuyama type congenital muscular dystrophy, 6 with myotonic dystrophy (MyD), and 5 with spinal muscular atrophy (SMA) type 2, and also serum sVCAM-1, sICAM-1, and sE-selectin in 9 healthy controls. The levels of sVCAM-1 in the patients with DMD were 367.0-852.0 ng/ml (552.8 +/', 'AD': 'Department of Neurology, Toneyama National Hospital, Toyonaka, Osaka.', 'CY': 'Japan', 'DA': '20020723', 'DCOM': '20020730', 'DP': '2002 Jul', 'EDAT': '2002/07/24 10:00', 'IP': '4', 'IS': '0029-0831', 'JID': '0215224', 'LA': 'jpn', 'LR': '20021101', 'MHDA': '2002/07/31 10:01', 'PG': '326-9', 'PMID': '12134684', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'No To Hattatsu 2002 Jul;34(4):326-9.', 'TA': 'No To Hattatsu', 'TI': '[Soluble adhesion molecules in muscular dystrophy]', 'UI': '22130743', 'VI': '34', 'MH': ['Adolescent', 'Adult', 'Child', 'Child, Preschool', 'E-Selectin/blood', 'English Abstract', 'Female', 'Human', 'Intercellular Adhesion Molecule-1/*blood', 'Male', 'Muscular Dystrophies/*blood', 'Solubility', 'Vascular Cell Adhesion Molecule-1/*blood'], 'RN': ['0 (E-Selectin)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Saito, Toshio', 'Yamamoto, Yuko', 'Shinno, Susumu']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '502100', '_score': 8.912884, '_source': {'CY': 'China', 'DA': '20030127', 'DCOM': '20030212', 'DP': '2000 Jul', 'EDAT': '2003/01/28 04:00', 'IP': '3', 'IS': '0559-7765', 'JID': '20730140R', 'LA': 'chi', 'MHDA': '2003/02/14 04:00', 'PG': '257-60', 'PMID': '12545718', 'PST': 'ppublish', 'PT': 'Review, Tutorial', 'RF': '8', 'SB': 'IM', 'SO': 'Sheng Li Ke Xue Jin Zhan 2000 Jul;31(3):257-60.', 'TA': 'Sheng Li Ke Xue Jin Zhan', 'TI': '[Adhesion molecules and inflammation following cerebral ischemia]', 'UI': '22434060', 'VI': '31', 'MH': ['Animal', 'Brain Ischemia/complications/*metabolism', 'Cell Adhesion Molecules/biosynthesis', 'Human', 'Inflammation/etiology/*metabolism', 'Integrins/biosynthesis', 'Intercellular Adhesion Molecule-1/*biosynthesis', 'Selectins/*biosynthesis', 'Vascular Cell Adhesion Molecule-1/biosynthesis'], 'RN': ['0 (Cell Adhesion Molecules)', '0 (Integrins)', '0 (Selectins)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Xu, H L', 'Fen, Y P']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '249039', '_score': 8.834926, '_source': {'AB': 'The structures of two potential anti-human immunodeficiency virus type 1 (HIV-1) non-nucleoside reverse transcriptase inhibitors (NNRTI), namely 1-(2-chlorobenzoyl)-2-phenylpyrrolidine-2-carboxamide, C(18)H(17)ClN(2)O(2), and 1-(2-furoyl)-2-phenylpyrrolidine-2-carboxamide, C(16)H(16)N(2)O(3), have been investigated by X-ray diffraction and the butterfly-like conformation established in each case. The pyrrolidine ring has the same half-chair conformation in both structures.', 'AD': 'Molecule Structure Research Centre, National Academy of Sciences RA, Azatutyan Ave. 26, 375014 Yerevan, Republic of Armenia. rafael@msrc.am', 'AID': 'S0108270102008636 [pii]', 'CY': 'Denmark', 'DA': '20020702', 'DCOM': '20020917', 'DP': '2002 Jul', 'EDAT': '2002/07/03 10:00', 'IP': 'Pt 7', 'IS': '0108-2701', 'JID': '8305826', 'LA': 'eng', 'MHDA': '2002/07/03 10:00', 'PG': 'o386-8', 'PHST': '2002/Jun/12 [epublish]', 'PMID': '12094054', 'PST': 'ppublish', 'PT': 'Journal Article', 'SO': 'Acta Crystallogr C 2002 Jul;58(Pt 7):o386-8.', 'TA': 'Acta Crystallogr C', 'TI': '1-Substituted derivatives of 2-phenylpyrrolidine-2-carboxamide.', 'UI': '22087998', 'VI': '58', 'FAU': ['Tamazyan, Rafael', 'Karapetyan, Harutyun', 'Martirosyan, Ashot', 'Hovhannesyan, Vahagn', 'Gasparyan, Sahak']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '63151', '_score': 8.751207, '_source': {'AID': 'S0021915001007572 [pii]', 'CY': 'Ireland', 'DA': '20020307', 'DCOM': '20020716', 'DP': '2002 Mar', 'EDAT': '2002/03/08 10:00', 'IP': '1', 'IS': '0021-9150', 'JID': '0242543', 'LA': 'eng', 'LR': '20020808', 'MHDA': '2002/07/18 10:01', 'PG': '243-4', 'PMID': '11882338', 'PST': 'ppublish', 'PT': 'Letter', 'SB': 'IM', 'SO': 'Atherosclerosis 2002 Mar;161(1):243-4.', 'TA': 'Atherosclerosis', 'TI': 'Serum levels of adhesion molecules correlate with insulin resistance.', 'UI': '21876766', 'VI': '161', 'MH': ['Cell Adhesion Molecules/*blood', 'Diabetes Mellitus, Non-Insulin-Dependent/blood', 'E-Selectin/blood', 'Human', '*Insulin Resistance', 'Intercellular Adhesion Molecule-1/blood', 'Male', 'Middle Age', 'Vascular Cell Adhesion Molecule-1/blood'], 'RN': ['0 (Cell Adhesion Molecules)', '0 (E-Selectin)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': ['Matsumoto, Kazunari', 'Sera, Yasunori', 'Miyake, Seibei', 'Ueki, Yukitaka']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '443112', '_score': 8.751207, '_source': {'AB': 'Adhesion molecules are important in cell-cell and cell-basement membrane interactions. They are intimately involved in inflammatory reactions and a role in tumour progression has been postulated. E-selectin, intercellular adhesion molecule-1 (ICAM-1), and vascular cell adhesion molecule-1 (VCAM-1) play a role in cell adhesion to the vascular endothelium, and may have a role in tumour cell dissemination. Soluble forms of these molecules have been described and this study was established to examine these adhesion molecules in patients with breast carcinoma. Serum was obtained from 92 patients with breast carcinoma and 31 age-matched patients with benign breast disease. All samples were obtained prior to surgery. Soluble levels of E-selectin, ICAM-1, and VCAM-1 were significantly elevated in patients with Stage 4 disease compared with controls. (E-selectin 88.6 (47.9) versus 51.4 (18.4) ng/ml; P<0.001: ICAM-1 447 (249) versus 244 (79) ng/ml; P<0.001: VCAM-1 779 (159) versus 552 (135) ng/ml; P<0.001 results expressed on mean (SEM) SD placed above this.). The prognostic value of the adhesion molecules was examined. In patients with Stage 2 disease, elevated VCAM-1 was predictive of decreased survival, even when corrected for T and N status. Adhesion molecules are elevated in patients with advanced disease and elevation in VCAM-1 has prognostic significance in patients with breast carcinoma.', 'AD': 'National Breast Cancer Research Institute and Department of Surgery, University College Hospital, Galway, Ireland.', 'AID': 'S0959804902002186 [pii]', 'CY': 'England', 'DA': '20021120', 'DCOM': '20030121', 'DP': '2002 Nov', 'EDAT': '2002/11/21 04:00', 'IP': '17', 'IS': '0959-8049', 'JID': '9005373', 'LA': 'eng', 'MHDA': '2003/01/22 04:00', 'PG': '2252-7', 'PMID': '12441261', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Eur J Cancer 2002 Nov;38(17):2252-7.', 'TA': 'Eur J Cancer', 'TI': 'Soluble adhesion molecules (E-selectin, ICAM-1 and VCAM-1) in breast carcinoma.', 'UI': '22328984', 'VI': '38', 'MH': ['Breast Neoplasms/*metabolism/pathology', 'E-Selectin/*metabolism', 'Female', 'Human', 'Intercellular Adhesion Molecule-1/*metabolism', 'Neoplasm Staging/methods', 'Prognosis', 'Survival Analysis', 'Tumor Markers, Biological/*metabolism', 'Vascular Cell Adhesion Molecule-1/*metabolism'], 'RN': ['0 (E-Selectin)', '0 (Tumor Markers, Biological)', '0 (Vascular Cell Adhesion Molecule-1)', '126547-89-5 (Intercellular Adhesion Molecule-1)'], 'FAU': [\"O'Hanlon, D M\", 'Fitzsimons, H', 'Lynch, J', 'Tormey, S', 'Malone, C', 'Given, H F']}}]}}\n",
      "{'took': 27, 'timed_out': False, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}, 'hits': {'total': {'value': 31569, 'relation': 'eq'}, 'max_score': 8.257235, 'hits': [{'_index': 'genomics-stem', '_type': '_doc', '_id': '14835', '_score': 8.257235, '_source': {'AB': 'Black pigment cells, melanophores, e.g. located in the epidermis and dermis of frogs, are large flat cells having intracellular black pigment granules, called melanosomes. Due to a large size, high optical contrast, and quick response to drugs, melanophores are attractive as biosensors as well as for model studies of intracellular processes; e.g. organelle transport and G-protein coupled receptors. The geometry of melanosomes from African clawed toad, Xenopus laevis, has been measured using scanning force microscopy (SFM). Three-dimensional images from SFM were used to measure height, width, and length of the melanosomes (100 from aggregated cells and 100 from dispersed cells). The volumes of melanosomes isolated from aggregated and dispersed melanophores were significantly different (P < 0.05, n=200). The average ellipsoidal volume was 0.14+/-0.01 (aggregated) and 0.17+/-0.01 microm3 (dispersed), a difference of 18%. The average major diameter was 810+/-20 and 880+/-20 nm for aggregated and dispersed melanosomes, respectively. To our knowledge, this is the first time SFM has been used to study melanosomes. This may provide an alternative non-destructive technique that may be particularly suitable for studying morphological aspects of various melanin granules.', 'AD': 'Department of Biomedical Engineering, Linkoping University, Sweden. marte@ifm.liu.se', 'CY': 'Denmark', 'DA': '20011231', 'DCOM': '20020626', 'DP': '2001 Dec', 'EDAT': '2002/01/05 10:00', 'IP': '6', 'IS': '0893-5785', 'JID': '8800247', 'LA': 'eng', 'MHDA': '2002/06/27 10:01', 'PG': '445-9', 'PMID': '11775056', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Pigment Cell Res 2001 Dec;14(6):445-9.', 'TA': 'Pigment Cell Res', 'TI': 'Volume changes of individual melanosomes measured by scanning force microscopy.', 'UI': '21630794', 'VI': '14', 'MH': ['Animal', 'Cell Aggregation/physiology', 'Cell Size/physiology', 'Melanosomes/metabolism/*ultrastructure', 'Microscopy, Atomic Force', 'Molecular Motors/metabolism/ultrastructure', 'Skin/metabolism/*ultrastructure', \"Support, Non-U.S. Gov't\", 'Xenopus laevis/*anatomy & histology/metabolism'], 'RN': ['0 (Molecular Motors)'], 'FAU': ['Testorf, M F', 'Roback, K', 'Lundstrom, I', 'Svensson, S P']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '23266', '_score': 8.257235, '_source': {'AB': 'Among filamentous fungi, mitosis has been studied in-depth in just a few species. The mitotic apparatuses in the ascomycetous Fusarium spp. are the most clearly and readily visualized in vivo within this group; fluorescent labeling is unnecessary. This superior cytological tractability has enabled detailed studies and revealing experiments that have led the way toward a more complete understanding of fungal mitosis. Some of the most important discoveries include the role of half-spindles in development of the bipolar spindle, the existence of true kinetochores in fungi, the unorthodox chromosome configurations and movements comprising metaphase and anaphase A, the attachment of astral microtubules to the plasmalemma, the role of the astral pulling force in elongating the spindle, an inwardly directed force within the spindle, and microtubule cross-bridging in both spindle and asters. Recent research has focused on the roles of microtubuleassociated motor proteins in Fusarium solani f. sp. pisi (anamorph of Nectria haematococca). Cytoplasmic dynein was shown to be involved in the development and/or maintenance of mitotic asters and necessary for motility and functionality of the interphase spindle pole body. The inwardly directed force within the anaphase spindle was shown to be produced by a kinesin-related protein, NhKRP1. Because of its superior cytological tractability, the considerable and unique knowledge we have of many aspects of its mitosis, and its genetic tractability, Fusarium solani f. sp. pisi is a good choice for further investigations of mitosis in filamentous fungi.', 'AD': 'Department of Plant Pathology, College of Agriculture and Life Sciences, Cornell University, Ithaca, New York 14853, USA.', 'CY': 'United States', 'DA': '20020123', 'DCOM': '20020621', 'DP': '2002', 'EDAT': '2002/01/24 10:00', 'IS': '0074-7696', 'JID': '2985180R', 'LA': 'eng', 'MHDA': '2002/06/22 10:01', 'PG': '239-63', 'PMID': '11804038', 'PST': 'ppublish', 'PT': 'Review, Tutorial', 'RF': '67', 'SB': 'IM', 'SO': 'Int Rev Cytol 2002;212:239-63.', 'TA': 'Int Rev Cytol', 'TI': 'Mitosis and motor proteins in the filamentous ascomycete, Nectria haematococca, and some related fungi.', 'UI': '21662383', 'VI': '212', 'MH': ['Ascomycota/*metabolism/ultrastructure', 'Fusarium/*metabolism/ultrastructure', 'Microtubules/metabolism/ultrastructure', 'Mitosis/*physiology', 'Mitotic Spindle Apparatus/*metabolism/ultrastructure', 'Models, Biological', 'Molecular Motors/*metabolism', \"Support, U.S. Gov't, Non-P.H.S.\"], 'RN': ['0 (Molecular Motors)'], 'FAU': ['Aist, James R']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '23267', '_score': 8.257235, '_source': {'AB': 'There has been a great deal of interest in how the microtubule array of the axon is established and maintained. In an early model, it was proposed that microtubules are actively transported from the cell body of the neuron down the length of the axon. This model has been contested over the years in favor of very different models based on stationary microtubules. It appears that a corner has finally been turned in this long-standing controversy. It is now clear that cells contain molecular motor proteins capable of transporting microtubules and that microtubule transport is an essential component in the formation of microtubule arrays across many cells types. A wide variety of cell biological approaches have provided strong indirect evidence that microtubules are indeed transported within axons, and new live-cell imaging approaches are beginning to permit the direct visualization of this transport. The molecules and mechanisms that transport microtubules within axons are also under intense study.', 'AD': 'Department of Neurobiology and Anatomy, MCP Hahnemann University, Philadelphia, Pennsylvania 19129, USA.', 'CY': 'United States', 'DA': '20020123', 'DCOM': '20020621', 'DP': '2002', 'EDAT': '2002/01/24 10:00', 'IS': '0074-7696', 'JID': '2985180R', 'LA': 'eng', 'MHDA': '2002/06/22 10:01', 'PG': '41-62', 'PMID': '11804039', 'PST': 'ppublish', 'PT': 'Review, Tutorial', 'RF': '49', 'SB': 'IM', 'SO': 'Int Rev Cytol 2002;212:41-62.', 'TA': 'Int Rev Cytol', 'TI': 'Microtubule transport in the axon.', 'UI': '21662379', 'VI': '212', 'MH': ['Animal', 'Axonal Transport/drug effects/*physiology', 'Axons/*metabolism/ultrastructure', 'Human', 'Microtubules/*metabolism', 'Molecular Motors/*physiology', 'Nervous System/*metabolism/ultrastructure', 'Protein Transport/drug effects/*physiology', \"Support, U.S. Gov't, P.H.S.\"], 'RN': ['0 (Molecular Motors)'], 'FAU': ['Baas, Peter W']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '27670', '_score': 8.257235, '_source': {'AD': 'The Journal of General Physiology, USA.', 'CY': 'United States', 'DA': '20020129', 'DCOM': '20020926', 'DP': '2002 Feb', 'EDAT': '2002/01/30 10:00', 'IP': '2', 'IS': '0022-1295', 'JID': '2985110R', 'LA': 'eng', 'MHDA': '2002/09/27 06:00', 'PG': '117-22', 'PMID': '11815661', 'PST': 'ppublish', 'PT': 'Congresses', 'SB': 'IM', 'SO': 'J Gen Physiol 2002 Feb;119(2):117-22.', 'TA': 'J Gen Physiol', 'TI': 'The 55th Annual Meeting and Symposium of the Society of General Physiologists: molecular motors.', 'UI': '21674145', 'VI': '119', 'MH': ['Animal', 'Human', 'Molecular Motors/*physiology', 'Muscle Contraction/physiology', 'Physiology/*methods'], 'RN': ['0 (Molecular Motors)'], 'FAU': ['Andersen, Olaf S']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '53568', '_score': 8.257235, '_source': {'AD': 'Howard Hughes Medical Institute, Department of Cellular and Molecular Pharmacology, Department of Biochemistry and Biophysics, University of California-San Francisco, San Francisco, California, USA.', 'CY': 'United States', 'DA': '20020228', 'DCOM': '20020820', 'DP': '2001', 'EDAT': '2002/03/01 10:00', 'IS': '0065-3233', 'JID': '0116732', 'LA': 'eng', 'MHDA': '2002/08/21 10:01', 'PG': '345-90', 'PMID': '11868277', 'PST': 'ppublish', 'PT': 'Review, Academic', 'RF': '197', 'SB': 'IM', 'SO': 'Adv Protein Chem 2001;59:345-90.', 'TA': 'Adv Protein Chem', 'TI': 'Functional genomic approaches to understanding molecular chaperones and stress responses.', 'UI': '21857850', 'VI': '59', 'MH': ['*Genome', 'Molecular Chaperones/*physiology', '*Oxidative Stress', \"Support, Non-U.S. Gov't\"], 'RN': ['0 (Molecular Chaperones)'], 'FAU': ['Travers, K J', 'Patil, C K', 'Weissman, J S']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '85834', '_score': 8.257235, '_source': {'AB': 'We consider a particle diffusing in a one-dimensional periodic lattice with arbitrary transition rates between nearest-neighbor sites. We show rigorously that the ratio of the drift velocity V to the diffusion coefficient D has the upper bound 2N/d, where N is the number of nodes in an elementary cell and d denotes its length. Applying this result to a model of a molecular motor introduced by Fisher and Kolomeisky [Proc. Natl. Acad. Sci. USA 96, 6597 (1999)] we show that the so called Einstein force, which sets the lower bound for the force exerted by a molecular motor, is bounded from above by 2k(B)TN/d irrespective of the actual values of the jump rates between internal states of the motor.', 'AD': 'Institute of Theoretical Physics, University of Wroclaw, plac Maxa Borna 9, 50204 Wroclaw, Poland. zkoza@ift.uni.wroc.pl', 'CY': 'United States', 'DA': '20020322', 'DCOM': '20020610', 'DP': '2002 Mar', 'EDAT': '2002/03/23 10:00', 'IP': '3 Pt 1', 'IS': '1539-3755', 'JID': '101136452', 'LA': 'eng', 'LR': '20030103', 'MHDA': '2002/06/11 10:01', 'PG': '031905', 'PHST': '2002/Feb/12 [aheadofprint]', 'PMID': '11909107', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Phys Rev E Stat Nonlin Soft Matter Phys 2002 Mar;65(3 Pt 1):031905.', 'TA': 'Phys Rev E Stat Nonlin Soft Matter Phys', 'TI': 'Maximal force exerted by a molecular motor.', 'UI': '21907427', 'VI': '65', 'MH': ['*Biophysics', 'Models, Theoretical', 'Molecular Motors/*chemistry/*physiology', \"Support, Non-U.S. Gov't\"], 'RN': ['0 (Molecular Motors)'], 'FAU': ['Koza, Zbigniew']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '93266', '_score': 8.257235, '_source': {'AB': 'Animal', 'AD': 'Department of Biology, 208 Mueller Laboratory, Pennsylvania State University, University Park, PA 16802, USA. jhm10@psu.edu', 'AID': '022052899 [pii]', 'CY': 'United States', 'DA': '20020403', 'DCOM': '20020425', 'DP': '2002 Apr 2', 'EDAT': '2002/03/28 10:00', 'IP': '7', 'IS': '0027-8424', 'JID': '7505876', 'LA': 'eng', 'LR': '20030103', 'MHDA': '2002/04/26 10:01', 'PG': '4161-6', 'PHST': '2002/Mar/26 [aheadofprint]', 'PMID': '11917097', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'Proc Natl Acad Sci U S A 2002 Apr 2;99(7):4161-6.', 'TA': 'Proc Natl Acad Sci U S A', 'TI': 'Molecules, muscles, and machines: universal performance characteristics of motors.', 'UI': '21927582', 'VI': '99', 'MH': ['Animal', 'Human', '*Molecular Motors', 'Movement', 'Muscles/*physiology', \"Support, U.S. Gov't, Non-P.H.S.\"], 'RN': ['0 (Molecular Motors)'], 'FAU': ['Marden, James H', 'Allen, Lee R']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '165464', '_score': 8.257235, '_source': {'AB': 'Lineal structures in biological tissue support a wide variety of physiological functions, including membrane stabilization, vascular perfusion, and cell-to-cell communication. In 1953, Smith and Guttman demonstrated a stereological method to estimate the total length density (Lv) of linear objects based on random intersections with a two-dimensional sampling probe. Several methods have been developed to ensure the required isotropy of object-probe intersections, including isotropic-uniform-random (IUR) sections, vertical-uniform-random (VUR) slices, and isotropic virtual planes. The disadvantages of these methods are the requirements for inconvenient section orientations (IUR, VUR) or complex counting rules at multiple focal planes (isotropic virtual planes). To overcome these limitations we report a convenient and straightforward approach to estimate Lv and total length, L, for linear objects on tissue sections cut at any arbitrary orientation. The approach presented here uses spherical probes that are inherently isotropic, combined with unbiased fractionator sampling, to demonstrate total L estimation for thin nerve fibres in dorsal hippocampus of the mouse brain.', 'AD': 'Laboratory of Neurosciences, Gerontology Research Center (NIA/NIH) Baltimore, MD, U.S.A. School of Materials Science and Engineering, Georgia Institute of Technology, Atlanta, GA 30332-0245, USA. disector@friend.ly.net', 'AID': '1006 [pii]', 'CY': 'England', 'DA': '20020509', 'DCOM': '20020627', 'DP': '2002 Apr', 'EDAT': '2002/05/10 10:00', 'ID': 'NS16735/NS/NINDS', 'IP': 'Pt 1', 'IS': '0022-2720', 'JID': '0204522', 'LA': 'eng', 'MHDA': '2002/06/28 10:01', 'PG': '54-64', 'PMID': '12000563', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'J Microsc 2002 Apr;206(Pt 1):54-64.', 'TA': 'J Microsc', 'TI': 'Stereological length estimation using spherical probes.', 'UI': '21996286', 'VI': '206', 'MH': ['Animal', 'Anisotropy', 'Biometry/methods', 'Brain/anatomy & histology', 'Dentate Gyrus/anatomy & histology', 'Hippocampus/anatomy & histology', 'Male', 'Mice', 'Mice, Inbred Strains', 'Microspheres', 'Molecular Probes/*metabolism', 'Nerve Fibers', \"Support, Non-U.S. Gov't\", \"Support, U.S. Gov't, Non-P.H.S.\", \"Support, U.S. Gov't, P.H.S.\"], 'RN': ['0 (Molecular Probes)'], 'FAU': ['Mouton, P R', 'Gokhale, A M', 'Ward, N L', 'West, M J']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '174737', '_score': 8.257235, '_source': {'AB': 'The contributions of key molecules predicted to align chromosomes at the center of the mitotic spindle have been recently examined. New results dictate that models for how chromosomes align during the early stages of mitosis must be revised to integrate properties of microtubule-based motor proteins as well as microtubule dynamics.', 'AD': 'Laboratory of Chemistry and Cell Biology, The Rockefeller University, New York, NY 10021, USA. Kapoor@rockefeller.edu', 'AID': 'jcb.200202073 [pii]', 'CY': 'United States', 'DA': '20020515', 'DCOM': '20020617', 'DP': '2002 May 13', 'EDAT': '2002/05/16 10:00', 'ID': 'GM51542/GM/NIGMS', 'IP': '4', 'IS': '0021-9525', 'JID': '0375356', 'LA': 'eng', 'LR': '20030103', 'MHDA': '2002/06/18 10:01', 'PG': '551-6', 'PHST': '2002/May/13 [aheadofprint]', 'PMID': '12011106', 'PST': 'ppublish', 'PT': 'Review, Tutorial', 'RF': '42', 'SB': 'IM', 'SO': 'J Cell Biol 2002 May 13;157(4):551-6.', 'TA': 'J Cell Biol', 'TI': 'Searching for the middle ground: mechanisms of chromosome alignment during mitosis.', 'UI': '22007219', 'VI': '157', 'MH': ['Animal', 'Biomechanics', 'Chromosome Pairing/*genetics', 'Human', 'Kinetochores/*physiology', 'Microtubules/*genetics', 'Mitosis/*genetics', 'Mitotic Spindle Apparatus/*genetics', 'Models, Biological', 'Molecular Motors/*genetics', \"Support, U.S. Gov't, P.H.S.\"], 'RN': ['0 (Molecular Motors)'], 'FAU': ['Kapoor, Tarun M', 'Compton, Duane A']}}, {'_index': 'genomics-stem', '_type': '_doc', '_id': '184972', '_score': 8.257235, '_source': {'AB': 'The correlation ratchet (CR) has been proposed as a possible mechanism for the operation of biomolecular motors. To test whether a CR mechanism is applicable to any particular motor, it is necessary to evaluate the unloaded velocity and the stall force of the CR mechanism. This paper is concerned with efficient numerical procedures for the determination of these quantities. In particular, the numerical results indicate that the stall force of a CR mechanism is quite limited, despite wide variations of the CR parameters.', 'AD': 'Department of Mathematics, Michigan State University, East Lansing, MI 48824, USAchiu@mth.msu.edu.', 'AID': '10.1007/s002850100129 [doi]', 'CY': 'Germany', 'DA': '20020521', 'DCOM': '20020813', 'DP': '2002 May', 'EDAT': '2002/05/22 10:00', 'IP': '5', 'IS': '0303-6812', 'JID': '7502105', 'LA': 'eng', 'MHDA': '2002/08/14 10:01', 'PG': '479-501', 'PMID': '12021986', 'PST': 'ppublish', 'PT': 'Journal Article', 'SB': 'IM', 'SO': 'J Math Biol 2002 May;44(5):479-501.', 'TA': 'J Math Biol', 'TI': 'Numerical analysis of the flux and force generated by a correlation ratchet mechanism.', 'UI': '22015948', 'VI': '44', 'MH': ['Biomechanics', '*Models, Biological', 'Molecular Motors/*physiology', 'Numerical Analysis, Computer-Assisted', \"Support, U.S. Gov't, Non-P.H.S.\"], 'RN': ['0 (Molecular Motors)'], 'FAU': ['Chiu, Chichia', 'Peskin, Charles S']}}]}}\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "# BEGIN ANSWER\n",
    "\n",
    "body_molecule = {\n",
    "    \"query\": {\n",
    "        \"query_string\": {\n",
    "            \"query\": \"molecule\"\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "body_molecular = {\n",
    "  \"query\": {\n",
    "        \"query_string\": {\n",
    "            \"query\": \"molecular\"\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "response_molecule = es.search(index='genomics-stem', body=body_molecule, track_total_hits=True)\n",
    "response_molecular = es.search(index='genomics-stem', body=body_molecular, track_total_hits=True)\n",
    "\n",
    "print(response_molecule)\n",
    "print(response_molecular)\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comment here about the eventual different results you get \n",
    "# -> words that are stemmed, like molecule and molecules, should improve the retrieval results (in terms of amount of retrieved dos)\n",
    "# -> words that are not stemmed, like 'molecular' should not see much differernt results\n",
    "\n",
    "# the point of the exercise is not to have the same results with molecule and molecular, but \n",
    "# what the stemmer does, and reason after that.\n",
    "\n",
    "# It looks like the result of retriving molecule has increased. This is due the fact the for example molecule and molecules have the same root word.\n",
    "# As for molecular, it does not have a root that matches molecule. So the retrived number does not change. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------------------------------------------------------------------\n",
    "--------------------------------------------------------------------------\n",
    "--------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 04: Search models \n",
    "\n",
    "\n",
    "<span style=\"background:red; color: white;\">__You are advised to work on this part after Lecture 03__</span>\n",
    "\n",
    "\n",
    "### Background\n",
    "The way documents are indexed influences the performance of the IR systems. \n",
    "Elasticsearch [Mappings](https://www.elastic.co/guide/en/elasticsearch/reference/7.8/mapping.html) define how a document, and its properties (fields) are stored and indexed, but also provides tools to implement and execute different document similarity measures (i.e. search models).  When using a different configuration of an ElasticSearch Mapping, the document collection needs to be re-indexed (or a new index need to be created - use the functions we provided above to do that).\n",
    "\n",
    "> See again: [Index Settings and Mappings](https://www.elastic.co/guide/en/elasticsearch/reference/7.8/indices-create-index.html).\n",
    "> _Note: the default model (similarity) in ElasticSearch is BM25. Different models need to be specified (see example)._\n",
    "\n",
    "For instance, we can add a new field `\"title-abstract\"` that uses the  [similarity measure](https://www.elastic.co/guide/en/elasticsearch/reference/7.8/similarity.html) _Boolean_, and let it serve as an index for the fields `\"TI\"` and `\"AB\"` (title and abstract):\n",
    "\n",
    "> Plase note that if you want to use the `boolean` similarity for the single fields, you need to specify it for each field. Otherwise, the default BM25 will be used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "import elasticsearch\n",
    "import elasticsearch.helpers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(263080, [])"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "boolean = {\n",
    "  \"settings\" : {\n",
    "    # a single shard, so we do not suffer from approximate document frequencies\n",
    "    \"number_of_shards\" : 1\n",
    "  },\n",
    "  \"mappings\": {\n",
    "      \"properties\": {\n",
    "        \"AB\": {\n",
    "          \"type\": \"text\",\n",
    "          \"copy_to\": \"title-abstract\",\n",
    "          \"similarity\": \"boolean\"\n",
    "        },\n",
    "        \"TI\": {\n",
    "          \"type\": \"text\",\n",
    "          \"copy_to\": \"title-abstract\",\n",
    "          \"similarity\": \"boolean\"\n",
    "        },\n",
    "        \"title-abstract\": {  # compound field\n",
    "          \"type\": \"text\",\n",
    "          \"similarity\": \"boolean\"\n",
    "        }\n",
    "      }\n",
    "  }\n",
    "}\n",
    "\n",
    "es = elasticsearch.Elasticsearch('localhost')\n",
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-bool', body=boolean)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Most changes to the mappings cannot be done on an existing index. Some (for instance\n",
    "> similarity measures) can be changed if the index is first closed. Nevertheless, we \n",
    "> will in this notebook _re-index_ the collection for every change to the mappings\n",
    "> using the function `index_documents()` that we defined above. Mappings (and settings)\n",
    "> can be passed to the function using the `body` parameter.\n",
    "\n",
    "<span style=\"background:#444; color: white;\">__We suggest you to create different indices using different models of search (according to the available disk space on your VM). This will avoid that changes are not correctly applied, and you won't see the expected results.__</span>\n",
    "\n",
    "<span style=\"background:#444; color: white;\">E.g. for the 'boolean' model, we created the 'genomics-bool' index.</span>\n",
    "\n",
    "Let's have a look at the mappings and settings for our index as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'genomics-bool': {'aliases': {},\n",
       "  'mappings': {'properties': {'AB': {'type': 'text',\n",
       "     'copy_to': ['title-abstract'],\n",
       "     'similarity': 'boolean'},\n",
       "    'AD': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'AID': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'CI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'CIN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'CN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'CON': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'CY': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'DA': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'DCOM': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'DP': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'EDAT': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'EIN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'FAU': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'ID': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'IP': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'IS': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'JID': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'LA': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'LR': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'MH': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'MHDA': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'ORI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PG': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PHST': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PMID': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PS': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PST': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PT': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'RF': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'RIN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'RN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'ROF': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'RPF': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'RPI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'SB': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'SI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'SO': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'SPIN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'TA': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'TI': {'type': 'text',\n",
       "     'copy_to': ['title-abstract'],\n",
       "     'similarity': 'boolean'},\n",
       "    'TT': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'UI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'UIN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'UOF': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'VI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'title-abstract': {'type': 'text', 'similarity': 'boolean'}}},\n",
       "  'settings': {'index': {'routing': {'allocation': {'include': {'_tier_preference': 'data_content'}}},\n",
       "    'number_of_shards': '1',\n",
       "    'provided_name': 'genomics-bool',\n",
       "    'creation_date': '1696168468548',\n",
       "    'number_of_replicas': '1',\n",
       "    'uuid': '1J6zvaSRRMm9kaZdm4Zq-g',\n",
       "    'version': {'created': '7171399'}}}}}"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "es.indices.get(index='genomics-bool')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's search our new field `\"title-abstract\"` as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'took': 7369,\n",
       " 'timed_out': False,\n",
       " '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0},\n",
       " 'hits': {'total': {'value': 3185, 'relation': 'eq'},\n",
       "  'max_score': 1.0,\n",
       "  'hits': [{'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '125',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'Previous studies from our laboratories revealed the susceptibility of Leishmania sp. to glibenclamide (GLIB), a potassium channel blocker which selectively interacts with adenosine-binding-cassette transporters. In the present work, we analyzed whether the drug sensitivity of intracellular amastigotes correlates with changes in macrophage features that are related to their function as antigen-presenting cells. We provide evidence that in BALB/c murine macrophages, GLIB induced a decrease in the interferon-gamma-stimulated expression of major histocompatibility complex class II molecules and the co-stimulatory molecule CD86 (B7-2). Furthermore, it caused a decrease in the interleukin-1 secretion by macrophages. The data indicate that the treatment with GLIB inhibits the Th2 development and polarizes macrophage functions towards the induction of a protective Th1 response.',\n",
       "     'AD': 'Laboratory of Molecular Physiology, Instituto de Medicina Experimental, Facultad de Medicina, Universidad Central de Venezuela, Caracas. aiponte@reacciun.ve',\n",
       "     'CY': 'United States',\n",
       "     'DA': '20011105',\n",
       "     'DCOM': '20020412',\n",
       "     'DP': '2001 Aug',\n",
       "     'EDAT': '2001/11/06 10:00',\n",
       "     'IP': '3',\n",
       "     'IS': '0892-3973',\n",
       "     'JID': '8800150',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/04/16 10:01',\n",
       "     'PG': '477-86',\n",
       "     'PMID': '11694036',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Immunopharmacol Immunotoxicol 2001 Aug;23(3):477-86.',\n",
       "     'TA': 'Immunopharmacol Immunotoxicol',\n",
       "     'TI': 'Experimental leishmaniasis: the glibenclamide-triggered decrease in parasite growth correlates with changes in macrophage features.',\n",
       "     'UI': '21545398',\n",
       "     'VI': '23',\n",
       "     'MH': ['Animal',\n",
       "      'Antigens, CD/metabolism',\n",
       "      'Glyburide/*therapeutic use',\n",
       "      'Histocompatibility Antigens Class II/metabolism',\n",
       "      'In Vitro',\n",
       "      'Interleukin-1/metabolism',\n",
       "      'Leishmania major/drug effects/growth & development',\n",
       "      'Leishmaniasis/*drug therapy/*immunology/parasitology',\n",
       "      'Macrophages/*drug effects/*immunology/parasitology',\n",
       "      'Membrane Glycoproteins/metabolism',\n",
       "      'Mice',\n",
       "      'Mice, Inbred BALB C',\n",
       "      \"Support, Non-U.S. Gov't\"],\n",
       "     'RN': ['0 (Antigens, CD)',\n",
       "      '0 (B7-2 protein)',\n",
       "      '0 (Histocompatibility Antigens Class II)',\n",
       "      '0 (Interleukin-1)',\n",
       "      '0 (Membrane Glycoproteins)',\n",
       "      '10238-21-8 (Glyburide)'],\n",
       "     'FAU': ['Ponte-Sucre, A', 'Figarella, K', 'Moll, H']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '639',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'Osteoblast differentiation is a multistep process that involves critical spatial and temporal regulation of cellular processes marked by the presence of a large number of differentially expressed molecules. To identify key functional molecules, we used differential messenger RNA (mRNA) display and compared RNA populations isolated from the defined transition phases (proliferation, matrix formation, and mineralization) of the MC3T3-E1 osteoblast-like cell line. Using this approach, a complementary DNA (cDNA) fragment was isolated and identified as neuroleukin (NLK), a multifunctional cytokine also known as autocrine motility factor (AMF), phosphoglucose isomerase (PGI; phosphohexose isomerase [PHI]), and maturation factor (MF). Northern analysis showed NLK temporal expression during MC3T3-E1 cell differentiation with a 3.5-fold increase during matrix formation and mineralization. Immunocytochemical studies revealed the presence of NLK in MC3T3-E1 cells as well as in the surrounding matrix, consistent with a secreted molecule. In contrast, the NLK receptor protein was detected primarily on the cell membrane. In subsequent studies, a high level of NLK expression was identified in osteoblasts and superficial articular chondrocytes in bone of 1-, 4-, and 8-month-old normal mice, as well as in fibroblasts, proliferating chondrocytes, and osteoblasts within a fracture callus. However, NLK was not evident in hypertrophic chondrocytes or osteocytes. In addition, treatment of MC3T3 cells with 6-phosphogluconic acid (6PGA; a NLK inhibitor) resulted in diminishing alkaline phosphatase (ALP) activity and mineralization in MC3T3-E1 cells, especially during the matrix formation stage of differentiating cells. Taken together, these data show specific expression of NLK in discrete populations of bone and cartilage cells and suggest a possible role for this secreted protein in bone development and regeneration.',\n",
       "     'AD': 'Department of Biomedical Engineering, State University of New York at Stony Brook, 11794-2580, USA.',\n",
       "     'CY': 'United States',\n",
       "     'DA': '20011107',\n",
       "     'DCOM': '20020422',\n",
       "     'DP': '2001 Nov',\n",
       "     'EDAT': '2001/11/08 10:00',\n",
       "     'IP': '11',\n",
       "     'IS': '0884-0431',\n",
       "     'JID': '8610640',\n",
       "     'LA': 'eng',\n",
       "     'LR': '20020610',\n",
       "     'MHDA': '2002/04/23 10:01',\n",
       "     'PG': '1994-2004',\n",
       "     'PMID': '11697795',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'S',\n",
       "     'SO': 'J Bone Miner Res 2001 Nov;16(11):1994-2004.',\n",
       "     'TA': 'J Bone Miner Res',\n",
       "     'TI': 'Differential expression of neuroleukin in osseous tissues and its involvement in mineralization during osteoblast differentiation.',\n",
       "     'UI': '21554149',\n",
       "     'VI': '16',\n",
       "     'MH': ['3T3 Cells',\n",
       "      'Animal',\n",
       "      'Bone Development/genetics/physiology',\n",
       "      'Cell Differentiation/drug effects',\n",
       "      'Fracture Healing/genetics/physiology',\n",
       "      'Gene Expression',\n",
       "      'Gluconates/pharmacology',\n",
       "      'Immunohistochemistry',\n",
       "      'Mice',\n",
       "      'Minerals/metabolism',\n",
       "      'Neuroleukin/antagonists & inhibitors/*genetics',\n",
       "      'Osteoblasts/*cytology/drug effects/*metabolism',\n",
       "      'RNA, Messenger/genetics/metabolism',\n",
       "      'Receptors, Growth Factor/metabolism',\n",
       "      \"Support, Non-U.S. Gov't\",\n",
       "      \"Support, U.S. Gov't, Non-P.H.S.\"],\n",
       "     'RN': ['0 (Gluconates)',\n",
       "      '0 (Minerals)',\n",
       "      '0 (Neuroleukin)',\n",
       "      '0 (RNA, Messenger)',\n",
       "      '0 (Receptors, Growth Factor)',\n",
       "      '921-62-0 (6-phosphogluconate)'],\n",
       "     'FAU': ['Zhi, J', 'Sommerfeldt, D W', 'Rubin, C T', 'Hadjiargyrou, M']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '713',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'CD38 identifies a surface molecule with multi-functional activity. Its prognostic importance in B-cell chronic lymphocytic leukemia (B-CLL) is currently under investigation in view of the fact that two different groups have recently indicated that CD38 expression could be an independent prognostic marker in B-CLL. We analyzed the clinico-biological features of 61 immunologically typical (CD5+CD23+) B-CLL patients stratified according to the CD38 expression. Twenty-two (36%) patients expressed CD38 in more than 30% of CD19-positive cells and were considered as CD38-positive B-CLL. Atypical morphology (p 0.02), peripheral blood lymphocytosis (p 0.01) and diffuse histopathologic bone marrow pattern (p 0.003) were findings found to be closely associated with CD38 expression. On the other hand, A and B Binet stages (p 0.02) and interstitial bone marrow involvement (p 0.005) were more represented in the CD38-negative B-CLL group. Trisomy 12 was detected more frequently in the CD38-positive B-CLL group, while 13q14 deletions mainly occurred in CD38-negative group (p 0.005). Finally, median survival of CD38-positive B-CLL patients was 90 months, while it was not reached at 180 months in CD38-negative patients. Taken together, our data strongly suggest that the evaluation of CD38 expression may identify two groups patients with B-CLL greatly differing in their clinico-biological features.',\n",
       "     'AD': 'Division of Hematology, IRCCS Casa Sollievo della Sofferenza Hospital, 71013 San Giovanni Rotondo, Italy. ematologia@operapadrepio.it',\n",
       "     'CY': 'Switzerland',\n",
       "     'DA': '20011108',\n",
       "     'DCOM': '20020516',\n",
       "     'DP': '2001 Jun',\n",
       "     'EDAT': '2001/11/09 10:00',\n",
       "     'IP': '1-2',\n",
       "     'IS': '1042-8194',\n",
       "     'JID': '9007422',\n",
       "     'LA': 'eng',\n",
       "     'LR': '20021213',\n",
       "     'MHDA': '2002/05/17 10:01',\n",
       "     'PG': '109-14',\n",
       "     'PMID': '11699197',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Leuk Lymphoma 2001 Jun;42(1-2):109-14.',\n",
       "     'TA': 'Leuk Lymphoma',\n",
       "     'TI': 'CD38 expression correlates with adverse biological features and predicts poor clinical outcome in B-cell chronic lymphocytic leukemia.',\n",
       "     'UI': '21556526',\n",
       "     'VI': '42',\n",
       "     'MH': ['ADP-ribosyl Cyclase',\n",
       "      'Adult',\n",
       "      'Aged',\n",
       "      'Antigens, Differentiation/*biosynthesis',\n",
       "      'Antigens, Neoplasm/*biosynthesis',\n",
       "      'Biological Markers/analysis',\n",
       "      'Cohort Studies',\n",
       "      'Cytogenetic Analysis',\n",
       "      'Human',\n",
       "      'Leukemia, B-Cell, Chronic/diagnosis/*immunology/mortality',\n",
       "      'Middle Age',\n",
       "      'NAD+ Nucleosidase/*biosynthesis',\n",
       "      'Prognosis',\n",
       "      'Survival Analysis',\n",
       "      'Treatment Outcome'],\n",
       "     'RN': ['0 (Antigens, Differentiation)',\n",
       "      '0 (Antigens, Neoplasm)',\n",
       "      '0 (Biological Markers)',\n",
       "      'EC 3.2.2.5 (ADP-ribosyl Cyclase)',\n",
       "      'EC 3.2.2.5 (CD38 antigen)',\n",
       "      'EC 3.2.2.5 (NAD+ Nucleosidase)'],\n",
       "     'FAU': [\"D'Arena, G\",\n",
       "      'Musto, P',\n",
       "      'Cascavilla, N',\n",
       "      \"Dell'Olio, M\",\n",
       "      'Di Renzo, N',\n",
       "      'Perla, G',\n",
       "      'Savino, L',\n",
       "      'Carotenuto, M']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '737',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'The emigration of lymphocytes from blood into lymph nodes is regulated by the expression of the adhesion molecule, L-selectin on the lymphocyte surface which arrests the rolling of the cell on the vessel wall and allows firmer adhesive interactions to develop. The expression of L-selectin on B-CLL lymphocytes is less than half that on normal lymphocytes and this difference correlates with an impaired capacity of B-CLL lymphocytes to migrate beneath a monolayer of human umbilical vein endothelial cells. Both the B-cell and T-cell lymphocytes from normal subjects and B-CLL patients show down-regulation of L-selectin and CD23 after transendothelial migration. The reduced expression of L-selectin on B-CLL lymphocytes leads to a relative \"trapping\" of these cells in the vascular space and is one factor contributing to the elevation of peripheral lymphocyte count.',\n",
       "     'AD': 'Sydney University, Department of Medicine, Nepean Hospital, Penrith, NSW 2750, Australia.',\n",
       "     'CY': 'Switzerland',\n",
       "     'DA': '20011108',\n",
       "     'DCOM': '20020516',\n",
       "     'DP': '2001 Jun',\n",
       "     'EDAT': '2001/11/09 10:00',\n",
       "     'IP': '1-2',\n",
       "     'IS': '1042-8194',\n",
       "     'JID': '9007422',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/05/17 10:01',\n",
       "     'PG': '5-12',\n",
       "     'PMID': '11699221',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Review, Tutorial',\n",
       "     'RF': '34',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Leuk Lymphoma 2001 Jun;42(1-2):5-12.',\n",
       "     'TA': 'Leuk Lymphoma',\n",
       "     'TI': 'Impaired transendothelial migration of B-CLL lymphocytes: a defect linked to low L-selectin expression.',\n",
       "     'UI': '21556515',\n",
       "     'VI': '42',\n",
       "     'MH': ['B-Lymphocytes/pathology',\n",
       "      'Chemotaxis, Leukocyte',\n",
       "      'Down-Regulation',\n",
       "      'Endothelium, Vascular/*physiopathology',\n",
       "      'Human',\n",
       "      'L-Selectin/biosynthesis/physiology',\n",
       "      'Leukemia, B-Cell, Chronic/*pathology/physiopathology'],\n",
       "     'RN': ['126880-86-2 (L-Selectin)'],\n",
       "     'FAU': ['Gu, B', 'Dao, L P', 'Wiley, J']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '858',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'Monolayers of seven fractions of natural lipids (phosphatidyl inositol, sulfoquinovosyl diacylglycerol, phosphatidylcholine, digalactosyl diacylglycerol, phosphatidyl glycerol, phosphatidylethanolamine, monogalactosyl diacylglycerol), isolated from the photoautotrophic cell culture of the moss Marchantia polymorpha grown under normal and light-stress conditions, have been prepared for the first time. We have shown that the high-intensity light affects the area occupied by the lipid molecule. In the case of digalactosyl diacylglycerol and phosphatidyl glycerol fractions, after the light stress the area significantly increased from 0.50 to 0.80 nm2 and from 0.47 to 0.63 nm2, respectively, and in the case of the sulfoquinovosyl diacylglycerol fraction, the area decreased from 0.40 to 0.32 nm2. These results are in agreement with our previous data on the redistribution of the double bonds in the aliphatic chains of these lipids and can be used to characterize the state of the lipid bilayer of the thylakoid membranes.',\n",
       "     'AD': 'Shemyakin and Ovchinnikov Institute of Bioorganic Chemistry, Russian Academy of Sciences, Moscow.',\n",
       "     'CY': 'Switzerland',\n",
       "     'DA': '20011108',\n",
       "     'DCOM': '20020411',\n",
       "     'DP': '2001 Jul',\n",
       "     'EDAT': '2001/11/09 10:00',\n",
       "     'IP': '5',\n",
       "     'IS': '1023-6597',\n",
       "     'JID': '9517472',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/04/12 10:01',\n",
       "     'PG': '579-85',\n",
       "     'PMID': '11699862',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Membr Cell Biol 2001 Jul;14(5):579-85.',\n",
       "     'TA': 'Membr Cell Biol',\n",
       "     'TI': 'A study of moss (Marchantia polymorpha) thylakoid membrane lipids in monolayers.',\n",
       "     'UI': '21556221',\n",
       "     'VI': '14',\n",
       "     'MH': ['Cells, Cultured/chemistry/radiation effects',\n",
       "      'Chromatography, Thin Layer',\n",
       "      'Diglycerides/chemistry/isolation & purification',\n",
       "      'Glycolipids/chemistry/isolation & purification',\n",
       "      'Light',\n",
       "      'Membrane Lipids/*chemistry/isolation & purification',\n",
       "      'Mosses/chemistry/radiation effects',\n",
       "      'Phosphatidylcholines/chemistry/isolation & purification',\n",
       "      'Phosphatidylethanolamines/chemistry/isolation & purification',\n",
       "      'Phosphatidylglycerols/chemistry/isolation & purification',\n",
       "      'Phosphatidylinositols/chemistry/isolation & purification',\n",
       "      'Plants/*chemistry/radiation effects',\n",
       "      \"Support, Non-U.S. Gov't\",\n",
       "      'Thylakoids/*chemistry/radiation effects'],\n",
       "     'RN': ['0 (Diglycerides)',\n",
       "      '0 (Glycolipids)',\n",
       "      '0 (Membrane Lipids)',\n",
       "      '0 (Phosphatidylcholines)',\n",
       "      '0 (Phosphatidylethanolamines)',\n",
       "      '0 (Phosphatidylglycerols)',\n",
       "      '0 (Phosphatidylinositols)',\n",
       "      '0 (digalactosyldiglycerides)',\n",
       "      '0 (monogalactosyldiglycerides)',\n",
       "      '0 (sulfoquinovosyl diglyceride)'],\n",
       "     'FAU': ['Volchenkova, T A',\n",
       "      'Kalabina, N A',\n",
       "      'Schaefer, C',\n",
       "      'Zubov, V P']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '1206',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'Photodynamic therapy (PDT) is the treatment of tumors or dysplasic tissue with drugs that produce cytotoxic metabolites when exposed to light. Aminolevulinic acid HCl (5-aminolevulinic acid HCl; ALA) is a prodrug that is metabolized intracellularly to form the photosensitizing molecule protoporphyrin (PpIX). When PpIX is activated by light, cytotoxic reactive oxygen species and free radicals are generated. ALA can diffuse through skin and preferentially localizes in tumors and dysplasic tissue; subsequent exposure of PpIX-loaded tumor cells to light can destroy the tumor. After application of a 20% solution of ALA to actinic keratosis lesions of the head, PpIX (as measured by skin fluorescence) peaked 11 hours after treatment and the mean clearance half-life was 30 hours. In phase II trials 10 J/cm2 of blue light (wavelength = 417 nm) delivered at 10 mW/cm2 for 1000 seconds was found to provide maximal therapeutic effect on lesions of the head after treatment with 20% ALA. In phase III trials of ALA PDT in 241 patients with lesions of the head 72% of patients had a complete response to treatment at 12 weeks versus 20% of those treated with vehicle and light alone. Some of these patients had been re-treated at 8 weeks. In these trials 12% of ALA-treated patients and 37.5% of those receiving vehicle whose lesions had cleared at 8 weeks had relapsed at 12 weeks. When the total number of lesions were considered the recurrence rate was 5 and 27.9% for ALA',\n",
       "     'AD': 'Adis International Limited, Auckland, New Zealand. demail@adis.co.nz',\n",
       "     'CY': 'New Zealand',\n",
       "     'DA': '20011112',\n",
       "     'DCOM': '20020716',\n",
       "     'DP': '2000 Mar-Apr',\n",
       "     'EDAT': '2001/11/13 10:00',\n",
       "     'IP': '2',\n",
       "     'IS': '1175-0561',\n",
       "     'JID': '100895290',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/07/18 10:01',\n",
       "     'PG': '133-9; discussion 140-1',\n",
       "     'PMID': '11702313',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Review, Tutorial',\n",
       "     'RF': '42',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Am J Clin Dermatol 2000 Mar-Apr;1(2):133-9; discussion 140-1.',\n",
       "     'TA': 'Am J Clin Dermatol',\n",
       "     'TI': 'Topical aminolevulinic acid HCl photodynamic therapy.',\n",
       "     'UI': '21559338',\n",
       "     'VI': '1',\n",
       "     'MH': ['Administration, Topical',\n",
       "      'Aminolevulinic Acid/*administration & dosage/adverse effects',\n",
       "      'Clinical Trials, Phase II',\n",
       "      'Clinical Trials, Phase III',\n",
       "      'Human',\n",
       "      '*Photochemotherapy',\n",
       "      'Photosensitizing Agents/*administration & dosage/adverse effects'],\n",
       "     'RN': ['0 (Photosensitizing Agents)', '106-60-5 (Aminolevulinic Acid)'],\n",
       "     'FAU': ['Ormrod, D', 'Jarvis, B']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '1684',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'We describe an approach for the rapid mapping of epitopes within a malaria antigen using a combination of phage display techniques. Phage display of antigen fragments identifies the location of the epitopes, then random peptide libraries displayed on phage are employed to identify accurately amino acids involved in the epitope. Finally, phage display of mutant fragments confirms the role of each residue in the epitope. This approach was applied to the apical membrane antigen-1 (AMA1), which is a leading candidate for inclusion in a vaccine directed against the asexual blood stages of Plasmodium falciparum. As part of the effort both to understand the function of AMA1 in the parasite life cycle and to define the specificity of protective immune responses, a panel of monoclonal antibodies (MAbs) was generated to obtain binding reagents to the various domains within the molecule. There is a pressing need to determine rapidly the regions recognized by these antibodies and the structural requirements required within AMA1 for high affinity binding of the MAbs. Using phage displaying random AMA1 fragments, it was shown that MAb5G8 recognizes a short linear epitope within the pro-domain of AMA1 whereas the epitope recognized by MAb 1F9 is reduction sensitive and resides within a disulphide-bonded 57 amino acid sub-domain of domain-1. Phage displaying random peptide libraries and mutant AMA1 fragments were employed for fine mapping of the MAb5G8 core epitope to a three-residue sequence in the AMA1 prodomain.',\n",
       "     'AD': 'Department of Biochemistry, La Trobe University, Bundoora, 3083, Victoria, Australia.',\n",
       "     'CY': 'England',\n",
       "     'DA': '20011114',\n",
       "     'DCOM': '20020402',\n",
       "     'DP': '2001 Sep',\n",
       "     'EDAT': '2001/11/15 10:00',\n",
       "     'IP': '9',\n",
       "     'IS': '0269-2139',\n",
       "     'JID': '8801484',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/04/03 10:01',\n",
       "     'PG': '691-8',\n",
       "     'PMID': '11707616',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Protein Eng 2001 Sep;14(9):691-8.',\n",
       "     'TA': 'Protein Eng',\n",
       "     'TI': 'Rapid and precise epitope mapping of monoclonal antibodies against Plasmodium falciparum AMA1 by combined phage display of fragments and random peptides.',\n",
       "     'UI': '21564501',\n",
       "     'VI': '14',\n",
       "     'MH': ['Amino Acid Sequence',\n",
       "      'Animal',\n",
       "      'Antibodies, Monoclonal/genetics/*immunology',\n",
       "      'Antibody Specificity',\n",
       "      'Antigens, Protozoan/chemistry/genetics/*immunology',\n",
       "      'Bacteriophages/genetics',\n",
       "      'Disulfides/chemistry',\n",
       "      'Epitope Mapping/*methods',\n",
       "      'Epitopes/immunology',\n",
       "      'Erythrocytes/parasitology',\n",
       "      'Membrane Proteins/chemistry/*immunology',\n",
       "      'Molecular Sequence Data',\n",
       "      'Mutation',\n",
       "      'Peptide Library',\n",
       "      'Peptides/genetics/immunology',\n",
       "      'Plasmodium falciparum/*immunology',\n",
       "      'Protein Binding',\n",
       "      'Protein Structure, Tertiary',\n",
       "      'Protozoan Proteins/chemistry/*immunology',\n",
       "      'Sequence Alignment',\n",
       "      \"Support, Non-U.S. Gov't\"],\n",
       "     'RN': ['0 (Antibodies, Monoclonal)',\n",
       "      '0 (Antigens, Protozoan)',\n",
       "      '0 (Disulfides)',\n",
       "      '0 (Epitopes)',\n",
       "      '0 (Membrane Proteins)',\n",
       "      '0 (Peptide Library)',\n",
       "      '0 (Peptides)',\n",
       "      '0 (Protozoan Proteins)',\n",
       "      '0 (apical membrane antigen I)'],\n",
       "     'FAU': ['Coley, A M',\n",
       "      'Campanale, N V',\n",
       "      'Casey, J L',\n",
       "      'Hodder, A N',\n",
       "      'Crewther, P E',\n",
       "      'Anders, R F',\n",
       "      'Tilley, L M',\n",
       "      'Foley, M']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '2029',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'Inflammatory cells are thought to be instrumental in the pathophysiology of pulmonary diseases, and control of their recruitment and activation in the lung would appear to be an attractive strategy for therapeutic intervention. Interleukin-8 and related CXC chemokines are involved in the function of neutrophils and T cells, and have been implicated in several lung diseases. Small-molecule antagonists of the interleukin-8 receptors have been identified, which may help elucidate the role of interleukin-8 and related chemokines in the pathophysiology of lung diseases.',\n",
       "     'AD': 'Respiratory, Inflammation and Respiratory Pathogens Center for Excellence in Drug Discovery, GlaxoSmithKline, King of Prussia, PA 19406, USA. douglas_w_hay@sbphrd.com',\n",
       "     'CY': 'England',\n",
       "     'DA': '20011119',\n",
       "     'DCOM': '20020408',\n",
       "     'DP': '2001 Jun',\n",
       "     'EDAT': '2001/11/20 10:00',\n",
       "     'IP': '3',\n",
       "     'IS': '1471-4892',\n",
       "     'JID': '100966133',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/04/09 10:01',\n",
       "     'PG': '242-7',\n",
       "     'PMID': '11712746',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Review, Tutorial',\n",
       "     'RF': '57',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Curr Opin Pharmacol 2001 Jun;1(3):242-7.',\n",
       "     'TA': 'Curr Opin Pharmacol',\n",
       "     'TI': 'Interleukin-8 receptor antagonists in pulmonary diseases.',\n",
       "     'UI': '21569164',\n",
       "     'VI': '1',\n",
       "     'MH': ['Animal',\n",
       "      'Human',\n",
       "      'Interleukin-8/metabolism/*physiology',\n",
       "      'Lung Diseases/*drug therapy',\n",
       "      'Receptors, Interleukin-8A/*antagonists & inhibitors/physiology'],\n",
       "     'RN': ['0 (Interleukin-8)', '0 (Receptors, Interleukin-8A)'],\n",
       "     'FAU': ['Hay, D W', 'Sarau, H M']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '2046',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'PROBLEM: It has been suggested that specific mechanisms commonly used by different cellular systems to evade immunologic recognition are involved in the development of endometriosis. To gain insight into this aspect, we looked at the relationship between two of these mechanisms in endometrial stroma and the melanoma system for which the ability to create an environment of immune privilege has been well established. METHOD OF STUDY: Media conditioned by endometrial stromal cultures and malignant melanoma A375 were examined to test their effects on peripheral blood mononuclear cell-mediated cytotoxicity directed against K562 target. Moreover, these media were tested for the concentration of the soluble form of intercellular adhesion molecule-1 (sICAM-1), which has been suggested as a marker for spreading potential. RESULTS: Media conditioned by endometrial stromal cultures exerted a significant suppressive effect on cell cytotoxicity when compared with those derived from malignant melanoma Moreover, the constitutive release of sICAM-1 was significantly higher in supernatants from endometrial stromal than in melanoma cells. CONCLUSIONS: These results indicate that two specific properties suggested to be involved in the ability of tumor cells to evade the immune system are more pronounced in the endometrium than in a malignant melanoma. Since the properties evaluated have been previously demonstrated to be even more notable in endometrial samples derived from endometriosis patients, a role of these mechanisms in the development of the disease may be hypothesized.',\n",
       "     'AD': 'Istituto Scientifico Centro Auxologico Italiano, Milano, Italy. paola.vigano@unimi.it',\n",
       "     'CY': 'Denmark',\n",
       "     'DA': '20011119',\n",
       "     'DCOM': '20020404',\n",
       "     'DP': '2001 Nov',\n",
       "     'EDAT': '2001/11/20 10:00',\n",
       "     'IP': '5',\n",
       "     'IS': '1046-7408',\n",
       "     'JID': '8912860',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/04/18 10:01',\n",
       "     'PG': '342-8',\n",
       "     'PMID': '11712763',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Am J Reprod Immunol 2001 Nov;46(5):342-8.',\n",
       "     'TA': 'Am J Reprod Immunol',\n",
       "     'TI': 'Suppression of natural killer cell function and production of soluble ICAM-1: endometrial stroma versus melanoma.',\n",
       "     'UI': '21569181',\n",
       "     'VI': '46',\n",
       "     'MH': ['Comparative Study',\n",
       "      'Cytotoxicity, Immunologic',\n",
       "      'Endometriosis/etiology',\n",
       "      'Endometrium/*immunology/metabolism',\n",
       "      'Female',\n",
       "      'Human',\n",
       "      '*Immune Tolerance',\n",
       "      'Intercellular Adhesion Molecule-1/*biosynthesis',\n",
       "      'Interleukin-1/pharmacology',\n",
       "      'Killer Cells, Natural/*immunology',\n",
       "      'Melanoma/etiology/*immunology/metabolism',\n",
       "      'Stromal Cells/metabolism',\n",
       "      'Tumor Cells, Cultured'],\n",
       "     'RN': ['0 (Interleukin-1)',\n",
       "      '126547-89-5 (Intercellular Adhesion Molecule-1)'],\n",
       "     'FAU': ['Vigano, P',\n",
       "      'Somigliana, E',\n",
       "      'Di Blasio, A M',\n",
       "      'Cozzolino, S',\n",
       "      'Candiani, M',\n",
       "      'Vignali, M']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '2420',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': \"In this article we describe detailed pathological and molecular genetics studies in a consanguineous kindred with Pendred's syndrome. The index patient was a 53-year-old female patient with congenital deafness and goiter. Her parents were first-degree cousins. She had a large goiter (150 g) that had been present since childhood. One of her sisters and a niece are also deaf and have goiter as well. The presence of Pendred's syndrome was confirmed by a positive perchlorate test and the demonstration of a Mondini malformation. Thyroid function tests (under levothyroxine [LT4] therapy) were in the euthyroid range with a thyrotropin [TSH] level of 2.8 microU/mL (0.2-3.2), a serum total thyroxine (T4) of 90 nmol/L (54-142), and a serum total triiodothyronine (T3) of 2.7 nmol/L (0.8-2.4). Total thyroidectomy was performed, and the mass in the right lobe was found to have invaded adjacent tissues. The histopathological findings were consistent with a follicular carcinoma with areas of anaplastic transformation and lung metastasis. The patient was treated twice with 100 mCi 131iodine (3,700 MBq) and received suppressive doses of LT4. Postoperatively, the serum thyroglobulin (Tg) levels remained markedly elevated (2,352 to 41,336 ng/mL). The patient died of a sudden severe episode of hemoptysis. Sequence analysis of the PDS gene performed with DNA from the two relatives with Pendred's syndrome revealed the presence of a deletion of thymidine 279 in exon 3, a point mutation that results in a frameshift and a premature stop codon at codon 96 in the pendrin molecule. We concluded that prolonged TSH stimulation because of iodine deficiency or dyshormonogenesis in combination with mutations of oncogenes and/or tumor suppressor genes, may result in the development of follicular thyroid carcinomas that undergo transformation into anaplastic cancers. It is likely that these pathogenetic mechanisms have been involved in the development of aggressive metastatic thyroid cancer in this unusual patient with Pendred's syndrome.\",\n",
       "     'AD': 'Thyroid Unit, Hospital das Clinicas FMUSP, Sao Paulo, Brazil.',\n",
       "     'CY': 'United States',\n",
       "     'DA': '20011121',\n",
       "     'DCOM': '20020522',\n",
       "     'DP': '2001 Oct',\n",
       "     'EDAT': '2001/11/22 10:00',\n",
       "     'IP': '10',\n",
       "     'IS': '1050-7256',\n",
       "     'JID': '9104317',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/05/25 10:01',\n",
       "     'PG': '981-8',\n",
       "     'PMID': '11716048',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Thyroid 2001 Oct;11(10):981-8.',\n",
       "     'TA': 'Thyroid',\n",
       "     'TI': \"Aggressive metastatic follicular thyroid carcinoma with anaplastic transformation arising from a long-standing goiter in a patient with Pendred's syndrome.\",\n",
       "     'UI': '21572133',\n",
       "     'VI': '11',\n",
       "     'MH': ['Adult',\n",
       "      'Amino Acid Sequence',\n",
       "      'Anaplasia',\n",
       "      'Base Sequence',\n",
       "      'Carrier Proteins/genetics',\n",
       "      'Case Report',\n",
       "      'Consanguinity',\n",
       "      'DNA/genetics',\n",
       "      'DNA Mutational Analysis',\n",
       "      'Deafness/*complications/congenital/*genetics',\n",
       "      'Female',\n",
       "      'Goiter/*complications/*genetics',\n",
       "      'Human',\n",
       "      'Lung Neoplasms/secondary',\n",
       "      'Male',\n",
       "      'Middle Age',\n",
       "      'Mutation',\n",
       "      'Pedigree',\n",
       "      \"Support, Non-U.S. Gov't\",\n",
       "      'Syndrome',\n",
       "      'Thyroid Neoplasms/*complications/genetics/pathology'],\n",
       "     'RN': ['0 (Carrier Proteins)',\n",
       "      '0 (sulfate transporter)',\n",
       "      '9007-49-2 (DNA)'],\n",
       "     'FAU': ['Camargo, R',\n",
       "      'Limbert, E',\n",
       "      'Gillam, M',\n",
       "      'Henriques, M M',\n",
       "      'Fernandes, C',\n",
       "      'Catarino, A L',\n",
       "      'Soares, J',\n",
       "      'Alves, V A',\n",
       "      'Kopp, P',\n",
       "      'Medeiros-Neto, G']}}]}}"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"molecule\"\n",
    "search_type = \"dfs_query_then_fetch\" # this will use exact document frequencies even for multiple shards\n",
    "body = {\n",
    "  \"query\": {\n",
    "    \"match\" : { \"title-abstract\" : query }\n",
    "  },\n",
    "  \"size\": 10\n",
    "}\n",
    "es.search(index=\"genomics-bool\", search_type=search_type, body=body)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 04.A: _new run and evaluation_\n",
    "Create a new run file (e.g. `boolean.run`), compute the retrieval performance with the function `print_trec_eval()` and compare the results with the baseline run file `baseline.run`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for boolean.run\n",
      "mean success_at_1              0.1316\n",
      "mean success_at_5              0.1316\n",
      "mean success_at_10             0.2368\n",
      "mean r_precision               0.06673\n",
      "mean precision_at_1            0.1316\n",
      "mean precision_at_5            0.03684\n",
      "mean precision_at_10           0.03684\n",
      "mean precision_at_50           0.007368\n",
      "mean precision_at_100          0.003684\n",
      "mean precision_at_recall_00    0.1467\n",
      "mean precision_at_recall_01    0.1467\n",
      "mean precision_at_recall_02    0.1254\n",
      "mean precision_at_recall_03    0.0891\n",
      "mean precision_at_recall_04    0.08033\n",
      "mean precision_at_recall_05    0.08033\n",
      "mean precision_at_recall_06    0.04211\n",
      "mean precision_at_recall_07    0.04211\n",
      "mean precision_at_recall_08    0.02895\n",
      "mean precision_at_recall_09    0.02895\n",
      "mean precision_at_recall_10    0.02895\n",
      "mean average_precision         0.06247\n",
      "Results for baseline.run\n",
      "mean success_at_1              0.1579\n",
      "mean success_at_5              0.2105\n",
      "mean success_at_10             0.3158\n",
      "mean r_precision               0.07989\n",
      "mean precision_at_1            0.1579\n",
      "mean precision_at_5            0.06316\n",
      "mean precision_at_10           0.04737\n",
      "mean precision_at_50           0.009474\n",
      "mean precision_at_100          0.004737\n",
      "mean precision_at_recall_00    0.195\n",
      "mean precision_at_recall_01    0.195\n",
      "mean precision_at_recall_02    0.1687\n",
      "mean precision_at_recall_03    0.1424\n",
      "mean precision_at_recall_04    0.1298\n",
      "mean precision_at_recall_05    0.1298\n",
      "mean precision_at_recall_06    0.07093\n",
      "mean precision_at_recall_07    0.07093\n",
      "mean precision_at_recall_08    0.03333\n",
      "mean precision_at_recall_09    0.03333\n",
      "mean precision_at_recall_10    0.03333\n",
      "mean average_precision         0.08827\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# write your code here\n",
    "# BEGIN ANSWER\n",
    "                \n",
    "# connect to ES server             \n",
    "es = elasticsearch.Elasticsearch('localhost')\n",
    "# Write the results of the queries contained in the topic file `'data/training-queries-simple.txt'` \n",
    "# to the run file `'baseline.run'`, and name this test as `test01`\n",
    "make_trec_run(es, 'data01/FIR-s05-training-queries-simple.txt', 'boolean.run', \"genomics-bool\", run_name='test01')\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'boolean.run')\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'baseline.run')\n",
    "\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 04.B: _Language models_\n",
    "\n",
    "Custom similarities can be configured by tuning the parameters of the built-in similarities. Read more about these (expert) options in the [similarity module](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/index-modules-similarity.html).\n",
    "\n",
    "> Tip: the example similarity settings have to be used in a `\"settings\"` object.\n",
    "> Check your settings and mappings with: `es.indices.get(index='NAME-OF-INDEX')`.\n",
    "\n",
    "__1. Make a run that uses Language Models with [Jelinek-Mercer smoothing](http://lucene.apache.org/core/5_2_1/core/org/apache/lucene/search/similarities/LMJelinekMercerSimilarity.html) (linear interpolation smoothing) on the field `\"all\"` that indexes the fields `\"TI\"` and `\"AB\"`. Use the parameter `lambda=0.2`.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "lmjelinekmercer = {\n",
    "  # BEGIN ANSWER\n",
    "    \"settings\": {\n",
    "    \"number_of_shards\": 1,  # single shard for accurate document frequencies\n",
    "    \"similarity\": {\n",
    "      \"lm_jelinek_mercer\": {\n",
    "        \"type\": \"LMJelinekMercer\",\n",
    "        \"lambda\": 0.2\n",
    "      }\n",
    "    }\n",
    "  },\n",
    "  \"mappings\": {\n",
    "    \"properties\": {\n",
    "      \"AB\": {\n",
    "        \"type\": \"text\",\n",
    "        \"copy_to\": \"title-abstract\",\n",
    "        \"similarity\": \"lm_jelinek_mercer\"\n",
    "      },\n",
    "      \"TI\": {\n",
    "        \"type\": \"text\",\n",
    "        \"copy_to\": \"title-abstract\",\n",
    "        \"similarity\": \"lm_jelinek_mercer\"\n",
    "      },\n",
    "      \"title-abstract\": {\n",
    "        \"type\": \"text\",\n",
    "        \"similarity\": \"lm_jelinek_mercer\"\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  # END ANSWER\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-jm', body=lmjelinekmercer)\n",
    "make_trec_run(es, 'data01/FIR-s05-training-queries-simple.txt', 'lmjelinekmercer.run', 'genomics-jm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__2. Make a run that uses Language Models with [Dirichelet smoothing](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-similarity.html) to index the fields `\"TI\"` and `\"AB\"`. Use the parameter `mu=2000`.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "dirichlet = {\n",
    "  # BEGIN ANSWER\n",
    "    \"settings\": {\n",
    "    \"number_of_shards\": 1,  # single shard for accurate document frequencies\n",
    "    \"similarity\": {\n",
    "      \"lm_dirichlet\": {\n",
    "        \"type\": \"LMDirichlet\",\n",
    "        \"mu\": 2000\n",
    "      }\n",
    "    }\n",
    "  },\n",
    "  \"mappings\": {\n",
    "    \"properties\": {\n",
    "      \"AB\": {\n",
    "        \"type\": \"text\",\n",
    "        \"copy_to\": \"title-abstract\",\n",
    "        \"similarity\": \"lm_dirichlet\"\n",
    "      },\n",
    "      \"TI\": {\n",
    "        \"type\": \"text\",\n",
    "        \"copy_to\": \"title-abstract\",\n",
    "        \"similarity\": \"lm_dirichlet\"\n",
    "      },\n",
    "      \"title-abstract\": {\n",
    "        \"type\": \"text\",\n",
    "        \"similarity\": \"lm_dirichlet\"\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  # END ANSWER\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-dirichlet', body=dirichlet)\n",
    "make_trec_run(es, 'data01/FIR-s05-training-queries-simple.txt', 'dirichlet.run', 'genomics-dirichlet')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 04.C: _Model comparison_\n",
    "\n",
    "\n",
    "__1. Compute the performance results of the `lmjelinekmercer.run` and `dirichelet.run`. Compare them with those of the `baseline.run` and `boolean.run`. Evaluate the runs using the `print_trec_eval` function. Performing statistical tests may help strengthen your claims.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for boolean.run\n",
      "mean success_at_1              0.1316\n",
      "mean success_at_5              0.1316\n",
      "mean success_at_10             0.2368\n",
      "mean r_precision               0.06673\n",
      "mean precision_at_1            0.1316\n",
      "mean precision_at_5            0.03684\n",
      "mean precision_at_10           0.03684\n",
      "mean precision_at_50           0.007368\n",
      "mean precision_at_100          0.003684\n",
      "mean precision_at_recall_00    0.1467\n",
      "mean precision_at_recall_01    0.1467\n",
      "mean precision_at_recall_02    0.1254\n",
      "mean precision_at_recall_03    0.0891\n",
      "mean precision_at_recall_04    0.08033\n",
      "mean precision_at_recall_05    0.08033\n",
      "mean precision_at_recall_06    0.04211\n",
      "mean precision_at_recall_07    0.04211\n",
      "mean precision_at_recall_08    0.02895\n",
      "mean precision_at_recall_09    0.02895\n",
      "mean precision_at_recall_10    0.02895\n",
      "mean average_precision         0.06247\n",
      "Results for baseline.run\n",
      "mean success_at_1              0.1579\n",
      "mean success_at_5              0.2105\n",
      "mean success_at_10             0.3158\n",
      "mean r_precision               0.07989\n",
      "mean precision_at_1            0.1579\n",
      "mean precision_at_5            0.06316\n",
      "mean precision_at_10           0.04737\n",
      "mean precision_at_50           0.009474\n",
      "mean precision_at_100          0.004737\n",
      "mean precision_at_recall_00    0.195\n",
      "mean precision_at_recall_01    0.195\n",
      "mean precision_at_recall_02    0.1687\n",
      "mean precision_at_recall_03    0.1424\n",
      "mean precision_at_recall_04    0.1298\n",
      "mean precision_at_recall_05    0.1298\n",
      "mean precision_at_recall_06    0.07093\n",
      "mean precision_at_recall_07    0.07093\n",
      "mean precision_at_recall_08    0.03333\n",
      "mean precision_at_recall_09    0.03333\n",
      "mean precision_at_recall_10    0.03333\n",
      "mean average_precision         0.08827\n",
      "Results for lmjelinekmercer.run\n",
      "mean success_at_1              0.1316\n",
      "mean success_at_5              0.1842\n",
      "mean success_at_10             0.3158\n",
      "mean r_precision               0.08866\n",
      "mean precision_at_1            0.1316\n",
      "mean precision_at_5            0.06316\n",
      "mean precision_at_10           0.04737\n",
      "mean precision_at_50           0.009474\n",
      "mean precision_at_100          0.004737\n",
      "mean precision_at_recall_00    0.174\n",
      "mean precision_at_recall_01    0.174\n",
      "mean precision_at_recall_02    0.1711\n",
      "mean precision_at_recall_03    0.1447\n",
      "mean precision_at_recall_04    0.1272\n",
      "mean precision_at_recall_05    0.1272\n",
      "mean precision_at_recall_06    0.08443\n",
      "mean precision_at_recall_07    0.08443\n",
      "mean precision_at_recall_08    0.04825\n",
      "mean precision_at_recall_09    0.04825\n",
      "mean precision_at_recall_10    0.04825\n",
      "mean average_precision         0.07826\n",
      "Results for dirichlet.run\n",
      "mean success_at_1              0.1316\n",
      "mean success_at_5              0.2105\n",
      "mean success_at_10             0.2368\n",
      "mean r_precision               0.07989\n",
      "mean precision_at_1            0.1316\n",
      "mean precision_at_5            0.06842\n",
      "mean precision_at_10           0.03947\n",
      "mean precision_at_50           0.007895\n",
      "mean precision_at_100          0.003947\n",
      "mean precision_at_recall_00    0.1814\n",
      "mean precision_at_recall_01    0.1814\n",
      "mean precision_at_recall_02    0.1682\n",
      "mean precision_at_recall_03    0.1331\n",
      "mean precision_at_recall_04    0.1031\n",
      "mean precision_at_recall_05    0.1031\n",
      "mean precision_at_recall_06    0.06031\n",
      "mean precision_at_recall_07    0.06031\n",
      "mean precision_at_recall_08    0.0307\n",
      "mean precision_at_recall_09    0.0307\n",
      "mean precision_at_recall_10    0.0307\n",
      "mean average_precision         0.08169\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# your comments here\n",
    "# BEGIN ANSWER\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'boolean.run')\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'baseline.run')\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'lmjelinekmercer.run')\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'dirichlet.run')\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top20 retrieved documents baseline.run\n",
      "1\tQ0\t11824903\t1\t36.454147\ttest01\n",
      "1\tQ0\t11907576\t2\t28.477657\ttest01\n",
      "1\tQ0\t12167152\t3\t27.794035\ttest01\n",
      "1\tQ0\t11971762\t4\t26.188068\ttest01\n",
      "1\tQ0\t12645611\t5\t24.40221\ttest01\n",
      "1\tQ0\t12393857\t6\t24.205702\ttest01\n",
      "1\tQ0\t12018119\t7\t23.876041\ttest01\n",
      "1\tQ0\t12148460\t8\t23.304388\ttest01\n",
      "1\tQ0\t12399461\t9\t23.044468\ttest01\n",
      "1\tQ0\t12203811\t10\t22.933348\ttest01\n",
      "\n",
      "Top20 retrieved documents boolean.run\n",
      "1\tQ0\t11824903\t1\t5.0\ttest01\n",
      "1\tQ0\t11900595\t2\t4.0\ttest01\n",
      "1\tQ0\t11907576\t3\t4.0\ttest01\n",
      "1\tQ0\t11908956\t4\t4.0\ttest01\n",
      "1\tQ0\t12124912\t5\t4.0\ttest01\n",
      "1\tQ0\t12135477\t6\t4.0\ttest01\n",
      "1\tQ0\t12237935\t7\t4.0\ttest01\n",
      "1\tQ0\t12362432\t8\t4.0\ttest01\n",
      "1\tQ0\t12457850\t9\t4.0\ttest01\n",
      "1\tQ0\t12468090\t10\t4.0\ttest01\n",
      "\n",
      "Top20 retrieved documents lmjelinekmercer.run\n",
      "1\tQ0\t11824903\t1\t41.689125\ttest\n",
      "1\tQ0\t11907576\t2\t30.441002\ttest\n",
      "1\tQ0\t11971762\t3\t26.20203\ttest\n",
      "1\tQ0\t12645611\t4\t24.916845\ttest\n",
      "1\tQ0\t12167152\t5\t24.725616\ttest\n",
      "1\tQ0\t12503676\t6\t23.688574\ttest\n",
      "1\tQ0\t12393857\t7\t23.459831\ttest\n",
      "1\tQ0\t12141135\t8\t23.36098\ttest\n",
      "1\tQ0\t12519948\t9\t23.232643\ttest\n",
      "1\tQ0\t12148460\t10\t23.031034\ttest\n",
      "\n",
      "Top20 retrieved documents dirichlet.run\n",
      "1\tQ0\t11824903\t1\t19.445662\ttest\n",
      "1\tQ0\t11907576\t2\t15.004266\ttest\n",
      "1\tQ0\t11993511\t3\t11.828305\ttest\n",
      "1\tQ0\t11937514\t4\t11.817793\ttest\n",
      "1\tQ0\t12018119\t5\t11.62272\ttest\n",
      "1\tQ0\t12028792\t6\t11.205576\ttest\n",
      "1\tQ0\t12231201\t7\t11.190425\ttest\n",
      "1\tQ0\t11900595\t8\t11.169551\ttest\n",
      "1\tQ0\t12525161\t9\t11.0874605\ttest\n",
      "1\tQ0\t11908956\t10\t11.046291\ttest\n"
     ]
    }
   ],
   "source": [
    "print('Top20 retrieved documents baseline.run')\n",
    "! head -10 baseline.run\n",
    "\n",
    "print('\\nTop20 retrieved documents boolean.run')\n",
    "! head -10 boolean.run\n",
    "\n",
    "print('\\nTop20 retrieved documents lmjelinekmercer.run')\n",
    "! head -10 lmjelinekmercer.run\n",
    "\n",
    "print('\\nTop20 retrieved documents dirichlet.run')\n",
    "! head -10 dirichlet.run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "__2. Provide below your comments and interpretations of the results. Why, in your opinion, one model of search is better than the others?__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "# answer as a comment:\n",
    "# It looks like lmjelinekmercer outperformed the baseline, boolean and dirichlet by a significant margin.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example: _ElasticSearch Analyzers for tokenization_\n",
    "\n",
    "The amount and quality of the tokens used to construct the inverted index are of great importance. In ElasticSearch, mappings and settings also allow specifying what [Analyzer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis.html) is used to tokenize your documents and queries. In the mappings below, use the _Dutch_ analyzer for the field `\"all\"`):\n",
    "\n",
    "> Usually, the same analyzer should be applied to documents and queries, but \n",
    "> Elasticsearch allows you to specify a `\"search_analyzer\"` that is used on \n",
    "> your queries (which we do not need to use in the assignment)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyzer_test = {\n",
    "  \"mappings\": {\n",
    "      \"properties\": {\n",
    "        \"all\": {\n",
    "          \"type\": \"text\",\n",
    "          \"analyzer\": \"dutch\"\n",
    "        }\n",
    "      }\n",
    "  }\n",
    "}\n",
    "\n",
    "# create the index, but don't index any documents:\n",
    "create_index(es, 'test-tokens', body=analyzer_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The analyzer defined for the `\"all\"` field can be tested [as follows](https://elasticsearch-py.readthedocs.io/en/master/api.html#indices). Translated to English the text says: _\"This is a Dutch sentence\"_. \n",
    "\n",
    "> The following script identifies the tokens (based on the use of the dutch tokenizer): try with different tokenizers and different sentences to see how the tokens are created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'tokens': [{'end_offset': 20,\n",
      "             'position': 2,\n",
      "             'start_offset': 9,\n",
      "             'token': 'nederland',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 27,\n",
      "             'position': 3,\n",
      "             'start_offset': 21,\n",
      "             'token': 'zinn',\n",
      "             'type': '<ALPHANUM>'}]}\n"
     ]
    }
   ],
   "source": [
    "from pprint import pprint # pretty print\n",
    "\n",
    "body = { \"field\": \"all\", \"text\": \"dit zijn nederlandse zinnen\"}\n",
    "tokens = es.indices.analyze(index='test-tokens', body=body)\n",
    "pprint(tokens)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BONUS PART: _Implement your own similarity measure_ \n",
    "\n",
    "We have only seen the results of using the analyzer to queries. The analyzer results from the _documents_ are available using the `termvectors()` function, as follows for document `id=3`: (Additionally, we can get overall field statistics, such as the number of documents)\n",
    "\n",
    "> First, index the collection again. While waiting, have a coffee or tea :) \n",
    "\n",
    "> `id=3` refers to the internal document identifiers, so not to the Pubmed identifier.\n",
    "\n",
    "_The bonus exercise is not mandatory. It can compensate for lower grades of other exercises._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'_index': 'genomics-base',\n",
       " '_type': '_doc',\n",
       " '_id': '3',\n",
       " '_version': 1,\n",
       " 'found': True,\n",
       " 'took': 70,\n",
       " 'term_vectors': {'TI': {'field_statistics': {'sum_doc_freq': 3071968,\n",
       "    'doc_count': 263080,\n",
       "    'sum_ttf': 3235174},\n",
       "   'terms': {'against': {'doc_freq': 1744,\n",
       "     'ttf': 1752,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 8}]},\n",
       "    'alignment': {'doc_freq': 76,\n",
       "     'ttf': 79,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 18}]},\n",
       "    'an': {'doc_freq': 11933,\n",
       "     'ttf': 12173,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 19}]},\n",
       "    'and': {'doc_freq': 98220,\n",
       "     'ttf': 114781,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 12}]},\n",
       "    'application': {'doc_freq': 1365,\n",
       "     'ttf': 1367,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 20}]},\n",
       "    'binding': {'doc_freq': 2641,\n",
       "     'ttf': 2750,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 23}]},\n",
       "    'carbonyl': {'doc_freq': 64,\n",
       "     'ttf': 67,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 13}]},\n",
       "    'changes': {'doc_freq': 2488,\n",
       "     'ttf': 2507,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 16}]},\n",
       "    'chemical': {'doc_freq': 820,\n",
       "     'ttf': 827,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 14}]},\n",
       "    'couplings': {'doc_freq': 34,\n",
       "     'ttf': 34,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 11}]},\n",
       "    'dipolar': {'doc_freq': 56,\n",
       "     'ttf': 56,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 10}]},\n",
       "    'direct': {'doc_freq': 780,\n",
       "     'ttf': 783,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 0}]},\n",
       "    'high': {'doc_freq': 3674,\n",
       "     'ttf': 3765,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 4}]},\n",
       "    'maltose': {'doc_freq': 15,\n",
       "     'ttf': 15,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 22}]},\n",
       "    'molecular': {'doc_freq': 3099,\n",
       "     'ttf': 3116,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 5}]},\n",
       "    'of': {'doc_freq': 152723,\n",
       "     'ttf': 203186,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 3}]},\n",
       "    'protein': {'doc_freq': 7047,\n",
       "     'ttf': 7547,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 24}]},\n",
       "    'proteins': {'doc_freq': 1988,\n",
       "     'ttf': 2016,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 7}]},\n",
       "    'refinement': {'doc_freq': 35,\n",
       "     'ttf': 35,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 2}]},\n",
       "    'residual': {'doc_freq': 226,\n",
       "     'ttf': 227,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 9}]},\n",
       "    'shift': {'doc_freq': 200,\n",
       "     'ttf': 200,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 15}]},\n",
       "    'structure': {'doc_freq': 2445,\n",
       "     'ttf': 2481,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 1}]},\n",
       "    'to': {'doc_freq': 27203,\n",
       "     'ttf': 28693,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 21}]},\n",
       "    'upon': {'doc_freq': 249,\n",
       "     'ttf': 250,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 17}]},\n",
       "    'weight': {'doc_freq': 814,\n",
       "     'ttf': 835,\n",
       "     'term_freq': 1,\n",
       "     'tokens': [{'position': 6}]}}}}}"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import elasticsearch\n",
    "es = elasticsearch.Elasticsearch(host='localhost')\n",
    "\n",
    "# index_documents(es, 'data/FIR-05-medline.json', 'genomics-base')\n",
    "\n",
    "es.termvectors(index=\"genomics-base\", id=\"3\", fields=\"TI\", \n",
    "               term_statistics=True, field_statistics=True, offsets=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Implement the BM25 similarity\n",
    "\n",
    "Complete the function `bm25_similarity()` below by implementing the BM25 similarity as described by in Section 11.4.3 of [Manning, Raghavan and Schuetze, Chapter 11](https://nlp.stanford.edu/IR-book/pdf/11prob.pdf). Are you able to replicate the score of ElasitcSearch (9.55)? If not, are you using a different variant of the BM25 model? Provide your comments in plain text."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9.552309228956993"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "import math\n",
    "\n",
    "# math.log(x) computes the logarithm of x\n",
    "\n",
    "def bm25_similarity (query, doc_id):\n",
    "\n",
    "    # Get the query tokens (see above)\n",
    "    query_tokens = es.indices.analyze(index='genomics-base', body={\"field\":\"TI\", \"text\": query})\n",
    "    tokens = query_tokens['tokens']\n",
    "\n",
    "    # Get the term vector for doc_id and the field statistics\n",
    "    term_vector = es.termvectors(index=\"genomics-base\", id=doc_id, fields=\"TI\", \n",
    "                  term_statistics=True, field_statistics=True, offsets=False)\n",
    "    vector = term_vector['term_vectors']['TI']['terms']\n",
    "    f_stats = term_vector['term_vectors']['TI']['field_statistics']\n",
    "\n",
    "    # The answer should sum over 'tokens', check if the tokens exists in the 'vector',\n",
    "    # and if so, add the appropriate value to 'similarity'.\n",
    "    # Tip: add print statements to your code to see what each variable contains.\n",
    "    \n",
    "    similarity = 0\n",
    "\n",
    "    # BEGIN ANSWER\n",
    "    N = f_stats['doc_count']\n",
    "    avgdl = f_stats['sum_ttf'] / f_stats['doc_count']\n",
    "    k1=1.2\n",
    "    b=0.75\n",
    "    for token in tokens:\n",
    "        term = token['token']\n",
    "        if term in vector:\n",
    "            f = vector[term]['term_freq']\n",
    "            df = vector[term]['doc_freq']\n",
    "            \n",
    "            idf = math.log((N - df + 0.5) / (df + 0.5) + 1.0)\n",
    "            numerator = f * (k1 + 1)\n",
    "            denominator = f + k1 * (1 - b + b * (len(vector) / avgdl))\n",
    "            \n",
    "            similarity += idf * (numerator / denominator)\n",
    "    # END ANSWER\n",
    "    return similarity\n",
    "\n",
    "bm25_similarity(\"structure refinement\", 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "# eventual comments here\n",
    "# The score was 9.276"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comments:\n",
    "\n",
    "The score was 9.276 very close to the 9.55.\n",
    "\n",
    "The equation used for the implementation is the following:\n",
    "\n",
    "$$\n",
    " \\text{BM25}(q, d) = \\sum_{i=1}^{n} \\text{IDF}(q_i) \\cdot \\frac{f(q_i, d) \\cdot (k_1 + 1)}{f(q_i, d) + k_1 \\cdot (1 - b + b \\cdot \\frac{|d|}{\\text{avgdl}})}\n",
    "$$\n",
    "\n",
    "with starting parameter b = 0.75 and k1 = 1.5\n",
    "\n",
    "However playing with k1 and reducing it to 1.2 can achieve the exact result  which is 9.552309"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See below the 'reference score' computed by ElasticSearch:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9.552309\n"
     ]
    }
   ],
   "source": [
    "body = {\n",
    "  \"query\": {\n",
    "    \"match\" : { \"TI\" : \"structure refinement\" }\n",
    "  }\n",
    "}\n",
    "explain = es.explain(index=\"genomics-base\", id=\"3\", body=body)\n",
    "print (explain['explanation']['value'])  # BM25 score computed by ElasticSearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
